<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.7"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::ModelGeneVar Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.7 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1ModelGeneVar-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::ModelGeneVar Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute and model the per-gene variances in log-expression data.  
 <a href="classscran_1_1ModelGeneVar.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="ModelGeneVar_8hpp_source.html">ModelGeneVar.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVar_1_1AverageBlockResults.html">AverageBlockResults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of variance modelling without blocks.">Results</a> of variance modelling with average statistics across blocks.  <a href="structscran_1_1ModelGeneVar_1_1AverageBlockResults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVar_1_1BlockResults.html">BlockResults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of variance modelling without blocks.">Results</a> of variance modelling with blocks.  <a href="structscran_1_1ModelGeneVar_1_1BlockResults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameters for variance modelling.  <a href="structscran_1_1ModelGeneVar_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of variance modelling without blocks.">Results</a> of variance modelling without blocks.  <a href="structscran_1_1ModelGeneVar_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a07e39b7f6cb9d3c3e93288b66b9a1a72"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a07e39b7f6cb9d3c3e93288b66b9a1a72">set_span</a> (double s=<a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#a63b8885ad46a87667df93955e683fd98">FitTrendVar::Defaults::span</a>)</td></tr>
<tr class="separator:a07e39b7f6cb9d3c3e93288b66b9a1a72"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e6af4f828d2a3bbe238961d6190c0c5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a4e6af4f828d2a3bbe238961d6190c0c5">set_minimum_mean</a> (double m=<a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#a34486e740e69ef045edb53603852f4e5">FitTrendVar::Defaults::minimum_mean</a>)</td></tr>
<tr class="separator:a4e6af4f828d2a3bbe238961d6190c0c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae1461cc0e4594e07182c2a38d7674b2f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#ae1461cc0e4594e07182c2a38d7674b2f">set_use_fixed_width</a> (bool u=<a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#a4e82a47a7c8c3f499d5dfe67bd12d331">FitTrendVar::Defaults::use_fixed_width</a>)</td></tr>
<tr class="separator:ae1461cc0e4594e07182c2a38d7674b2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d0df9fe08e31cfa4f4d89776fec012e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a7d0df9fe08e31cfa4f4d89776fec012e">set_fixed_width</a> (double f=<a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#ad43c19ee320fd351c2165677060b79d0">FitTrendVar::Defaults::fixed_width</a>)</td></tr>
<tr class="separator:a7d0df9fe08e31cfa4f4d89776fec012e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1ffb19a339e0ea8474ed13c593167334"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a1ffb19a339e0ea8474ed13c593167334">set_minimum_window_count</a> (int c=<a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#ac17dc9c3a9866f92227fe9454a02bc90">FitTrendVar::Defaults::minimum_window_count</a>)</td></tr>
<tr class="separator:a1ffb19a339e0ea8474ed13c593167334"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c7fdd291fd6784010827acc230e35bf"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a6c7fdd291fd6784010827acc230e35bf">set_block_weight_policy</a> (<a class="el" href="namespacescran.html#abcd041d625290609e221552910ae2c34">WeightPolicy</a> w=<a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html#a56250b4747b3eccaa87aa9e6c8fb44b5">Defaults::block_weight_policy</a>)</td></tr>
<tr class="separator:a6c7fdd291fd6784010827acc230e35bf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b1974737ce30bef77140ba22aa7463c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a8b1974737ce30bef77140ba22aa7463c">set_variable_block_weight_parameters</a> (<a class="el" href="structscran_1_1VariableBlockWeightParameters.html">VariableBlockWeightParameters</a> v=<a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html#a5b47cec1ec3cfea85a5d6ce47c8cccfd">Defaults::variable_block_weight_parameters</a>)</td></tr>
<tr class="separator:a8b1974737ce30bef77140ba22aa7463c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4fb3af8f312e745080073bf78089a856"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a4fb3af8f312e745080073bf78089a856">set_num_threads</a> (int n=<a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html#a9ccac88ed4b0762bd8ea3528724d517f">Defaults::num_threads</a>)</td></tr>
<tr class="separator:a4fb3af8f312e745080073bf78089a856"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6032451d3498f053f02c236e7a51261b"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:a6032451d3498f053f02c236e7a51261b"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a6032451d3498f053f02c236e7a51261b">run</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, Stat_ *means, Stat_ *variances, Stat_ *fitted, Stat_ *residuals) const</td></tr>
<tr class="separator:a6032451d3498f053f02c236e7a51261b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06989254f3029205e295f161eef3c227"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:a06989254f3029205e295f161eef3c227"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a06989254f3029205e295f161eef3c227">run_blocked_with_average</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block, std::vector&lt; Stat_ * &gt; means, std::vector&lt; Stat_ * &gt; variances, std::vector&lt; Stat_ * &gt; fitted, std::vector&lt; Stat_ * &gt; residuals, Stat_ *ave_means, Stat_ *ave_variances, Stat_ *ave_fitted, Stat_ *ave_residuals) const</td></tr>
<tr class="separator:a06989254f3029205e295f161eef3c227"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a69a009e8b4ae09974104bce1eba48af7"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:a69a009e8b4ae09974104bce1eba48af7"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a69a009e8b4ae09974104bce1eba48af7">run_blocked</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block, std::vector&lt; Stat_ * &gt; means, std::vector&lt; Stat_ * &gt; variances, std::vector&lt; Stat_ * &gt; fitted, std::vector&lt; Stat_ * &gt; residuals) const</td></tr>
<tr class="separator:a69a009e8b4ae09974104bce1eba48af7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:add2cb49b24a34f81dd44e436088bd2b7"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ &gt; </td></tr>
<tr class="memitem:add2cb49b24a34f81dd44e436088bd2b7"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#add2cb49b24a34f81dd44e436088bd2b7">run</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat) const</td></tr>
<tr class="separator:add2cb49b24a34f81dd44e436088bd2b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae344252fbc72fe47d6adb10960eee363"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ &gt; </td></tr>
<tr class="memitem:ae344252fbc72fe47d6adb10960eee363"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1ModelGeneVar_1_1BlockResults.html">BlockResults</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#ae344252fbc72fe47d6adb10960eee363">run_blocked</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block) const</td></tr>
<tr class="separator:ae344252fbc72fe47d6adb10960eee363"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab94dede3b7e62016b7c90c918d01e6a7"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ &gt; </td></tr>
<tr class="memitem:ab94dede3b7e62016b7c90c918d01e6a7"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1ModelGeneVar_1_1AverageBlockResults.html">AverageBlockResults</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#ab94dede3b7e62016b7c90c918d01e6a7">run_blocked_with_average</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block) const</td></tr>
<tr class="separator:ab94dede3b7e62016b7c90c918d01e6a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Compute and model the per-gene variances in log-expression data. </p>
<p>This scans through a log-transformed normalized expression matrix (e.g., from <code><a class="el" href="classscran_1_1LogNormCounts.html" title="Compute log-normalized expression values.">LogNormCounts</a></code>) and computes per-feature means and variances. It then fits a trend to the variances with respect to the means using <code><a class="el" href="classscran_1_1FitTrendVar.html" title="Fit a mean-variance trend to log-count data.">FitTrendVar</a></code>. We assume that most genes at any given abundance are not highly variable, such that the fitted value of the trend is interpreted as the "uninteresting" variance - this is mostly attributed to technical variation like sequencing noise, but can also represent constitutive biological noise like transcriptional bursting. Under this assumption, the residual can be treated as a quantification of biologically interesting variation, and can be used to identify relevant features for downstream analyses. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a07e39b7f6cb9d3c3e93288b66b9a1a72" name="a07e39b7f6cb9d3c3e93288b66b9a1a72"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a07e39b7f6cb9d3c3e93288b66b9a1a72">&#9670;&#160;</a></span>set_span()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_span </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#a63b8885ad46a87667df93955e683fd98">FitTrendVar::Defaults::span</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>See <code><a class="el" href="classscran_1_1FitTrendVar.html#af90421c977f14a07cbc287f6dd250a11">FitTrendVar::set_span()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="a4e6af4f828d2a3bbe238961d6190c0c5" name="a4e6af4f828d2a3bbe238961d6190c0c5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4e6af4f828d2a3bbe238961d6190c0c5">&#9670;&#160;</a></span>set_minimum_mean()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_minimum_mean </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>m</em> = <code><a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#a34486e740e69ef045edb53603852f4e5">FitTrendVar::Defaults::minimum_mean</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>See <code><a class="el" href="classscran_1_1FitTrendVar.html#ac65ea3b3886bed5c1cb42c6638aeadf8">FitTrendVar::set_minimum_mean()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="ae1461cc0e4594e07182c2a38d7674b2f" name="ae1461cc0e4594e07182c2a38d7674b2f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae1461cc0e4594e07182c2a38d7674b2f">&#9670;&#160;</a></span>set_use_fixed_width()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_use_fixed_width </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>u</em> = <code><a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#a4e82a47a7c8c3f499d5dfe67bd12d331">FitTrendVar::Defaults::use_fixed_width</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">u</td><td>See <code><a class="el" href="classscran_1_1FitTrendVar.html#a609b9a40d30c9d878d7d43ef01f8f478">FitTrendVar::set_use_fixed_width()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="a7d0df9fe08e31cfa4f4d89776fec012e" name="a7d0df9fe08e31cfa4f4d89776fec012e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7d0df9fe08e31cfa4f4d89776fec012e">&#9670;&#160;</a></span>set_fixed_width()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_fixed_width </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>f</em> = <code><a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#ad43c19ee320fd351c2165677060b79d0">FitTrendVar::Defaults::fixed_width</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>See <code><a class="el" href="classscran_1_1FitTrendVar.html#a84961856ccbb988bba8bad076ecb967e">FitTrendVar::set_fixed_width()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="a1ffb19a339e0ea8474ed13c593167334" name="a1ffb19a339e0ea8474ed13c593167334"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1ffb19a339e0ea8474ed13c593167334">&#9670;&#160;</a></span>set_minimum_window_count()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_minimum_window_count </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1FitTrendVar_1_1Defaults.html#ac17dc9c3a9866f92227fe9454a02bc90">FitTrendVar::Defaults::minimum_window_count</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>See <code><a class="el" href="classscran_1_1FitTrendVar.html#acba7ac552070357990d2a4f762890364">FitTrendVar::set_minimum_window_count()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="a6c7fdd291fd6784010827acc230e35bf" name="a6c7fdd291fd6784010827acc230e35bf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6c7fdd291fd6784010827acc230e35bf">&#9670;&#160;</a></span>set_block_weight_policy()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_block_weight_policy </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="namespacescran.html#abcd041d625290609e221552910ae2c34">WeightPolicy</a>&#160;</td>
          <td class="paramname"><em>w</em> = <code><a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html#a56250b4747b3eccaa87aa9e6c8fb44b5">Defaults::block_weight_policy</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">p</td><td>Weighting policy to use for averaging statistics across blocks.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> instance. </dd></dl>

</div>
</div>
<a id="a8b1974737ce30bef77140ba22aa7463c" name="a8b1974737ce30bef77140ba22aa7463c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8b1974737ce30bef77140ba22aa7463c">&#9670;&#160;</a></span>set_variable_block_weight_parameters()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_variable_block_weight_parameters </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structscran_1_1VariableBlockWeightParameters.html">VariableBlockWeightParameters</a>&#160;</td>
          <td class="paramname"><em>v</em> = <code><a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html#a5b47cec1ec3cfea85a5d6ce47c8cccfd">Defaults::variable_block_weight_parameters</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">v</td><td>Parameters for the variable block weights, see <code><a class="el" href="namespacescran.html#a128f43444ef77988f27ae5b91cef6b13">variable_block_weight()</a></code> for more details. Only used when the block weight policy is set to <code>WeightPolicy::VARIABLE</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> instance. </dd></dl>

</div>
</div>
<a id="a4fb3af8f312e745080073bf78089a856" name="a4fb3af8f312e745080073bf78089a856"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4fb3af8f312e745080073bf78089a856">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp; scran::ModelGeneVar::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1ModelGeneVar_1_1Defaults.html#a9ccac88ed4b0762bd8ea3528724d517f">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="a6032451d3498f053f02c236e7a51261b" name="a6032451d3498f053f02c236e7a51261b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6032451d3498f053f02c236e7a51261b">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Stat_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVar::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the mean of each feature. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the variance of each feature. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the fitted value of the trend. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the residual from the trend for each feature.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>means</code>, <code>variances</code>, <code>fitted</code> and <code>residuals</code> are filled with the relevant statistics. </dd></dl>

</div>
</div>
<a id="a06989254f3029205e295f161eef3c227" name="a06989254f3029205e295f161eef3c227"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a06989254f3029205e295f161eef3c227">&#9670;&#160;</a></span>run_blocked_with_average() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVar::run_blocked_with_average </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>residuals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking. The mean and variance of each gene is computed separately for all cells in each block, and a separate trend is fitted to each block to obtain residuals. This ensures that sample and batch effects do not confound the variance estimates.</p>
<p>We also compute the average of each statistic across blocks, using the weighting strategy described in <code>weight_block()</code>. The average residual is particularly useful for feature selection with <code><a class="el" href="classscran_1_1ChooseHVGs.html" title="Choose highly variable genes for downstream analyses.">ChooseHVGs</a></code>.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type to hold the block IDs. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing a 0-based block ID for each cell - see <code><a class="el" href="namespacescran.html#a7f1658794e621bcd27fd33b91422a4a3">tabulate_ids()</a></code> for more details. This can also be a <code>nullptr</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the mean of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the variance of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the fitted value of the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the residual from the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_means</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average mean across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_variances</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average variance across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_fitted</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average fitted value across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_residuals</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average residual across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a69a009e8b4ae09974104bce1eba48af7" name="a69a009e8b4ae09974104bce1eba48af7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a69a009e8b4ae09974104bce1eba48af7">&#9670;&#160;</a></span>run_blocked() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVar::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking. This is similar to <code><a class="el" href="classscran_1_1ModelGeneVar.html#a06989254f3029205e295f161eef3c227">run_blocked_with_average()</a></code> but omits the calculation of the averaged statistics across blocks.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type to hold the block IDs. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing block identifiers - see <code><a class="el" href="namespacescran.html#a7f1658794e621bcd27fd33b91422a4a3">tabulate_ids()</a></code> for more details. This can also be a <code>nullptr</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the mean of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the variance of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the fitted value of the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the residual from the trend for each feature in the corresponding block of cells. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="add2cb49b24a34f81dd44e436088bd2b7" name="add2cb49b24a34f81dd44e436088bd2b7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#add2cb49b24a34f81dd44e436088bd2b7">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html">Results</a> scran::ModelGeneVar::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of variance modelling without blocks.">Results</a></code> object containing the results of the variance modelling. </dd></dl>

</div>
</div>
<a id="ae344252fbc72fe47d6adb10960eee363" name="ae344252fbc72fe47d6adb10960eee363"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae344252fbc72fe47d6adb10960eee363">&#9670;&#160;</a></span>run_blocked() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1ModelGeneVar_1_1BlockResults.html">BlockResults</a> scran::ModelGeneVar::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking, see <code><a class="el" href="classscran_1_1ModelGeneVar.html#a69a009e8b4ae09974104bce1eba48af7">run_blocked()</a></code> for details.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type, to hold the block IDs.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of block identifiers, see <code><a class="el" href="classscran_1_1ModelGeneVar.html#a69a009e8b4ae09974104bce1eba48af7">run_blocked()</a></code> for details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1ModelGeneVar_1_1BlockResults.html" title="Results of variance modelling with blocks.">BlockResults</a></code> object containing the results of the variance modelling in each block. </dd></dl>

</div>
</div>
<a id="ab94dede3b7e62016b7c90c918d01e6a7" name="ab94dede3b7e62016b7c90c918d01e6a7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab94dede3b7e62016b7c90c918d01e6a7">&#9670;&#160;</a></span>run_blocked_with_average() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1ModelGeneVar_1_1AverageBlockResults.html">AverageBlockResults</a> scran::ModelGeneVar::run_blocked_with_average </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking, see <code><a class="el" href="classscran_1_1ModelGeneVar.html#a06989254f3029205e295f161eef3c227">run_blocked_with_average()</a></code> for details.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type, to hold the block IDs.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of block identifiers, see <code><a class="el" href="classscran_1_1ModelGeneVar.html#a69a009e8b4ae09974104bce1eba48af7">run_blocked()</a></code> for details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An <code><a class="el" href="structscran_1_1ModelGeneVar_1_1AverageBlockResults.html" title="Results of variance modelling with average statistics across blocks.">AverageBlockResults</a></code> object containing the results of the variance modelling in each block, along with the average statistics across blocks. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/feature_selection/<a class="el" href="ModelGeneVar_8hpp_source.html">ModelGeneVar.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.7
</small></address>
</body>
</html>
