<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::AssignReferenceClusters Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1AssignReferenceClusters-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::AssignReferenceClusters Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Assign cells to their closest reference cluster.  
 <a href="classscran_1_1AssignReferenceClusters.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="AssignReferenceClusters_8hpp_source.html">AssignReferenceClusters.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter settings.  <a href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-types" name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:a7f3df554269927a6765c6aade796713e"><td class="memTemplParams" colspan="2">template&lt;typename Index , typename Float &gt; </td></tr>
<tr class="memitem:a7f3df554269927a6765c6aade796713e"><td class="memTemplItemLeft" align="right" valign="top">using&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a7f3df554269927a6765c6aade796713e">Prebuilt</a> = std::vector&lt; std::shared_ptr&lt; knncolle::Base&lt; Index, Float &gt; &gt; &gt;</td></tr>
<tr class="separator:a7f3df554269927a6765c6aade796713e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ab295f391ce6a4b3fe7132172c2cf1447"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#ab295f391ce6a4b3fe7132172c2cf1447">set_quantile</a> (double q=<a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#aab7491df01633f6afb701604129983ec">Defaults::quantile</a>)</td></tr>
<tr class="separator:ab295f391ce6a4b3fe7132172c2cf1447"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab38f0aaa22ad854580dd327ed9200ca"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#aab38f0aaa22ad854580dd327ed9200ca">set_num_threads</a> (int n=<a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#a9b5045a586478d126a3cae31250f19b7">Defaults::num_threads</a>)</td></tr>
<tr class="separator:aab38f0aaa22ad854580dd327ed9200ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1bd20ccb640ac8ed19009261f6e02426"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a1bd20ccb640ac8ed19009261f6e02426">set_approximate</a> (int a=<a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#a8c33bb06ce4b8595d837cc57d14f090d">Defaults::approximate</a>)</td></tr>
<tr class="separator:a1bd20ccb640ac8ed19009261f6e02426"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad8025731b479f39402310d92afaa1a17"><td class="memTemplParams" colspan="2">template&lt;typename Index  = int, typename Float , typename Cluster &gt; </td></tr>
<tr class="memitem:ad8025731b479f39402310d92afaa1a17"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a7f3df554269927a6765c6aade796713e">Prebuilt</a>&lt; Index, Float &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#ad8025731b479f39402310d92afaa1a17">build</a> (int ndim, size_t nref, const Float *ref, const Cluster *clusters) const</td></tr>
<tr class="separator:ad8025731b479f39402310d92afaa1a17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a27551b6bdc2e2001773dcabb2298cb02"><td class="memTemplParams" colspan="2">template&lt;typename Index  = int, typename Float , typename Cluster &gt; </td></tr>
<tr class="memitem:a27551b6bdc2e2001773dcabb2298cb02"><td class="memTemplItemLeft" align="right" valign="top">std::vector&lt; Cluster &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a27551b6bdc2e2001773dcabb2298cb02">run</a> (int ndim, size_t nref, const Float *ref, const Cluster *clusters, size_t ntest, const Float *test) const</td></tr>
<tr class="separator:a27551b6bdc2e2001773dcabb2298cb02"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a538b7a34291539c20176177635bdc4f0"><td class="memTemplParams" colspan="2">template&lt;typename Index  = int, typename Float , typename Cluster &gt; </td></tr>
<tr class="memitem:a538b7a34291539c20176177635bdc4f0"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a538b7a34291539c20176177635bdc4f0">run</a> (int ndim, size_t nref, const Float *ref, const Cluster *clusters, size_t ntest, const Float *test, Cluster *output) const</td></tr>
<tr class="separator:a538b7a34291539c20176177635bdc4f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a80532111405f30a7af327c862ccd8763"><td class="memTemplParams" colspan="2">template&lt;typename Index , typename Float , typename Cluster &gt; </td></tr>
<tr class="memitem:a80532111405f30a7af327c862ccd8763"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a80532111405f30a7af327c862ccd8763">run</a> (const <a class="el" href="classscran_1_1AssignReferenceClusters.html#a7f3df554269927a6765c6aade796713e">Prebuilt</a>&lt; Index, Float &gt; &amp;built, size_t ntest, const Float *test, Cluster *output) const</td></tr>
<tr class="separator:a80532111405f30a7af327c862ccd8763"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Assign cells to their closest reference cluster. </p>
<p >This uses a <a href="https://bioconductor.org/packages/SingleR"><b>SingleR</b></a>-like approach to assign a cell to a cluster in a reference dataset. The idea is that we have a test dataset containing cells with no clusters, and a reference dataset containing cluster assignments. For each test cell, we compute a score for each reference cluster, defined as a certain (low) quantile of the distances to all reference cells in that cluster. The test cell is then assigned to the cluster with the lowest quantile.</p>
<p >We use this approach as it adjusts for differences in the number of reference cells in each cluster. For comparison, a simpler approach would be to just define the cluster of a test cell as the majority of the cluster assignments of its neighbors in the reference dataset. This approach would favor assignment of test cells to clusters with more reference cells, simply because it is easier to obtain a majority. Our approach is also better than using the closest centroid of each reference cluster as the quantile accounts for heterogeneity within each cluster; more dispersed clusters are not overly penalized for having many observations far from the centroid.</p>
<p >The reference dataset is usually generated as a subset of the test dataset, e.g., using <code><a class="el" href="classscran_1_1DownsampleByNeighbors.html" title="Downsample a dataset based on its neighbors.">DownsampleByNeighbors</a></code>. In this manner, we can quickly cluster on the subset and then propagate assignments back to the full dataset. Note that this function expects low-dimensional coordinates as input - see <a href="https://ltla.github.io/singlepp"><b>singlepp</b></a> for classification based on the original count matrix. </p>
</div><h2 class="groupheader">Member Typedef Documentation</h2>
<a id="a7f3df554269927a6765c6aade796713e" name="a7f3df554269927a6765c6aade796713e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7f3df554269927a6765c6aade796713e">&#9670;&nbsp;</a></span>Prebuilt</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Index , typename Float &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">using <a class="el" href="classscran_1_1AssignReferenceClusters.html#a7f3df554269927a6765c6aade796713e">scran::AssignReferenceClusters::Prebuilt</a> =  std::vector&lt;std::shared_ptr&lt;knncolle::Base&lt;Index, Float&gt; &gt; &gt;</td>
        </tr>
      </table>
</div><div class="memdoc">
<p >Vector of neighbor search indices for the reference clusters. This has length equal to the number of clusters, where each entry contains a search index for the corresponding cluster in <code>clusters</code> of <code><a class="el" href="classscran_1_1AssignReferenceClusters.html#ad8025731b479f39402310d92afaa1a17">build()</a></code>.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Index</td><td>Integer type for the cell indices. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the distances. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="ad8025731b479f39402310d92afaa1a17" name="ad8025731b479f39402310d92afaa1a17"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad8025731b479f39402310d92afaa1a17">&#9670;&nbsp;</a></span>build()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Index  = int, typename Float , typename Cluster &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1AssignReferenceClusters.html#a7f3df554269927a6765c6aade796713e">Prebuilt</a>&lt; Index, Float &gt; scran::AssignReferenceClusters::build </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>ndim</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>nref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Cluster *&#160;</td>
          <td class="paramname"><em>clusters</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Build the search indices from the reference dataset, to be used for cluster assignment in <code><a class="el" href="classscran_1_1AssignReferenceClusters.html#a27551b6bdc2e2001773dcabb2298cb02">run()</a></code>.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Index</td><td>Integer type for the cell indices. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the distances. </td></tr>
    <tr><td class="paramname">Cluster</td><td>Integer type for the cluster assignments.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">ndim</td><td>Number of dimensions. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">nref</td><td>Number of cells in the reference dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref</td><td>Pointer to a column-major array of coordinates of dimensions (rows) and cells (columns) for the reference dataset. This should be a low-dimensional embedding, e.g., from <code><a class="el" href="classscran_1_1RunPCA.html" title="Perform PCA on a gene-cell matrix.">RunPCA</a></code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">clusters</td><td>Pointer to an array of length equal to <code>nref</code>, containing the cluster assignment for each cell in the reference. All integers in $[0, N)$ should be present at least once, where $N$ is the total number of unique clusters.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The prebuilt indices, for use in <code><a class="el" href="classscran_1_1AssignReferenceClusters.html#a27551b6bdc2e2001773dcabb2298cb02">run()</a></code>. </dd></dl>

</div>
</div>
<a id="a80532111405f30a7af327c862ccd8763" name="a80532111405f30a7af327c862ccd8763"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a80532111405f30a7af327c862ccd8763">&#9670;&nbsp;</a></span>run() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Index , typename Float , typename Cluster &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::AssignReferenceClusters::run </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classscran_1_1AssignReferenceClusters.html#a7f3df554269927a6765c6aade796713e">Prebuilt</a>&lt; Index, Float &gt; &amp;&#160;</td>
          <td class="paramname"><em>built</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ntest</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float *&#160;</td>
          <td class="paramname"><em>test</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Cluster *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Assign cells in the test dataset to reference clusters based on pre-built search indices.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Index</td><td>Integer type for the cell indices. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the distances. </td></tr>
    <tr><td class="paramname">Cluster</td><td>Integer type for the cluster assignments.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">built</td><td>Prebuilt search indices created by <code><a class="el" href="classscran_1_1AssignReferenceClusters.html#ad8025731b479f39402310d92afaa1a17">build()</a></code>. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ntest</td><td>Number of cells in the test dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">test</td><td>Pointer to a column-major array of coordinates of dimensions (rows) and cells (columns) for the test dataset. This should be a low-dimensional embedding in the same space as that used to construct <code>built</code>. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array of length equal to <code>ntest</code>. On output, this is filled with the cluster assignments for each cell in the test dataset. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a27551b6bdc2e2001773dcabb2298cb02" name="a27551b6bdc2e2001773dcabb2298cb02"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a27551b6bdc2e2001773dcabb2298cb02">&#9670;&nbsp;</a></span>run() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Index  = int, typename Float , typename Cluster &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; Cluster &gt; scran::AssignReferenceClusters::run </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>ndim</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>nref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Cluster *&#160;</td>
          <td class="paramname"><em>clusters</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ntest</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float *&#160;</td>
          <td class="paramname"><em>test</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Assign cells in the test dataset to reference clusters.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Index</td><td>Integer type for the cell indices. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the distances. </td></tr>
    <tr><td class="paramname">Cluster</td><td>Integer type for the cluster assignments.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">ndim</td><td>Number of dimensions. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">nref</td><td>Number of cells in the reference dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref</td><td>Pointer to a column-major array of coordinates of dimensions (rows) and cells (columns) for the reference dataset. This should be a low-dimensional embedding, e.g., from <code><a class="el" href="classscran_1_1RunPCA.html" title="Perform PCA on a gene-cell matrix.">RunPCA</a></code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">clusters</td><td>Pointer to an array of length equal to <code>nref</code>, containing the cluster assignment for each cell in the reference. All integers in $[0, N)$ should be present at least once, where $N$ is the total number of unique clusters. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ntest</td><td>Number of cells in the test dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">test</td><td>Pointer to a column-major array of coordinates of dimensions (rows) and cells (columns) for the test dataset. This should be a low-dimensional embedding in the same space as <code>ref</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Vector of cluster assignments for each cell in the test dataset. </dd></dl>

</div>
</div>
<a id="a538b7a34291539c20176177635bdc4f0" name="a538b7a34291539c20176177635bdc4f0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a538b7a34291539c20176177635bdc4f0">&#9670;&nbsp;</a></span>run() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Index  = int, typename Float , typename Cluster &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::AssignReferenceClusters::run </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>ndim</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>nref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float *&#160;</td>
          <td class="paramname"><em>ref</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Cluster *&#160;</td>
          <td class="paramname"><em>clusters</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ntest</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Float *&#160;</td>
          <td class="paramname"><em>test</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Cluster *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Assign cells in the test dataset to reference clusters.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Index</td><td>Integer type for the cell indices. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the distances. </td></tr>
    <tr><td class="paramname">Cluster</td><td>Integer type for the cluster assignments.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">ndim</td><td>Number of dimensions. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">nref</td><td>Number of cells in the reference dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">ref</td><td>Pointer to a column-major array of coordinates of dimensions (rows) and cells (columns) for the reference dataset. This should be a low-dimensional embedding, e.g., from <code><a class="el" href="classscran_1_1RunPCA.html" title="Perform PCA on a gene-cell matrix.">RunPCA</a></code>. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">clusters</td><td>Pointer to an array of length equal to <code>nref</code>, containing the cluster assignment for each cell in the reference. All integers in $[0, N)$ should be present at least once, where $N$ is the total number of unique clusters. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ntest</td><td>Number of cells in the test dataset. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">test</td><td>Pointer to a column-major array of coordinates of dimensions (rows) and cells (columns) for the test dataset. This should be a low-dimensional embedding in the same space as <code>ref</code>. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array of length equal to <code>ntest</code>. On output, this is filled with the cluster assignments for each cell in the test dataset. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a1bd20ccb640ac8ed19009261f6e02426" name="a1bd20ccb640ac8ed19009261f6e02426"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1bd20ccb640ac8ed19009261f6e02426">&#9670;&nbsp;</a></span>set_approximate()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a> &amp; scran::AssignReferenceClusters::set_approximate </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>a</em> = <code><a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#a8c33bb06ce4b8595d837cc57d14f090d">Defaults::approximate</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>Whether approximate neighbor detection should be used.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1DownsampleByNeighbors.html" title="Downsample a dataset based on its neighbors.">DownsampleByNeighbors</a></code> object. </dd></dl>

</div>
</div>
<a id="aab38f0aaa22ad854580dd327ed9200ca" name="aab38f0aaa22ad854580dd327ed9200ca"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aab38f0aaa22ad854580dd327ed9200ca">&#9670;&nbsp;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a> &amp; scran::AssignReferenceClusters::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#a9b5045a586478d126a3cae31250f19b7">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use for neighbor detection.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1DownsampleByNeighbors.html" title="Downsample a dataset based on its neighbors.">DownsampleByNeighbors</a></code> object. </dd></dl>

</div>
</div>
<a id="ab295f391ce6a4b3fe7132172c2cf1447" name="ab295f391ce6a4b3fe7132172c2cf1447"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab295f391ce6a4b3fe7132172c2cf1447">&#9670;&nbsp;</a></span>set_quantile()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1AssignReferenceClusters.html">AssignReferenceClusters</a> &amp; scran::AssignReferenceClusters::set_quantile </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>q</em> = <code><a class="el" href="structscran_1_1AssignReferenceClusters_1_1Defaults.html#aab7491df01633f6afb701604129983ec">Defaults::quantile</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">q</td><td>Quantile to use for creating a per-cluster score. Smaller values focus more on the local neighborhood around each test cell, while larger values focus on the behavior of the bulk of the cluster.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1DownsampleByNeighbors.html" title="Downsample a dataset based on its neighbors.">DownsampleByNeighbors</a></code> object. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/clustering/<a class="el" href="AssignReferenceClusters_8hpp_source.html">AssignReferenceClusters.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
