<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::ModelGeneVar Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1ModelGeneVar-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">scran::ModelGeneVar Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Compute and model the per-gene variances in log-expression data.  
 <a href="classscran_1_1ModelGeneVar.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="ModelGeneVar_8hpp_source.html">ModelGeneVar.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of the variance modelling.">Results</a> of the variance modelling.  <a href="structscran_1_1ModelGeneVar_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a22d38f7f8a946a62ef79d3684bc6f196"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a22d38f7f8a946a62ef79d3684bc6f196">set_span</a> ()</td></tr>
<tr class="separator:a22d38f7f8a946a62ef79d3684bc6f196"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0caf12aaa24dc0ff5c9fb59eeaec896c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a0caf12aaa24dc0ff5c9fb59eeaec896c">set_span</a> (double s)</td></tr>
<tr class="separator:a0caf12aaa24dc0ff5c9fb59eeaec896c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a55af65ee9cbc05276637b8cef421314e"><td class="memTemplParams" colspan="2">template&lt;class MAT , typename BPTR &gt; </td></tr>
<tr class="memitem:a55af65ee9cbc05276637b8cef421314e"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#a55af65ee9cbc05276637b8cef421314e">run</a> (const MAT *mat, BPTR block)</td></tr>
<tr class="separator:a55af65ee9cbc05276637b8cef421314e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab3c138ffa759e22669fadd628c2bddd"><td class="memTemplParams" colspan="2">template&lt;class MAT , typename BPTR &gt; </td></tr>
<tr class="memitem:aab3c138ffa759e22669fadd628c2bddd"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVar.html#aab3c138ffa759e22669fadd628c2bddd">run</a> (const MAT *p, BPTR block, std::vector&lt; double * &gt; means, std::vector&lt; double * &gt; variances, std::vector&lt; double * &gt; fitted, std::vector&lt; double * &gt; residuals)</td></tr>
<tr class="separator:aab3c138ffa759e22669fadd628c2bddd"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Compute and model the per-gene variances in log-expression data. </p>
<p>This scans through a log-transformed normalized expression matrix (e.g., from <code><a class="el" href="classscran_1_1LogNormCounts.html" title="Compute log-normalized expression values.">LogNormCounts</a></code>) and computes per-feature means and variances. It then fits a trend to the variances with respect to the means using <code><a class="el" href="classscran_1_1FitTrendVar.html" title="Fit a mean-variance trend to log-count data.">FitTrendVar</a></code>. We assume that most genes at any given abundance are not highly variable, such that the fitted value of the trend is interpreted as the "uninteresting" variance - this is mostly attributed to technical variation like sequencing noise, but can also represent constitutive biological noise like transcriptional bursting. Under this assumption, the residual can be treated as a quantification of biologically interesting variation, and can be used to identify relevant features for downstream analyses. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a55af65ee9cbc05276637b8cef421314e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a55af65ee9cbc05276637b8cef421314e">&#9670;&nbsp;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class MAT , typename BPTR &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html">Results</a> scran::ModelGeneVar::run </td>
          <td>(</td>
          <td class="paramtype">const MAT *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">BPTR&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">MAT</td><td>Type of matrix, usually a <code>tatami::NumericMatrix</code>. </td></tr>
    <tr><td class="paramname">BPTR</td><td>Pointer to an integer type, to hold the block IDs.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of block identifiers. If provided, the array should be of length equal to <code>ncells</code>. Values should be integer IDs in <img class="formulaInl" alt="$[0, N)$" src="form_0.png"/> where <img class="formulaInl" alt="$N$" src="form_1.png"/> is the number of blocks. This can also be a <code>nullptr</code>, in which case all cells are assumed to belong to the same block.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of the variance modelling.">Results</a></code> object containing the results of the variance modelling. </dd></dl>

</div>
</div>
<a id="aab3c138ffa759e22669fadd628c2bddd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aab3c138ffa759e22669fadd628c2bddd">&#9670;&nbsp;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class MAT , typename BPTR &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVar::run </td>
          <td>(</td>
          <td class="paramtype">const MAT *&#160;</td>
          <td class="paramname"><em>p</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">BPTR&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt;&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt;&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt;&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; double * &gt;&#160;</td>
          <td class="paramname"><em>residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">MAT</td><td>Type of matrix, usually a <code>tatami::NumericMatrix</code>. </td></tr>
    <tr><td class="paramname">BPTR</td><td>Pointer to an integer type, to hold the block IDs.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of block identifiers, see <code><a class="el" href="classscran_1_1ModelGeneVar.html#a55af65ee9cbc05276637b8cef421314e">run()</a></code> for details. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the mean of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the variance of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the fitted value of the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the residual from the trend for each feature in the corresponding block of cells.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>means</code>, <code>variances</code>, <code>fitted</code> and <code>residuals</code> are filled with the relevant statistics. </dd></dl>

</div>
</div>
<a id="a22d38f7f8a946a62ef79d3684bc6f196"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a22d38f7f8a946a62ef79d3684bc6f196">&#9670;&nbsp;</a></span>set_span() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a>&amp; scran::ModelGeneVar::set_span </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Use the default span for the LOWESS smoother, see <a href="https://ltla.github.io/CppWeightedLowess">https://ltla.github.io/CppWeightedLowess</a> for details.</p>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<a id="a0caf12aaa24dc0ff5c9fb59eeaec896c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0caf12aaa24dc0ff5c9fb59eeaec896c">&#9670;&nbsp;</a></span>set_span() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVar.html">ModelGeneVar</a>&amp; scran::ModelGeneVar::set_span </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>s</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Set the span for <code><a class="el" href="classscran_1_1FitTrendVar.html" title="Fit a mean-variance trend to log-count data.">FitTrendVar</a></code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Span for the smoother, as a proportion of the total number of points.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code> object. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/feature_selection/<a class="el" href="ModelGeneVar_8hpp_source.html">ModelGeneVar.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
