<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::PerCellQCMetrics&lt; X &gt; Class Template Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1PerCellQCMetrics.html">PerCellQCMetrics</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1PerCellQCMetrics-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">scran::PerCellQCMetrics&lt; X &gt; Class Template Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Compute typical per-cell quality control metrics.  
 <a href="classscran_1_1PerCellQCMetrics.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="PerCellQCMetrics_8hpp_source.html">PerCellQCMetrics.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1PerCellQCMetrics_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Result store for QC metric calculations.  <a href="structscran_1_1PerCellQCMetrics_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a3805947ce4bf8d355d79d375c2a1eaf6"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQCMetrics.html#a3805947ce4bf8d355d79d375c2a1eaf6">PerCellQCMetrics</a> ()</td></tr>
<tr class="separator:a3805947ce4bf8d355d79d375c2a1eaf6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae7f6e6f6a45b0909bec33df36cbaf77d"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQCMetrics.html#ae7f6e6f6a45b0909bec33df36cbaf77d">PerCellQCMetrics</a> (const std::vector&lt; X * &gt; &amp;s)</td></tr>
<tr class="separator:ae7f6e6f6a45b0909bec33df36cbaf77d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab875f8d10b1265311d2d62a5a8bac8dd"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQCMetrics.html#ab875f8d10b1265311d2d62a5a8bac8dd">PerCellQCMetrics</a> (std::vector&lt; const X * &gt; s)</td></tr>
<tr class="separator:ab875f8d10b1265311d2d62a5a8bac8dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a945a454e04bfffb66b19e02027eaed71"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQCMetrics.html">PerCellQCMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQCMetrics.html#a945a454e04bfffb66b19e02027eaed71">set_subsets</a> (std::vector&lt; const X * &gt; s)</td></tr>
<tr class="separator:a945a454e04bfffb66b19e02027eaed71"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afad82e4abf25bcab9dd02b745b0cbb4a"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX &gt; </td></tr>
<tr class="memitem:afad82e4abf25bcab9dd02b745b0cbb4a"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1PerCellQCMetrics_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQCMetrics.html#afad82e4abf25bcab9dd02b745b0cbb4a">run</a> (const tatami::typed_matrix&lt; T, IDX &gt; *mat)</td></tr>
<tr class="separator:afad82e4abf25bcab9dd02b745b0cbb4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a151571be33c547178faa9e0c1eaecf29"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX , typename S , typename D , typename P &gt; </td></tr>
<tr class="memitem:a151571be33c547178faa9e0c1eaecf29"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQCMetrics.html#a151571be33c547178faa9e0c1eaecf29">run</a> (const tatami::typed_matrix&lt; T, IDX &gt; *mat, S *sums, D *detected, std::vector&lt; P * &gt; subset_proportions)</td></tr>
<tr class="separator:a151571be33c547178faa9e0c1eaecf29"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h3>template&lt;typename X = uint8_t&gt;<br />
class scran::PerCellQCMetrics&lt; X &gt;</h3>

<p>Compute typical per-cell quality control metrics. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">X</td><td>Type of the boolean flag to indicate whether a feature belongs to a feature subset.</td></tr>
  </table>
  </dd>
</dl>
<p>Given a feature-by-cell count matrix, this class computes several QC metrics:</p>
<ul>
<li>The total sum for each cell quantifies the efficiency of library preparation and sequencing.</li>
<li>The number of detected features also quantifies the library preparation efficiency, but with a greater focus on capturing the transcriptional complexity.</li>
<li>The interpretation of the subset proportions depend on the subsets. In the most common case of mitochondrial transcripts, higher proportions indicate cell damage. Spike-in proportions can be interpreted in a similar manner. </li>
</ul>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a3805947ce4bf8d355d79d375c2a1eaf6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3805947ce4bf8d355d79d375c2a1eaf6">&#9670;&nbsp;</a></span>PerCellQCMetrics() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename X  = uint8_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQCMetrics.html">scran::PerCellQCMetrics</a>&lt; X &gt;::<a class="el" href="classscran_1_1PerCellQCMetrics.html">PerCellQCMetrics</a> </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Default constructor. </p>

</div>
</div>
<a id="ae7f6e6f6a45b0909bec33df36cbaf77d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae7f6e6f6a45b0909bec33df36cbaf77d">&#9670;&nbsp;</a></span>PerCellQCMetrics() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename X  = uint8_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQCMetrics.html">scran::PerCellQCMetrics</a>&lt; X &gt;::<a class="el" href="classscran_1_1PerCellQCMetrics.html">PerCellQCMetrics</a> </td>
          <td>(</td>
          <td class="paramtype">const std::vector&lt; X * &gt; &amp;&#160;</td>
          <td class="paramname"><em>s</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>A vector of pointers to boolean arrays. Each array represents a feature subset and is of length equal to the number of features. Each value should be non-zero if the corresponding feature belongs to the subset; otherwise, it should be zero. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ab875f8d10b1265311d2d62a5a8bac8dd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab875f8d10b1265311d2d62a5a8bac8dd">&#9670;&nbsp;</a></span>PerCellQCMetrics() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename X  = uint8_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQCMetrics.html">scran::PerCellQCMetrics</a>&lt; X &gt;::<a class="el" href="classscran_1_1PerCellQCMetrics.html">PerCellQCMetrics</a> </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; const X * &gt;&#160;</td>
          <td class="paramname"><em>s</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>A vector of pointers to boolean arrays. Each array represents a feature subset and is of length equal to the number of features. Each value should be non-zero if the corresponding feature belongs to the subset; otherwise, it should be zero. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="afad82e4abf25bcab9dd02b745b0cbb4a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afad82e4abf25bcab9dd02b745b0cbb4a">&#9670;&nbsp;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename X  = uint8_t&gt; </div>
<div class="memtemplate">
template&lt;typename T , typename IDX &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1PerCellQCMetrics_1_1Results.html">Results</a> <a class="el" href="classscran_1_1PerCellQCMetrics.html">scran::PerCellQCMetrics</a>&lt; X &gt;::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::typed_matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute the QC metrics from an input matrix and return the results.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>Pointer to a feature-by-cells matrix.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1PerCellQCMetrics_1_1Results.html" title="Result store for QC metric calculations.">PerCellQCMetrics::Results</a></code> object containing the QC metrics. Subset proportions are returned depending on the subsets defined at construction or by <code><a class="el" href="classscran_1_1PerCellQCMetrics.html#a945a454e04bfffb66b19e02027eaed71">set_subsets()</a></code>. </dd></dl>

</div>
</div>
<a id="a151571be33c547178faa9e0c1eaecf29"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a151571be33c547178faa9e0c1eaecf29">&#9670;&nbsp;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename X  = uint8_t&gt; </div>
<div class="memtemplate">
template&lt;typename T , typename IDX , typename S , typename D , typename P &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void <a class="el" href="classscran_1_1PerCellQCMetrics.html">scran::PerCellQCMetrics</a>&lt; X &gt;::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::typed_matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">S *&#160;</td>
          <td class="paramname"><em>sums</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">D *&#160;</td>
          <td class="paramname"><em>detected</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; P * &gt;&#160;</td>
          <td class="paramname"><em>subset_proportions</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute the QC metrics from an input matrix and return the results.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells matrix. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">sums</td><td>Pointer to a (typically double-precision) array of length equal to the number of columns in <code>mat</code>. This is used to store the computed sums for all cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">detected</td><td>Pointer to a (typically integer) array of length equal to the number of columns in <code>mat</code>. This is used to store the number of detected features for all cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">subset_proportions</td><td>Vector of pointers (typically double-precision) arrays of length equal to the number of columns in <code>mat</code>. Each array corresponds to a feature subset and is used to store the proportion of counts in that subset across all cells.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>sums</code>, <code>detected</code>, and each array in <code>subset_proportions</code> is filled with the relevant statistics. </dd></dl>

</div>
</div>
<a id="a945a454e04bfffb66b19e02027eaed71"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a945a454e04bfffb66b19e02027eaed71">&#9670;&nbsp;</a></span>set_subsets()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename X  = uint8_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQCMetrics.html">PerCellQCMetrics</a>&amp; <a class="el" href="classscran_1_1PerCellQCMetrics.html">scran::PerCellQCMetrics</a>&lt; X &gt;::set_subsets </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; const X * &gt;&#160;</td>
          <td class="paramname"><em>s</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Define feature subsets of interest. This is most commonly used to define mitochondrial transcripts or ribosomal genes.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>A vector of pointers to boolean arrays indicating which features belong in which feature subsets. Each array represents a feature subset and is of length equal to the number of features. Each value should be non-zero if the corresponding feature belongs to the subset; otherwise, it should be zero.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1PerCellQCMetrics.html" title="Compute typical per-cell quality control metrics.">PerCellQCMetrics</a></code> object. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/quality_control/<a class="el" href="PerCellQCMetrics_8hpp_source.html">PerCellQCMetrics.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
