<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.5"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::BlockedPCA Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.5 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1BlockedPCA-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::BlockedPCA Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute PCA after blocking on an uninteresting factor.  
 <a href="classscran_1_1BlockedPCA.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="BlockedPCA_8hpp_source.html">BlockedPCA.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter settings.  <a href="structscran_1_1BlockedPCA_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Container for the PCA results.  <a href="structscran_1_1BlockedPCA_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aaab4ee71a29b1ac787a1a135719217d4"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BlockedPCA.html#aaab4ee71a29b1ac787a1a135719217d4">set_rank</a> (int r=<a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#a3f57557595e7081283a7ab177b7c8058">Defaults::rank</a>)</td></tr>
<tr class="separator:aaab4ee71a29b1ac787a1a135719217d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8ead45b90f2330baa950279a62262912"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BlockedPCA.html#a8ead45b90f2330baa950279a62262912">set_scale</a> (bool s=<a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#a07e46e180a76384576741323a456e2bf">Defaults::scale</a>)</td></tr>
<tr class="separator:a8ead45b90f2330baa950279a62262912"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a72c4787d4bf3345a269fa6a00e1aaa6c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BlockedPCA.html#a72c4787d4bf3345a269fa6a00e1aaa6c">set_transpose</a> (bool t=<a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#ad1782b5c77f985c57e8aa85ffd555a48">Defaults::transpose</a>)</td></tr>
<tr class="separator:a72c4787d4bf3345a269fa6a00e1aaa6c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0996bfb248a6a56f901b03877ce99c99"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BlockedPCA.html#a0996bfb248a6a56f901b03877ce99c99">set_num_threads</a> (int n=<a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#afea8e136a107c82c080c9cdfede04306">Defaults::num_threads</a>)</td></tr>
<tr class="separator:a0996bfb248a6a56f901b03877ce99c99"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7da1c5301c74328fc6f74d5772f760f5"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX , typename Block &gt; </td></tr>
<tr class="memitem:a7da1c5301c74328fc6f74d5772f760f5"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1BlockedPCA.html#a7da1c5301c74328fc6f74d5772f760f5">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Block *block) const</td></tr>
<tr class="separator:a7da1c5301c74328fc6f74d5772f760f5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0b961f19e04cbe6c80f1ada9ff44283a"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX , typename Block , typename X &gt; </td></tr>
<tr class="memitem:a0b961f19e04cbe6c80f1ada9ff44283a"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1BlockedPCA.html#a0b961f19e04cbe6c80f1ada9ff44283a">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Block *block, const X *features) const</td></tr>
<tr class="separator:a0b961f19e04cbe6c80f1ada9ff44283a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Compute PCA after blocking on an uninteresting factor. </p>
<p >A simple batch correction method involves centering the expression of each gene in each batch to remove systematic differences between batches. The corrected values are then used in PCA to obtain a batch-corrected low-dimensional representation of the dataset. Unfortunately, naively centering the expression values will discard sparsity and reduce the computational efficiency of the PCA. To avoid these drawbacks, <code><a class="el" href="classscran_1_1BlockedPCA.html" title="Compute PCA after blocking on an uninteresting factor.">BlockedPCA</a></code> defers the centering until the matrix multiplication of the IRLBA step. This yields the same results as the naive approach but is much faster as it can take advantage of efficient sparse operations. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a7da1c5301c74328fc6f74d5772f760f5" name="a7da1c5301c74328fc6f74d5772f760f5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7da1c5301c74328fc6f74d5772f760f5">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T , typename IDX , typename Block &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html">Results</a> scran::BlockedPCA::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block *&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Run the blocked PCA on an input gene-by-cell matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Floating point type for the data. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer type for the indices. </td></tr>
    <tr><td class="paramname">Block</td><td>Integer type for the blocking factor.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">mat</td><td>Pointer to the input matrix. Columns should contain cells while rows should contain genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells. This should contain the blocking factor as 0-based block assignments (i.e., for <code>n</code> blocks, block identities should run from 0 to <code>n-1</code> with at least one entry for each block.)</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html" title="Container for the PCA results.">Results</a></code> object containing the PCs and the variance explained. </dd></dl>

</div>
</div>
<a id="a0b961f19e04cbe6c80f1ada9ff44283a" name="a0b961f19e04cbe6c80f1ada9ff44283a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0b961f19e04cbe6c80f1ada9ff44283a">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T , typename IDX , typename Block , typename X &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html">Results</a> scran::BlockedPCA::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const X *&#160;</td>
          <td class="paramname"><em>features</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Run the blocked PCA on an input gene-by-cell matrix after filtering for genes of interest. We typically use the set of highly variable genes from <code><a class="el" href="classscran_1_1ChooseHVGs.html" title="Choose highly variable genes for downstream analyses.">ChooseHVGs</a></code>, with the aim being to improve computational efficiency and avoid random noise by removing lowly variable genes.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Floating point type for the data. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer type for the indices. </td></tr>
    <tr><td class="paramname">Block</td><td>Integer type for the blocking factor. </td></tr>
    <tr><td class="paramname">X</td><td>Integer type for the feature filter.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">mat</td><td>Pointer to the input matrix. Columns should contain cells while rows should contain genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells. This should contain the blocking factor as 0-based block assignments (i.e., for <code>n</code> blocks, block identities should run from 0 to <code>n-1</code> with at least one entry for each block.) </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">features</td><td>Pointer to an array of length equal to the number of genes. Each entry treated as a boolean specifying whether the corresponding genes should be used in the PCA.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1BlockedPCA_1_1Results.html" title="Container for the PCA results.">Results</a></code> object containing the PCs and the variance explained. </dd></dl>

</div>
</div>
<a id="a0996bfb248a6a56f901b03877ce99c99" name="a0996bfb248a6a56f901b03877ce99c99"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0996bfb248a6a56f901b03877ce99c99">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp; scran::BlockedPCA::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#afea8e136a107c82c080c9cdfede04306">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BlockedPCA.html" title="Compute PCA after blocking on an uninteresting factor.">BlockedPCA</a></code> instance. </dd></dl>

</div>
</div>
<a id="aaab4ee71a29b1ac787a1a135719217d4" name="aaab4ee71a29b1ac787a1a135719217d4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaab4ee71a29b1ac787a1a135719217d4">&#9670;&#160;</a></span>set_rank()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp; scran::BlockedPCA::set_rank </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>r</em> = <code><a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#a3f57557595e7081283a7ab177b7c8058">Defaults::rank</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">r</td><td>Number of PCs to compute. This should be smaller than the smaller dimension of the input matrix.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BlockedPCA.html" title="Compute PCA after blocking on an uninteresting factor.">BlockedPCA</a></code> instance. </dd></dl>

</div>
</div>
<a id="a8ead45b90f2330baa950279a62262912" name="a8ead45b90f2330baa950279a62262912"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8ead45b90f2330baa950279a62262912">&#9670;&#160;</a></span>set_scale()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp; scran::BlockedPCA::set_scale </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#a07e46e180a76384576741323a456e2bf">Defaults::scale</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Should genes be scaled to unit variance?</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BlockedPCA.html" title="Compute PCA after blocking on an uninteresting factor.">BlockedPCA</a></code> instance. </dd></dl>

</div>
</div>
<a id="a72c4787d4bf3345a269fa6a00e1aaa6c" name="a72c4787d4bf3345a269fa6a00e1aaa6c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a72c4787d4bf3345a269fa6a00e1aaa6c">&#9670;&#160;</a></span>set_transpose()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BlockedPCA.html">BlockedPCA</a> &amp; scran::BlockedPCA::set_transpose </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>t</em> = <code><a class="el" href="structscran_1_1BlockedPCA_1_1Defaults.html#ad1782b5c77f985c57e8aa85ffd555a48">Defaults::transpose</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">t</td><td>Should the PC matrix be transposed on output? If <code>true</code>, the output PC matrix is column-major with cells in the columns, which is compatible with downstream <b>libscran</b> steps.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BlockedPCA.html" title="Compute PCA after blocking on an uninteresting factor.">BlockedPCA</a></code> instance. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/dimensionality_reduction/<a class="el" href="BlockedPCA_8hpp_source.html">BlockedPCA.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.5
</small></address>
</body>
</html>
