<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::PerCellQcMetrics Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1PerCellQcMetrics-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::PerCellQcMetrics Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute a variety of per-cell quality control metrics from a count matrix.  
 <a href="classscran_1_1PerCellQcMetrics.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="PerCellQcMetrics_8hpp_source.html">PerCellQcMetrics.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Buffers.html">Buffers</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Buffers.html" title="Buffers for direct storage of the calculated statistics.">Buffers</a> for direct storage of the calculated statistics.  <a href="structscran_1_1PerCellQcMetrics_1_1Buffers.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameters.  <a href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Result store for QC metric calculations.  <a href="structscran_1_1PerCellQcMetrics_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a147354fd4a4b932454f4ce66c7565c19" id="r_a147354fd4a4b932454f4ce66c7565c19"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a147354fd4a4b932454f4ce66c7565c19">set_compute_total</a> (bool s=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#ac70e65816427d11c645e99454f49522f">Defaults::compute_total</a>)</td></tr>
<tr class="separator:a147354fd4a4b932454f4ce66c7565c19"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a83bddaf8a6650ba2cfccbc83fa88d83e" id="r_a83bddaf8a6650ba2cfccbc83fa88d83e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a83bddaf8a6650ba2cfccbc83fa88d83e">set_compute_detected</a> (bool s=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#adcfa2c7acd752a99f3550a3fc8115744">Defaults::compute_detected</a>)</td></tr>
<tr class="separator:a83bddaf8a6650ba2cfccbc83fa88d83e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e632f7f08a89e25b16f2517029de775" id="r_a5e632f7f08a89e25b16f2517029de775"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a5e632f7f08a89e25b16f2517029de775">set_compute_max_count</a> (bool s=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a8ac1372308cea6c0ae8d91bce3c1e736">Defaults::compute_max_count</a>)</td></tr>
<tr class="separator:a5e632f7f08a89e25b16f2517029de775"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a92f8632855b1938832874cb0768b9602" id="r_a92f8632855b1938832874cb0768b9602"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a92f8632855b1938832874cb0768b9602">set_compute_max_index</a> (bool s=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a2b23685a7e8d518d144e4c76958fafba">Defaults::compute_max_index</a>)</td></tr>
<tr class="separator:a92f8632855b1938832874cb0768b9602"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6ccf406735b93e19834586150a66924a" id="r_a6ccf406735b93e19834586150a66924a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a6ccf406735b93e19834586150a66924a">set_compute_subset_total</a> (bool s=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#acd70a04fddbce22c417dbc2d5d2320e3">Defaults::compute_subset_total</a>)</td></tr>
<tr class="separator:a6ccf406735b93e19834586150a66924a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a339a15dddd17a915ade05d630b93e425" id="r_a339a15dddd17a915ade05d630b93e425"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a339a15dddd17a915ade05d630b93e425">set_compute_subset_detected</a> (bool s=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a1766d462f5f7e50bb6e0977709dfdf76">Defaults::compute_subset_detected</a>)</td></tr>
<tr class="separator:a339a15dddd17a915ade05d630b93e425"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2f5ded669fe215baf0cf3c33a6716ca2" id="r_a2f5ded669fe215baf0cf3c33a6716ca2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#a2f5ded669fe215baf0cf3c33a6716ca2">set_num_threads</a> (int n=<a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a5359e0320b093635333c50edec6f8021">Defaults::num_threads</a>)</td></tr>
<tr class="separator:a2f5ded669fe215baf0cf3c33a6716ca2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab30af17fd509f86bb40ec52e02edbaa2" id="r_ab30af17fd509f86bb40ec52e02edbaa2"><td class="memTemplParams" colspan="2">template&lt;class Matrix , typename Subset  = const uint8_t*&gt; </td></tr>
<tr class="memitem:ab30af17fd509f86bb40ec52e02edbaa2"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run</a> (const Matrix *mat, const std::vector&lt; Subset &gt; &amp;subsets) const</td></tr>
<tr class="separator:ab30af17fd509f86bb40ec52e02edbaa2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acb0f0c664a592a3c05a26136e4d7a10b" id="r_acb0f0c664a592a3c05a26136e4d7a10b"><td class="memTemplParams" colspan="2">template&lt;class Matrix , typename Subset  = const uint8_t*, typename Float , typename Integer &gt; </td></tr>
<tr class="memitem:acb0f0c664a592a3c05a26136e4d7a10b"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellQcMetrics.html#acb0f0c664a592a3c05a26136e4d7a10b">run</a> (const Matrix *mat, const std::vector&lt; Subset &gt; &amp;subsets, <a class="el" href="structscran_1_1PerCellQcMetrics_1_1Buffers.html">Buffers</a>&lt; Float, Integer &gt; &amp;output) const</td></tr>
<tr class="separator:acb0f0c664a592a3c05a26136e4d7a10b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Compute a variety of per-cell quality control metrics from a count matrix. </p>
<p>Given a feature-by-cell count matrix, this class computes several QC metrics:</p>
<ul>
<li>The total sum for each cell, which represents the efficiency of library preparation and sequencing. Low totals indicate that the library was not successfully captured.</li>
<li>The number of detected features (i.e., with non-zero counts). This also quantifies the library preparation efficiency, but with a greater focus on capturing the transcriptional complexity.</li>
<li>The maximum count across all features. This is useful in situations where only one feature is expected to be present, e.g., CRISPR guides, hash tags.</li>
<li>The row index of the feature with the maximum count. If multiple features are tied for the maximum count, the earliest feature is reported.</li>
<li>The total count in pre-defined feature subsets. The exact interpretation depends on the nature of the subset - most commonly, one subset will contain all genes on the mitochondrial chromosome, where higher proportions of counts in the mitochondrial subset indicate cell damage due to loss of cytoplasmic transcripts. Spike-in proportions can be interpreted in a similar manner.</li>
<li>The number of detected features in pre-defined feature subsets. Analogous to the number of detected features for the entire feature space.</li>
</ul>
<p>These calculations are done in a single pass, avoiding the need to reload data from the count matrix. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a147354fd4a4b932454f4ce66c7565c19" name="a147354fd4a4b932454f4ce66c7565c19"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a147354fd4a4b932454f4ce66c7565c19">&#9670;&#160;</a></span>set_compute_total()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_compute_total </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#ac70e65816427d11c645e99454f49522f">Defaults::compute_total</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Whether to compute the total count for each cell. This option only affects the <code><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run()</a></code> method that returns a <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="a83bddaf8a6650ba2cfccbc83fa88d83e" name="a83bddaf8a6650ba2cfccbc83fa88d83e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a83bddaf8a6650ba2cfccbc83fa88d83e">&#9670;&#160;</a></span>set_compute_detected()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_compute_detected </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#adcfa2c7acd752a99f3550a3fc8115744">Defaults::compute_detected</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Whether to compute the number of detected features for each cell. This option only affects the <code><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run()</a></code> method that returns a <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="a5e632f7f08a89e25b16f2517029de775" name="a5e632f7f08a89e25b16f2517029de775"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5e632f7f08a89e25b16f2517029de775">&#9670;&#160;</a></span>set_compute_max_count()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_compute_max_count </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a8ac1372308cea6c0ae8d91bce3c1e736">Defaults::compute_max_count</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Whether to compute the maximmum count for each cell. This option only affects the <code><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run()</a></code> method that returns a <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="a92f8632855b1938832874cb0768b9602" name="a92f8632855b1938832874cb0768b9602"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a92f8632855b1938832874cb0768b9602">&#9670;&#160;</a></span>set_compute_max_index()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_compute_max_index </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a2b23685a7e8d518d144e4c76958fafba">Defaults::compute_max_index</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Whether to store the index of the feature with the maximum count for each cell. This option only affects the <code><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run()</a></code> method that returns a <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="a6ccf406735b93e19834586150a66924a" name="a6ccf406735b93e19834586150a66924a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6ccf406735b93e19834586150a66924a">&#9670;&#160;</a></span>set_compute_subset_total()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_compute_subset_total </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#acd70a04fddbce22c417dbc2d5d2320e3">Defaults::compute_subset_total</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Whether to compute the total count in each feature subset. This option only affects the <code><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run()</a></code> method that returns a <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="a339a15dddd17a915ade05d630b93e425" name="a339a15dddd17a915ade05d630b93e425"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a339a15dddd17a915ade05d630b93e425">&#9670;&#160;</a></span>set_compute_subset_detected()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_compute_subset_detected </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a1766d462f5f7e50bb6e0977709dfdf76">Defaults::compute_subset_detected</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>Whether to compute the number of detected features in each feature subset. This option only affects the <code><a class="el" href="classscran_1_1PerCellQcMetrics.html#ab30af17fd509f86bb40ec52e02edbaa2">run()</a></code> method that returns a <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="a2f5ded669fe215baf0cf3c33a6716ca2" name="a2f5ded669fe215baf0cf3c33a6716ca2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2f5ded669fe215baf0cf3c33a6716ca2">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1PerCellQcMetrics.html">PerCellQcMetrics</a> &amp; scran::PerCellQcMetrics::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Defaults.html#a5359e0320b093635333c50edec6f8021">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1PerCellQcMetrics.html" title="Compute a variety of per-cell quality control metrics from a count matrix.">PerCellQcMetrics</a></code> object. </dd></dl>

</div>
</div>
<a id="ab30af17fd509f86bb40ec52e02edbaa2" name="ab30af17fd509f86bb40ec52e02edbaa2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab30af17fd509f86bb40ec52e02edbaa2">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Matrix , typename Subset  = const uint8_t*&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html">Results</a> scran::PerCellQcMetrics::run </td>
          <td>(</td>
          <td class="paramtype">const Matrix *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; Subset &gt; &amp;&#160;</td>
          <td class="paramname"><em>subsets</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute the QC metrics from an input matrix and return the results.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Matrix</td><td>Type of matrix, usually a <code>tatami::NumericMatrix</code>. </td></tr>
    <tr><td class="paramname">Subset</td><td>Pointer to an array of values interpretable as booleans.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing counts. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">subsets</td><td>Vector of pointers to arrays of length equal to <code>mat-&gt;nrow()</code>. Each array represents a feature subset and indicating whether each feature in <code>mat</code> belongs to that subset. Users can pass <code>{}</code> if no subsets are to be used.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">Results</a></code> object containing the QC metrics. Subset totals are returned depending on the <code>subsets</code>. </dd></dl>

</div>
</div>
<a id="acb0f0c664a592a3c05a26136e4d7a10b" name="acb0f0c664a592a3c05a26136e4d7a10b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acb0f0c664a592a3c05a26136e4d7a10b">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Matrix , typename Subset  = const uint8_t*, typename Float , typename Integer &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::PerCellQcMetrics::run </td>
          <td>(</td>
          <td class="paramtype">const Matrix *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; Subset &gt; &amp;&#160;</td>
          <td class="paramname"><em>subsets</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Buffers.html">Buffers</a>&lt; Float, Integer &gt; &amp;&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute the QC metrics from an input matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Matrix</td><td>Type of matrix, usually a <code>tatami::NumericMatrix</code>. </td></tr>
    <tr><td class="paramname">Subset</td><td>Pointer to a type interpretable as boolean.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells matrix containing counts. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">subsets</td><td>Vector of pointers to arrays of length equal to <code>mat-&gt;nrow()</code>. Each array represents a feature subset and indicating whether each feature in <code>mat</code> belongs to that subset. Users can pass <code>{}</code> if no subsets are to be used. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>A <code><a class="el" href="structscran_1_1PerCellQcMetrics_1_1Buffers.html" title="Buffers for direct storage of the calculated statistics.">Buffers</a></code> object in which the computed statistics are to be stored. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/quality_control/<a class="el" href="PerCellQcMetrics_8hpp_source.html">PerCellQcMetrics.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8
</small></address>
</body>
</html>
