<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::PerCellAdtQcMetrics Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1PerCellAdtQcMetrics.html">PerCellAdtQcMetrics</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1PerCellAdtQcMetrics-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::PerCellAdtQcMetrics Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute per-cell quality control metrics from an ADT count matrix.  
 <a href="classscran_1_1PerCellAdtQcMetrics.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="PerCellAdtQcMetrics_8hpp_source.html">PerCellAdtQcMetrics.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1PerCellAdtQcMetrics_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Result store for QC metric calculations.  <a href="structscran_1_1PerCellAdtQcMetrics_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a66a311ed31fa3856867ea7918c4b982e"><td class="memTemplParams" colspan="2">template&lt;class Matrix , typename SubPtr  = const uint8_t*&gt; </td></tr>
<tr class="memitem:a66a311ed31fa3856867ea7918c4b982e"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1PerCellAdtQcMetrics_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellAdtQcMetrics.html#a66a311ed31fa3856867ea7918c4b982e">run</a> (const Matrix *mat, std::vector&lt; SubPtr &gt; subsets) const</td></tr>
<tr class="separator:a66a311ed31fa3856867ea7918c4b982e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8333b5f6a1f5b6803e377d47af49ca16"><td class="memTemplParams" colspan="2">template&lt;class Matrix , typename SubPtr  = const uint8_t*, typename Sum , typename Detected , typename SubTotal &gt; </td></tr>
<tr class="memitem:a8333b5f6a1f5b6803e377d47af49ca16"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1PerCellAdtQcMetrics.html#a8333b5f6a1f5b6803e377d47af49ca16">run</a> (const Matrix *mat, const std::vector&lt; SubPtr &gt; &amp;subsets, Sum *sums, Detected *detected, std::vector&lt; SubTotal * &gt; subset_totals) const</td></tr>
<tr class="separator:a8333b5f6a1f5b6803e377d47af49ca16"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Compute per-cell quality control metrics from an ADT count matrix. </p>
<p >Given a feature-by-cell ADT count matrix, this class computes several QC metrics:</p>
<ul>
<li>The total sum for each cell, which represents the efficiency of library preparation and sequencing. This is less useful as a QC metric for ADT data given that the total count may be strongly skewed by the presence or absence of a single feature. Nonetheless, we compute it, because why not.</li>
<li>The number of detected features per cell Even though ADTs are commonly applied in situations where few features are expressed, we still expect detectable coverage of most features due to ambient contamination. The absence of detectable coverage indicates that library preparation or sequencing depth was suboptimal.</li>
<li>The total sum of counts in pre-defined feature subsets. The exact interpretation depends on the nature of the subset - the most common use case involves isotype control (IgG) features. IgG antibodies should not bind to anything, so high coverage suggests that non-specific binding is a problem, e.g., due to antibody conjugates. </li>
</ul>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a8333b5f6a1f5b6803e377d47af49ca16" name="a8333b5f6a1f5b6803e377d47af49ca16"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8333b5f6a1f5b6803e377d47af49ca16">&#9670;&nbsp;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Matrix , typename SubPtr  = const uint8_t*, typename Sum , typename Detected , typename SubTotal &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::PerCellAdtQcMetrics::run </td>
          <td>(</td>
          <td class="paramtype">const Matrix *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; SubPtr &gt; &amp;&#160;</td>
          <td class="paramname"><em>subsets</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Sum *&#160;</td>
          <td class="paramname"><em>sums</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Detected *&#160;</td>
          <td class="paramname"><em>detected</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; SubTotal * &gt;&#160;</td>
          <td class="paramname"><em>subset_totals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Compute the QC metrics from an ADT count matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Matrix</td><td>Type of matrix, usually a <code>tatami::NumericMatrix</code>. </td></tr>
    <tr><td class="paramname">SubPtr</td><td>Pointer to a type interpretable as boolean. </td></tr>
    <tr><td class="paramname">Sum</td><td>Floating-point value, to store the sums. </td></tr>
    <tr><td class="paramname">Detected</td><td>Integer value, to store the number of detected features. </td></tr>
    <tr><td class="paramname">SubTotal</td><td>Floating point value, to store the subset proportions.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells matrix containing ADT counts. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">subsets</td><td>Vector of pointers to arrays of length equal to <code>mat-&gt;nrow()</code>. Each array represents a feature subset and indicating whether each feature in <code>mat</code> belongs to that subset. Users can pass <code>{}</code> if no subsets are to be used. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">sums</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is used to store the computed sums for all cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">detected</td><td>Pointer to an array of length equal to the number of columns in <code>mat</code>. This is used to store the number of detected features for all cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">subset_totals</td><td>Vector of pointers to arrays of length equal to the number of columns in <code>mat</code>. Each array corresponds to a feature subset and is used to store the total count in that subset across all cells. The vector should be of length equal to that of <code>subsets</code>. Users can pass <code>{}</code> if no subsets are used.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>sums</code>, <code>detected</code>, and each array in <code>subset_proportions</code> is filled with the relevant statistics. </dd></dl>

</div>
</div>
<a id="a66a311ed31fa3856867ea7918c4b982e" name="a66a311ed31fa3856867ea7918c4b982e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a66a311ed31fa3856867ea7918c4b982e">&#9670;&nbsp;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Matrix , typename SubPtr  = const uint8_t*&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1PerCellAdtQcMetrics_1_1Results.html">Results</a> scran::PerCellAdtQcMetrics::run </td>
          <td>(</td>
          <td class="paramtype">const Matrix *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; SubPtr &gt;&#160;</td>
          <td class="paramname"><em>subsets</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Compute the QC metrics from an ADT count matrix and return the results.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Matrix</td><td>Type of matrix, usually a <code>tatami::NumericMatrix</code>. </td></tr>
    <tr><td class="paramname">SubPtr</td><td>Pointer to a type interpretable as boolean.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells matrix containing ADT counts. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">subsets</td><td>Vector of pointers to arrays of length equal to <code>mat-&gt;nrow()</code>. Each array represents a feature subset and indicating whether each feature in <code>mat</code> belongs to that subset. Users can pass <code>{}</code> if no subsets are to be used.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1PerCellAdtQcMetrics_1_1Results.html" title="Result store for QC metric calculations.">PerCellAdtQcMetrics::Results</a></code> object containing the QC metrics. Subset proportions are returned depending on the <code>subsets</code>. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/quality_control/<a class="el" href="PerCellAdtQcMetrics_8hpp_source.html">PerCellAdtQcMetrics.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
