<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::ScoreMarkers Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1ScoreMarkers.html">ScoreMarkers</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classscran_1_1ScoreMarkers-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">scran::ScoreMarkers Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Score each gene as a candidate marker for each group of cells.  
 <a href="classscran_1_1ScoreMarkers.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="ScoreMarkers_8hpp_source.html">ScoreMarkers.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a9348af3ae204495ab4bb00ba2fdd5c4f"><td class="memItemLeft" align="right" valign="top"><a id="a9348af3ae204495ab4bb00ba2fdd5c4f"></a>
<a class="el" href="classscran_1_1ScoreMarkers.html">ScoreMarkers</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>set_threshold</b> (double t=0)</td></tr>
<tr class="separator:a9348af3ae204495ab4bb00ba2fdd5c4f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3f6596c20cf349535877bb68760c1664"><td class="memTemplParams" colspan="2"><a id="a3f6596c20cf349535877bb68760c1664"></a>
template&lt;class MAT , typename G , typename Stat &gt; </td></tr>
<tr class="memitem:a3f6596c20cf349535877bb68760c1664"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><b>run</b> (const MAT *p, const G *group, std::vector&lt; Stat * &gt; means, std::vector&lt; Stat * &gt; detected, std::vector&lt; std::vector&lt; Stat * &gt; &gt; cohen)</td></tr>
<tr class="separator:a3f6596c20cf349535877bb68760c1664"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a00cab93a1ab861c3574b909593118e53"><td class="memTemplParams" colspan="2"><a id="a00cab93a1ab861c3574b909593118e53"></a>
template&lt;class MAT , typename G , typename B , typename Stat &gt; </td></tr>
<tr class="memitem:a00cab93a1ab861c3574b909593118e53"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><b>run_blocked</b> (const MAT *p, const G *group, const B *block, std::vector&lt; std::vector&lt; Stat * &gt; &gt; means, std::vector&lt; std::vector&lt; Stat * &gt; &gt; detected, std::vector&lt; std::vector&lt; Stat * &gt; &gt; cohen)</td></tr>
<tr class="separator:a00cab93a1ab861c3574b909593118e53"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9c9f0977a0823495dc65e30ba38ece87"><td class="memTemplParams" colspan="2"><a id="a9c9f0977a0823495dc65e30ba38ece87"></a>
template&lt;class MAT , typename G &gt; </td></tr>
<tr class="memitem:a9c9f0977a0823495dc65e30ba38ece87"><td class="memTemplItemLeft" align="right" valign="top">Results&#160;</td><td class="memTemplItemRight" valign="bottom"><b>run</b> (const MAT *p, const G *group)</td></tr>
<tr class="separator:a9c9f0977a0823495dc65e30ba38ece87"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78206bf58c4f94f4f8bda9d716ac4b25"><td class="memTemplParams" colspan="2"><a id="a78206bf58c4f94f4f8bda9d716ac4b25"></a>
template&lt;class MAT , typename G , typename B &gt; </td></tr>
<tr class="memitem:a78206bf58c4f94f4f8bda9d716ac4b25"><td class="memTemplItemLeft" align="right" valign="top">Results&#160;</td><td class="memTemplItemRight" valign="bottom"><b>run_blocked</b> (const MAT *p, const G *group, const B *block)</td></tr>
<tr class="separator:a78206bf58c4f94f4f8bda9d716ac4b25"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a935df8beac99b9a4b5e0e6071973ad91"><td class="memItemLeft" align="right" valign="top"><a id="a935df8beac99b9a4b5e0e6071973ad91"></a>
static constexpr int&#160;</td><td class="memItemRight" valign="bottom"><b>nsummaries</b> = 4</td></tr>
<tr class="separator:a935df8beac99b9a4b5e0e6071973ad91"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Score each gene as a candidate marker for each group of cells. </p>
<p>Markers are identified by pairwise comparisons between groups of cells (e.g., clusters, cell types). Given <code>n</code> groups, each group has <code>n - 1</code> comparisons and thus <code>n - 1</code> effect sizes. For each group, scores for each gene are obtained by taking summary statistics - e.g., the minimum, median, mean - of the effect sizes across all of that group's comparisons. Users can then sort by any of these summaries to obtain a ranking of potential marker genes for each group. The choice of effect size and summary statistic determines the characteristics of the resulting marker set.</p>
<h1><a class="anchor" id="Effect"></a>
sizes</h1>
<p>Cohen's d is the standardized log-fold change between two groups. This is defined as the difference in the mean log-expression for each group scaled by the average standard deviation across the two groups. (Technically, we should use the pooled variance; however, this introduces some unpleasant asymmetry depending on the variance of the larger group, so we take a simple average instead.) Cohen's d is analogous to the t-statistic in a two-sample t-test and avoids spuriously large effect sizes from comparisons between highly variable groups. We can also interpret Cohen's d as the number of standard deviations between the two group means.</p>
<p>The definitions of these effect sizes are generalized when testing against a non-zero log-fold change threshold:</p>
<ul>
<li>Cohen's d is redefined as the standardized difference between the observed log-fold change and the specified threshold, analogous to the TREAT method from <b>limma</b>. Large positive values are only obtained when the observed log-fold change is significantly greater than the threshold.</li>
</ul>
<h1><a class="anchor" id="Summary"></a>
statistics</h1>
<p>The choice of summary statistic dictates the interpretation of the ranking. Given a group X:</p>
<ul>
<li>A large mean effect size indicates that the gene is upregulated in X compared to the average of the other groups. A small value indicates that the gene is downregulated in X instead. This is a good general-purpose summary statistic for ranking, usually by decreasing size to obtain upregulated markers in X.</li>
<li>A large median effect size indicates that the gene is upregulated in X compared to most (&gt;50%) other groups. A small value indicates that the gene is downregulated in X instead. This is also good for ranking, with the advantage of being more robust to outlier effects (but also the disadvantage of being less sensitive to strong effects in a minority of comparisons).</li>
<li>A large minimum effect size indicates that the gene is upregulated in X compared to all other groups. A small value indicates that the gene is downregulated in X compared to at least one other group. This is the most stringent summary as markers will only have extreme values if they are <em>uniquely</em> up/downregulated in X compared to every other group. However, it may not be effective if X is closely related to any of the groups.</li>
<li>The "minimum rank" (a.k.a. min-rank) is defined by ranking genes based on decreasing effect size <em>within</em> each comparison, and then taking the smallest rank <em>across</em> comparisons. A minimum rank of 1 means that the gene is the top upregulated gene in at least one comparison to another group. More generally, a minimum rank of T indicates that the gene is the T-th upregulated gene in at least one comparison. Applying a threshold on the minimum rank is useful for obtaining a set of genes that, in combination, are guaranteed to distinguish X from every other group.</li>
</ul>
<p>The exact definition of "large" and "small" depends on the choice of effect size. For Cohen's d, the value must be positive to be considered "large", and negative to be considered "small".</p>
<p>Additionally, these interpretations change slightly when a log-fold change threshold is supplied. For a positive threshold, only large values can be unambiguously interpreted as upregulation; small values do not have any guaranteed meaning. Conversely, for a negative threshold, only small values can be unambiguously interpreted as downregulation.</p>
<h1><a class="anchor" id="Other"></a>
statistics</h1>
<p>We report the mean log-expression of all cells in each cluster, as well as the proportion of cells with detectable expression in each cluster. </p>
</div><hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/differential_analysis/<a class="el" href="ScoreMarkers_8hpp_source.html">ScoreMarkers.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
