<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.7"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::BuildSNNGraph Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.7 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1BuildSNNGraph-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::BuildSNNGraph Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Build a shared nearest-neighbor graph with cells as nodes.  
 <a href="classscran_1_1BuildSNNGraph.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="BuildSNNGraph_8hpp_source.html">BuildSNNGraph.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter settings.  <a href="structscran_1_1BuildSNNGraph_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html" title="Results of SNN graph construction.">Results</a> of SNN graph construction.  <a href="structscran_1_1BuildSNNGraph_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-types" name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:ae7e2167f6c3cf02432a3cdc6e4583e2a"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#ae7e2167f6c3cf02432a3cdc6e4583e2a">Scheme</a> { <b>RANKED</b>
, <b>NUMBER</b>
, <b>JACCARD</b>
 }</td></tr>
<tr class="separator:ae7e2167f6c3cf02432a3cdc6e4583e2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a22b16cc067f034eecc097b79b903a017"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#a22b16cc067f034eecc097b79b903a017">set_neighbors</a> (int k=<a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#a6a376b256377055e23a96c3d9eb96365">Defaults::neighbors</a>)</td></tr>
<tr class="separator:a22b16cc067f034eecc097b79b903a017"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac5085b513afa0e47617833e7279c61cb"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#ac5085b513afa0e47617833e7279c61cb">set_approximate</a> (bool a=<a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#af6d0a1c84cd739442a6767171e0202e0">Defaults::approximate</a>)</td></tr>
<tr class="separator:ac5085b513afa0e47617833e7279c61cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b2863dda326ce82c1fba43b7660e05a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#a9b2863dda326ce82c1fba43b7660e05a">set_weighting_scheme</a> (<a class="el" href="classscran_1_1BuildSNNGraph.html#ae7e2167f6c3cf02432a3cdc6e4583e2a">Scheme</a> w=<a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#a258288cda0e643719d53a834424b56a6">Defaults::weighting_scheme</a>)</td></tr>
<tr class="separator:a9b2863dda326ce82c1fba43b7660e05a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a45fde3074ff964c1105f1ef29d78bad7"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#a45fde3074ff964c1105f1ef29d78bad7">set_num_threads</a> (int n=<a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#a0b0aa250902f24e75b95bdd42a8ccb86">Defaults::num_threads</a>)</td></tr>
<tr class="separator:a45fde3074ff964c1105f1ef29d78bad7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a261dde3f2b8878bcafa5dbdc6048aaed"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#a261dde3f2b8878bcafa5dbdc6048aaed">run</a> (size_t ndims, size_t ncells, const double *mat) const</td></tr>
<tr class="separator:a261dde3f2b8878bcafa5dbdc6048aaed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a61a324024d7179d5a509730d3c69b18d"><td class="memTemplParams" colspan="2">template&lt;class Algorithm &gt; </td></tr>
<tr class="memitem:a61a324024d7179d5a509730d3c69b18d"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#a61a324024d7179d5a509730d3c69b18d">run</a> (const Algorithm *search) const</td></tr>
<tr class="separator:a61a324024d7179d5a509730d3c69b18d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaa6688278f7b4bfc05c0941719b66c52"><td class="memTemplParams" colspan="2">template&lt;class Indices &gt; </td></tr>
<tr class="memitem:aaa6688278f7b4bfc05c0941719b66c52"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1BuildSNNGraph.html#aaa6688278f7b4bfc05c0941719b66c52">run</a> (const std::vector&lt; Indices &gt; &amp;indices) const</td></tr>
<tr class="separator:aaa6688278f7b4bfc05c0941719b66c52"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Build a shared nearest-neighbor graph with cells as nodes. </p>
<p>In a shared nearest neighbor graph, pairs of cells are connected to each other by an edge with weight determined from their shared nearest neighbors. If two cells are close together but have distinct sets of neighbors, the corresponding edge is downweighted as the two cells are unlikely to be part of the same neighborhood. In this manner, highly weighted edges will form within highly interconnected neighborhoods where many cells share the same neighbors. This provides a more sophisticated definition of similarity between cells compared to a simpler (unweighted) nearest neighbor graph that just focuses on immediate proximity.</p>
<p>A key parameter in the construction of the graph is the number of nearest neighbors $k$ to consider. Larger values increase the connectivity of the graph and reduce the granularity of any subsequent community detection steps (see <code>scran::ClusterSNNGraph</code>) at the cost of speed. The nearest neighbor search can be performed using either vantage point trees (exact) or with the Annoy algorithm (approximate) - see the <a href="https://github.com/LTLA/knncolle"><b>knncolle</b></a> library for details.</p>
<p>For the edges, a variety of weighting schemes are possible:</p>
<ul>
<li><code>RANKED</code> defines the weight between two nodes as $k - r/2$ where $r$ is the smallest sum of ranks for any shared neighboring node (Xu and Su, 2015). For the purposes of this ranking, each node has a rank of zero in its own nearest-neighbor set. More shared neighbors, or shared neighbors that are close to both observations, will generally yield larger weights.</li>
<li><code>NUMBER</code> defines the weight between two nodes as the number of shared nearest neighbors between them. The weight can range from zero to $k + 1$, as the node itself is included in its own nearest-neighbor set. This is a simpler scheme that is also slightly faster but does not account for the ranking of neighbors within each set.</li>
<li><code>JACCARD</code> defines the weight between two nodes as the Jaccard index of their neighbor sets. This weight can range from zero to 1, and is a monotonic transformation of the weight used by <code>NUMBER</code>.</li>
</ul>
<p>See the <code>ClusterSNNGraph</code> class to perform community detection on the graph returned by <code><a class="el" href="classscran_1_1BuildSNNGraph.html#a261dde3f2b8878bcafa5dbdc6048aaed">run()</a></code>.</p>
<dl class="section see"><dt>See also</dt><dd>Xu C and Su Z (2015). Identification of cell types from single-cell transcriptomes using a novel clustering method. <em>Bioinformatics</em> 31, 1974-80 </dd></dl>
</div><h2 class="groupheader">Member Enumeration Documentation</h2>
<a id="ae7e2167f6c3cf02432a3cdc6e4583e2a" name="ae7e2167f6c3cf02432a3cdc6e4583e2a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae7e2167f6c3cf02432a3cdc6e4583e2a">&#9670;&#160;</a></span>Scheme</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classscran_1_1BuildSNNGraph.html#ae7e2167f6c3cf02432a3cdc6e4583e2a">scran::BuildSNNGraph::Scheme</a></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Choices for the edge weighting scheme during graph construction. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a22b16cc067f034eecc097b79b903a017" name="a22b16cc067f034eecc097b79b903a017"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a22b16cc067f034eecc097b79b903a017">&#9670;&#160;</a></span>set_neighbors()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp; scran::BuildSNNGraph::set_neighbors </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>k</em> = <code><a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#a6a376b256377055e23a96c3d9eb96365">Defaults::neighbors</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">k</td><td>Number of neighbors to use in the nearest neighbor search.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BuildSNNGraph.html" title="Build a shared nearest-neighbor graph with cells as nodes.">BuildSNNGraph</a></code> object. </dd></dl>

</div>
</div>
<a id="ac5085b513afa0e47617833e7279c61cb" name="ac5085b513afa0e47617833e7279c61cb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac5085b513afa0e47617833e7279c61cb">&#9670;&#160;</a></span>set_approximate()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp; scran::BuildSNNGraph::set_approximate </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>a</em> = <code><a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#af6d0a1c84cd739442a6767171e0202e0">Defaults::approximate</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>Whether to perform an approximate nearest neighbor search.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BuildSNNGraph.html" title="Build a shared nearest-neighbor graph with cells as nodes.">BuildSNNGraph</a></code> object. </dd></dl>

</div>
</div>
<a id="a9b2863dda326ce82c1fba43b7660e05a" name="a9b2863dda326ce82c1fba43b7660e05a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9b2863dda326ce82c1fba43b7660e05a">&#9670;&#160;</a></span>set_weighting_scheme()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp; scran::BuildSNNGraph::set_weighting_scheme </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classscran_1_1BuildSNNGraph.html#ae7e2167f6c3cf02432a3cdc6e4583e2a">Scheme</a>&#160;</td>
          <td class="paramname"><em>w</em> = <code><a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#a258288cda0e643719d53a834424b56a6">Defaults::weighting_scheme</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">w</td><td>The edge weighting scheme to use.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BuildSNNGraph.html" title="Build a shared nearest-neighbor graph with cells as nodes.">BuildSNNGraph</a></code> object. </dd></dl>

</div>
</div>
<a id="a45fde3074ff964c1105f1ef29d78bad7" name="a45fde3074ff964c1105f1ef29d78bad7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a45fde3074ff964c1105f1ef29d78bad7">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1BuildSNNGraph.html">BuildSNNGraph</a> &amp; scran::BuildSNNGraph::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1BuildSNNGraph_1_1Defaults.html#a0b0aa250902f24e75b95bdd42a8ccb86">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1BuildSNNGraph.html" title="Build a shared nearest-neighbor graph with cells as nodes.">BuildSNNGraph</a></code> object. </dd></dl>

</div>
</div>
<a id="a261dde3f2b8878bcafa5dbdc6048aaed" name="a261dde3f2b8878bcafa5dbdc6048aaed"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a261dde3f2b8878bcafa5dbdc6048aaed">&#9670;&#160;</a></span>run() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a> scran::BuildSNNGraph::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ndims</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ncells</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double *&#160;</td>
          <td class="paramname"><em>mat</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ndims</td><td>Number of dimensions. </td></tr>
    <tr><td class="paramname">ncells</td><td>Number of cells. </td></tr>
    <tr><td class="paramname">mat</td><td>Pointer to an array of expression values or a low-dimensional representation thereof. Rows should be dimensions while columns should be cells. Data should be stored in column-major format.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The edges and weights of the constructed SNN graph. </dd></dl>

</div>
</div>
<a id="a61a324024d7179d5a509730d3c69b18d" name="a61a324024d7179d5a509730d3c69b18d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a61a324024d7179d5a509730d3c69b18d">&#9670;&#160;</a></span>run() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Algorithm &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a> scran::BuildSNNGraph::run </td>
          <td>(</td>
          <td class="paramtype">const Algorithm *&#160;</td>
          <td class="paramname"><em>search</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Algorithm</td><td>Any instance of a <code>knncolle::Base</code> subclass.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">search</td><td>Pointer to a <code>knncolle::Base</code> instance to use for the nearest-neighbor search.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The edges and weights of the constructed SNN graph. </dd></dl>

</div>
</div>
<a id="aaa6688278f7b4bfc05c0941719b66c52" name="aaa6688278f7b4bfc05c0941719b66c52"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaa6688278f7b4bfc05c0941719b66c52">&#9670;&#160;</a></span>run() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;class Indices &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1BuildSNNGraph_1_1Results.html">Results</a> scran::BuildSNNGraph::run </td>
          <td>(</td>
          <td class="paramtype">const std::vector&lt; Indices &gt; &amp;&#160;</td>
          <td class="paramname"><em>indices</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Indices</td><td>Vector of integer indices.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">indices</td><td>Vector of indices of the neighbors for each cell, sorted by increasing distance.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The edges and weights of the constructed SNN graph. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/clustering/<a class="el" href="BuildSNNGraph_8hpp_source.html">BuildSNNGraph.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.7
</small></address>
</body>
</html>
