<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.6"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::SuggestCrisprQcFilters Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.6 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html">SuggestCrisprQcFilters</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1SuggestCrisprQcFilters-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::SuggestCrisprQcFilters Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Create filters to identify low-quality cells from CRISPR-derived QC metrics.  
 <a href="classscran_1_1SuggestCrisprQcFilters.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="SuggestCrisprQcFilters_8hpp_source.html">SuggestCrisprQcFilters.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameters.  <a href="structscran_1_1SuggestCrisprQcFilters_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html" title="Thresholds to define outliers on each metric.">Thresholds</a> to define outliers on each metric.  <a href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a7655a9639ef9da450f802a7503cbdc6a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html">SuggestCrisprQcFilters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html#a7655a9639ef9da450f802a7503cbdc6a">set_num_mads</a> (double n=<a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Defaults.html#adb19d5cf195d351bd412763a0182e4f9">Defaults::num_mads</a>)</td></tr>
<tr class="separator:a7655a9639ef9da450f802a7503cbdc6a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a065021a998e26e58e9cd26a2186921da"><td class="memTemplParams" colspan="2">template&lt;typename Float , typename Integer &gt; </td></tr>
<tr class="memitem:a065021a998e26e58e9cd26a2186921da"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html#a065021a998e26e58e9cd26a2186921da">run</a> (size_t n, const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Buffers.html">PerCellCrisprQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;buffers) const</td></tr>
<tr class="separator:a065021a998e26e58e9cd26a2186921da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a773f2ed2bb22b2bbbf9104bd83aaf143"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html#a773f2ed2bb22b2bbbf9104bd83aaf143">run</a> (const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Results.html">PerCellCrisprQcMetrics::Results</a> &amp;metrics) const</td></tr>
<tr class="separator:a773f2ed2bb22b2bbbf9104bd83aaf143"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1f0cb4b8eb2ae7d7225716bfcea06f58"><td class="memTemplParams" colspan="2">template&lt;typename Block , typename Float , typename Integer &gt; </td></tr>
<tr class="memitem:a1f0cb4b8eb2ae7d7225716bfcea06f58"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html#a1f0cb4b8eb2ae7d7225716bfcea06f58">run_blocked</a> (size_t n, const Block *block, const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Buffers.html">PerCellCrisprQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;buffers) const</td></tr>
<tr class="separator:a1f0cb4b8eb2ae7d7225716bfcea06f58"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2a1f7c118ce66b309b2c5707d61be484"><td class="memTemplParams" colspan="2">template&lt;typename Block &gt; </td></tr>
<tr class="memitem:a2a1f7c118ce66b309b2c5707d61be484"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html#a2a1f7c118ce66b309b2c5707d61be484">run_blocked</a> (const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Results.html">PerCellCrisprQcMetrics::Results</a> &amp;metrics, const Block *block) const</td></tr>
<tr class="separator:a2a1f7c118ce66b309b2c5707d61be484"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Create filters to identify low-quality cells from CRISPR-derived QC metrics. </p>
<p>In CRISPR guide count matrices, the QC filtering decisions are somewhat different than those for the other modalities. Here, low-quality cells are defined as those with:</p>
<ul>
<li>A low maximum count. This indicates that library preparation or sequencing was suboptimal, or that the cell was not transfected with any guide construct, or that the cell failed to express that guide construct.</li>
</ul>
<p>Directly defining a threshold on the maximum count is somewhat tricky as unsuccessful transfection is not uncommon. This often results in a large subpopulation with low maximum counts, inflating the MAD and compromising the threshold calculation. Instead, we use the following approach:</p>
<ol type="1">
<li>Compute the median of the proportion of counts in the most abundant guide (i.e., the maximum proportion),</li>
<li>Subset the cells to only those with maximum proportions above the median,</li>
<li>Define a threshold for low outliers on the log-transformed maximum count within the subset.</li>
</ol>
<p>This assumes that over 50% of cells were successfully transfected with a single guide construct and have high maximum proportions. In contrast, unsuccessful transfections will be dominated by ambient contamination and have low proportions. By taking the subset above the median proportion, we remove all of the unsuccessful transfections and enrich for mostly-high-quality cells. From there, we can apply the usual outlier detection methods on the maximum count, with log-transformation to avoid a negative threshold.</p>
<p>Keep in mind that the maximum proportion is only used to define the subset for threshold calculation. Once the maximum count threshold is computed, they are applied to all cells, regardless of their maximum proportions. This allows us to recover good cells that would have been filtered out by our aggressive median subset. It also ensures that we do not remove cells transfected with multiple guides - such cells are not necessarily uninteresting, e.g., for examining interaction effects, so we will err on the side of caution and leave them in.</p>
<p>For datasets with multiple blocks, <code><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html#a1f0cb4b8eb2ae7d7225716bfcea06f58">SuggestCrisprQcFilters::run_blocked()</a></code> will compute block-specific thresholds for the maximum count. See comments in <code><a class="el" href="classscran_1_1SuggestRnaQcFilters.html" title="Create filters to identify low-quality cells from RNA-derived QC metrics.">SuggestRnaQcFilters</a></code> for more details. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a7655a9639ef9da450f802a7503cbdc6a" name="a7655a9639ef9da450f802a7503cbdc6a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7655a9639ef9da450f802a7503cbdc6a">&#9670;&#160;</a></span>set_num_mads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html">SuggestCrisprQcFilters</a> &amp; scran::SuggestCrisprQcFilters::set_num_mads </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Defaults.html#adb19d5cf195d351bd412763a0182e4f9">Defaults::num_mads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of MADs below the median, to define the threshold for outliers in the maximum count. This should be non-negative.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1SuggestCrisprQcFilters.html" title="Create filters to identify low-quality cells from CRISPR-derived QC metrics.">SuggestCrisprQcFilters</a></code> object. </dd></dl>

</div>
</div>
<a id="a065021a998e26e58e9cd26a2186921da" name="a065021a998e26e58e9cd26a2186921da"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a065021a998e26e58e9cd26a2186921da">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Float , typename Integer &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestCrisprQcFilters::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Buffers.html">PerCellCrisprQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;&#160;</td>
          <td class="paramname"><em>buffers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Float</td><td>Floating point type for the metrics. </td></tr>
    <tr><td class="paramname">Integer</td><td>Integer for the metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">buffers</td><td>Pointers to arrays of length <code>n</code>, containing the per-cell CRISPR-derived metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric. </dd></dl>

</div>
</div>
<a id="a773f2ed2bb22b2bbbf9104bd83aaf143" name="a773f2ed2bb22b2bbbf9104bd83aaf143"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a773f2ed2bb22b2bbbf9104bd83aaf143">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestCrisprQcFilters::run </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Results.html">PerCellCrisprQcMetrics::Results</a> &amp;&#160;</td>
          <td class="paramname"><em>metrics</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">metrics</td><td>Collection of arrays of length equal to the number of cells, containing the per-cell CRISPR-derived metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric. </dd></dl>

</div>
</div>
<a id="a1f0cb4b8eb2ae7d7225716bfcea06f58" name="a1f0cb4b8eb2ae7d7225716bfcea06f58"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1f0cb4b8eb2ae7d7225716bfcea06f58">&#9670;&#160;</a></span>run_blocked() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Block , typename Float , typename Integer &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestCrisprQcFilters::run_blocked </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Buffers.html">PerCellCrisprQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;&#160;</td>
          <td class="paramname"><em>buffers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Block</td><td>Integer type for the block assignments. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the metrics. </td></tr>
    <tr><td class="paramname">Integer</td><td>Integer for the metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length <code>n</code>, containing the block assignments for each cell. This may be <code>NULL</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">buffers</td><td>Pointers to arrays of length <code>n</code>, containing the per-cell CRISPR-derived metrics. Only <code>max_proportion</code> and <code>sums</code> are used; <code>detected</code> is ignored and does not need to be set.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric in each block. </dd></dl>

</div>
</div>
<a id="a2a1f7c118ce66b309b2c5707d61be484" name="a2a1f7c118ce66b309b2c5707d61be484"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2a1f7c118ce66b309b2c5707d61be484">&#9670;&#160;</a></span>run_blocked() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Block &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestCrisprQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestCrisprQcFilters::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellCrisprQcMetrics_1_1Results.html">PerCellCrisprQcMetrics::Results</a> &amp;&#160;</td>
          <td class="paramname"><em>metrics</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block *&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Block</td><td>Integer type for the block assignments.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">metrics</td><td>Collection of arrays of length equal to the number of cells, containing the per-cell CRISPR-derived metrics. Only <code>max_proportion</code> and <code>sums</code> are used; <code>detected</code> is ignored and does not need to be set. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing the block assignments for each cell. This may be <code>NULL</code>, in which case all cells are assumed to belong to the same block.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric in each block. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/quality_control/<a class="el" href="SuggestCrisprQcFilters_8hpp_source.html">SuggestCrisprQcFilters.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.6
</small></address>
</body>
</html>
