<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.5"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::ChooseHVGs Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.5 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1ChooseHVGs.html">ChooseHVGs</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1ChooseHVGs-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::ChooseHVGs Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Choose highly variable genes for downstream analyses.  
 <a href="classscran_1_1ChooseHVGs.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="ChooseHVGs_8hpp_source.html">ChooseHVGs.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ChooseHVGs_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default paramater settings.  <a href="structscran_1_1ChooseHVGs_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:aa68796b51ac32a9c0fd4451ac91d60a0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ChooseHVGs.html">ChooseHVGs</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ChooseHVGs.html#aa68796b51ac32a9c0fd4451ac91d60a0">set_top</a> (size_t t=<a class="el" href="structscran_1_1ChooseHVGs_1_1Defaults.html#ab47b622404c5eb3ed1835518e4164bf1">Defaults::top</a>)</td></tr>
<tr class="separator:aa68796b51ac32a9c0fd4451ac91d60a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8485437ec27c1b9d167fd00b38dad235"><td class="memTemplParams" colspan="2">template&lt;typename V , typename T &gt; </td></tr>
<tr class="memitem:a8485437ec27c1b9d167fd00b38dad235"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ChooseHVGs.html#a8485437ec27c1b9d167fd00b38dad235">run</a> (size_t n, const V *statistic, T *output) const</td></tr>
<tr class="separator:a8485437ec27c1b9d167fd00b38dad235"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afaa78fda6d2dfddd354b878ddabd911c"><td class="memTemplParams" colspan="2">template&lt;typename T  = uint8_t, typename V &gt; </td></tr>
<tr class="memitem:afaa78fda6d2dfddd354b878ddabd911c"><td class="memTemplItemLeft" align="right" valign="top">std::vector&lt; T &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ChooseHVGs.html#afaa78fda6d2dfddd354b878ddabd911c">run</a> (size_t n, const V *statistic) const</td></tr>
<tr class="separator:afaa78fda6d2dfddd354b878ddabd911c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Choose highly variable genes for downstream analyses. </p>
<p >This is done by selecting the <code>top</code> number of genes with the largest values for a precomputed variance statistic. We typically use the residual from the trend as computed by <code><a class="el" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVar</a></code>. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="afaa78fda6d2dfddd354b878ddabd911c" name="afaa78fda6d2dfddd354b878ddabd911c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afaa78fda6d2dfddd354b878ddabd911c">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T  = uint8_t, typename V &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; T &gt; scran::ChooseHVGs::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const V *&#160;</td>
          <td class="paramname"><em>statistic</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">V</td><td>Type of the variance statistic. </td></tr>
    <tr><td class="paramname">T</td><td>Type to be used as a boolean.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">statistic</td><td>Pointer to an array of length <code>n</code> containing the per-gene variance statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A vector of booleans of length <code>n</code>, indicating whether each gene is to be retained. </dd></dl>

</div>
</div>
<a id="a8485437ec27c1b9d167fd00b38dad235" name="a8485437ec27c1b9d167fd00b38dad235"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8485437ec27c1b9d167fd00b38dad235">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename V , typename T &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ChooseHVGs::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const V *&#160;</td>
          <td class="paramname"><em>statistic</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">T *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">V</td><td>Type of the variance statistic. </td></tr>
    <tr><td class="paramname">T</td><td>Type to be used as a boolean.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">statistic</td><td>Pointer to an array of length <code>n</code> containing the per-gene variance statistics. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array of length <code>n</code>, used to store a boolean flag.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>output</code> is filled with <code>true</code> if the gene is to be retained and <code>false</code> otherwise. </dd></dl>

</div>
</div>
<a id="aa68796b51ac32a9c0fd4451ac91d60a0" name="aa68796b51ac32a9c0fd4451ac91d60a0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa68796b51ac32a9c0fd4451ac91d60a0">&#9670;&#160;</a></span>set_top()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ChooseHVGs.html">ChooseHVGs</a> &amp; scran::ChooseHVGs::set_top </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>t</em> = <code><a class="el" href="structscran_1_1ChooseHVGs_1_1Defaults.html#ab47b622404c5eb3ed1835518e4164bf1">Defaults::top</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">t</td><td>The number of top genes to consider as highly variable genes.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ChooseHVGs.html" title="Choose highly variable genes for downstream analyses.">ChooseHVGs</a></code> object.</dd></dl>
<p>The choice is more or less arbitrary and is a trade-off between capturing more biological signal at the cost of increasing noise. Values of 1000-5000 seem appropriate for most use cases. Note that increasing the number of genes will also increase the size of the subdataset to be used in downstream computation. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/feature_selection/<a class="el" href="ChooseHVGs_8hpp_source.html">ChooseHVGs.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.5
</small></address>
</body>
</html>
