<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.7"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::GroupedSizeFactors Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.7 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1GroupedSizeFactors-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::GroupedSizeFactors Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute grouped size factors to handle composition bias.  
 <a href="classscran_1_1GroupedSizeFactors.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="GroupedSizeFactors_8hpp_source.html">GroupedSizeFactors.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter settings.  <a href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Result of the size factor calculations.  <a href="structscran_1_1GroupedSizeFactors_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a907245f06f970a65c5a7c1a4a96af0fc"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a907245f06f970a65c5a7c1a4a96af0fc">set_center</a> (bool c=<a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#acbe5c154e3b67420a1f2ef84ef304310">Defaults::center</a>)</td></tr>
<tr class="separator:a907245f06f970a65c5a7c1a4a96af0fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aec88f173e14eb0df8dcc8e1e9d6db1c0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#aec88f173e14eb0df8dcc8e1e9d6db1c0">set_prior_count</a> (double p=<a class="el" href="structscran_1_1MedianSizeFactors_1_1Defaults.html#a909251588d34d41237a64df1dd85552c">MedianSizeFactors::Defaults::prior_count</a>)</td></tr>
<tr class="separator:aec88f173e14eb0df8dcc8e1e9d6db1c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9471c099549a9d24c2fdc1bdbea98c7d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a9471c099549a9d24c2fdc1bdbea98c7d">set_handle_zeros</a> (bool z=<a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#a95d70f82976a430a11c20c1e9a5a068e">Defaults::handle_zeros</a>)</td></tr>
<tr class="separator:a9471c099549a9d24c2fdc1bdbea98c7d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a084f2ef8e3d651ebc4604c59850cb360"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a084f2ef8e3d651ebc4604c59850cb360">set_handle_non_finite</a> (bool n=<a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#a0586f933b7cd61f9f07daa8272360d2a">Defaults::handle_non_finite</a>)</td></tr>
<tr class="separator:a084f2ef8e3d651ebc4604c59850cb360"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab9121d8622080b615083d0d2b53a32f1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#ab9121d8622080b615083d0d2b53a32f1">set_num_threads</a> (int n=<a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#a26a37b3247734472ce814d356243c980">Defaults::num_threads</a>)</td></tr>
<tr class="separator:ab9121d8622080b615083d0d2b53a32f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afec8d5ba32a045c587f40c832973de2e"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX , typename Group , typename Out &gt; </td></tr>
<tr class="memitem:afec8d5ba32a045c587f40c832973de2e"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#afec8d5ba32a045c587f40c832973de2e">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Group *group, Out *output) const</td></tr>
<tr class="separator:afec8d5ba32a045c587f40c832973de2e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2dc18c86bde54ed4bda775d5f2bbac16"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX , typename Group , typename Out &gt; </td></tr>
<tr class="memitem:a2dc18c86bde54ed4bda775d5f2bbac16"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a2dc18c86bde54ed4bda775d5f2bbac16">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Group *group, size_t reference, Out *output) const</td></tr>
<tr class="separator:a2dc18c86bde54ed4bda775d5f2bbac16"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abda7647acd8947583d91e9b86d7b5461"><td class="memTemplParams" colspan="2">template&lt;typename Out  = double, typename T , typename IDX , typename Group &gt; </td></tr>
<tr class="memitem:abda7647acd8947583d91e9b86d7b5461"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a>&lt; Out &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#abda7647acd8947583d91e9b86d7b5461">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Group *group) const</td></tr>
<tr class="separator:abda7647acd8947583d91e9b86d7b5461"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8959b16acbe263ab172cdffc8a3efb36"><td class="memTemplParams" colspan="2">template&lt;typename Out  = double, typename T , typename IDX , typename Group &gt; </td></tr>
<tr class="memitem:a8959b16acbe263ab172cdffc8a3efb36"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a>&lt; Out &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a8959b16acbe263ab172cdffc8a3efb36">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Group *group, size_t reference) const</td></tr>
<tr class="separator:a8959b16acbe263ab172cdffc8a3efb36"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Compute grouped size factors to handle composition bias. </p>
<p>This implements the grouping approach described in Lun et al. (2016) whereby groups/clusters of cells are used to construct pseudo-cells. These pseudo-cells are normalized against each other using median-based size factors (see <code><a class="el" href="classscran_1_1MedianSizeFactors.html" title="Compute median-based size factors to handle composition bias.">MedianSizeFactors</a></code>) to obtain group-specific scaling factors. Each cell is then normalized against its pseudo-cell using the library size; each cell's size factor is defined as product of its library size-based factor and the median-based factor for its group.</p>
<p>This strategy leverages the reduced sparsity of the pseudo-cells to obtain sensible median-based size factors for removing composition biases, while still generating per-cell factors for computing a normalized single-cell expression matrix in <code><a class="el" href="classscran_1_1LogNormCounts.html" title="Compute log-normalized expression values.">LogNormCounts</a></code>. The assumption is that there are no composition biases within each group; thus, the supplied groupings should correspond to subpopulations in the data, typically generated by clustering.</p>
<dl class="section see"><dt>See also</dt><dd>Lun ATL, Bach K and Marioni JC (2016). Pooling across cells to normalize single-cell RNA sequencing data with many zero counts. <em>Genome Biol.</em> 17:75 </dd></dl>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a907245f06f970a65c5a7c1a4a96af0fc" name="a907245f06f970a65c5a7c1a4a96af0fc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a907245f06f970a65c5a7c1a4a96af0fc">&#9670;&#160;</a></span>set_center()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp; scran::GroupedSizeFactors::set_center </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#acbe5c154e3b67420a1f2ef84ef304310">Defaults::center</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to center the size factors to have a mean of unity. This is usually desirable for interpretation of relative values.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1GroupedSizeFactors.html" title="Compute grouped size factors to handle composition bias.">GroupedSizeFactors</a></code> object.</dd></dl>
<p>For more control over centering, this can be set to <code>false</code> and the resulting size factors can be passed to <code><a class="el" href="classscran_1_1CenterSizeFactors.html" title="Center size factors prior to scaling normalization.">CenterSizeFactors</a></code>. </p>

</div>
</div>
<a id="aec88f173e14eb0df8dcc8e1e9d6db1c0" name="aec88f173e14eb0df8dcc8e1e9d6db1c0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aec88f173e14eb0df8dcc8e1e9d6db1c0">&#9670;&#160;</a></span>set_prior_count()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp; scran::GroupedSizeFactors::set_prior_count </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>p</em> = <code><a class="el" href="structscran_1_1MedianSizeFactors_1_1Defaults.html#a909251588d34d41237a64df1dd85552c">MedianSizeFactors::Defaults::prior_count</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">p</td><td>Prior count for the library size shrinkage, see <code><a class="el" href="classscran_1_1MedianSizeFactors.html#a5507419988ca1925e8f60bdc99436e51">MedianSizeFactors::set_prior_count()</a></code> for details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1GroupedSizeFactors.html" title="Compute grouped size factors to handle composition bias.">GroupedSizeFactors</a></code> object. </dd></dl>

</div>
</div>
<a id="a9471c099549a9d24c2fdc1bdbea98c7d" name="a9471c099549a9d24c2fdc1bdbea98c7d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9471c099549a9d24c2fdc1bdbea98c7d">&#9670;&#160;</a></span>set_handle_zeros()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp; scran::GroupedSizeFactors::set_handle_zeros </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>z</em> = <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#a95d70f82976a430a11c20c1e9a5a068e">Defaults::handle_zeros</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Should we gracefully handle size factors of zero for the pseudo-cells? Note that this does not sanitize the per-cell size factors - to do so, users should call <code><a class="el" href="classscran_1_1SanitizeSizeFactors.html" title="Sanitize invalid size factors.">SanitizeSizeFactors</a></code> separately on the output of <code><a class="el" href="classscran_1_1GroupedSizeFactors.html#afec8d5ba32a045c587f40c832973de2e">run()</a></code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">z</td><td>Whether to replace pseudo-cell size factors of zero with the smallest non-zero size factor across pseudo-cells, see <code>SanitizeSizeFactors::set_handle_zeros()</code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1GroupedSizeFactors.html" title="Compute grouped size factors to handle composition bias.">GroupedSizeFactors</a></code> object. </dd></dl>

</div>
</div>
<a id="a084f2ef8e3d651ebc4604c59850cb360" name="a084f2ef8e3d651ebc4604c59850cb360"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a084f2ef8e3d651ebc4604c59850cb360">&#9670;&#160;</a></span>set_handle_non_finite()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp; scran::GroupedSizeFactors::set_handle_non_finite </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#a0586f933b7cd61f9f07daa8272360d2a">Defaults::handle_non_finite</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Should we gracefully handle non-finite size factors of zero for the pseudo-cells? Note that this does not sanitize the per-cell size factors - to do so, users should call <code><a class="el" href="classscran_1_1SanitizeSizeFactors.html" title="Sanitize invalid size factors.">SanitizeSizeFactors</a></code> separately on the output of <code><a class="el" href="classscran_1_1GroupedSizeFactors.html#afec8d5ba32a045c587f40c832973de2e">run()</a></code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Whether to replace non-finite pseudo-cell size factors with the largest finite size factor across pseudo-cells, see <code><a class="el" href="classscran_1_1SanitizeSizeFactors.html#ac74f5998ee7f28a55545e06badd64f18">SanitizeSizeFactors::set_handle_non_finite()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1GroupedSizeFactors.html" title="Compute grouped size factors to handle composition bias.">GroupedSizeFactors</a></code> object. </dd></dl>

</div>
</div>
<a id="ab9121d8622080b615083d0d2b53a32f1" name="ab9121d8622080b615083d0d2b53a32f1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab9121d8622080b615083d0d2b53a32f1">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp; scran::GroupedSizeFactors::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#a26a37b3247734472ce814d356243c980">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1AggregateAcrossCells.html" title="Aggregate expression values across cells.">AggregateAcrossCells</a></code> object. </dd></dl>

</div>
</div>
<a id="afec8d5ba32a045c587f40c832973de2e" name="afec8d5ba32a045c587f40c832973de2e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afec8d5ba32a045c587f40c832973de2e">&#9670;&#160;</a></span>run() <span class="overload">[1/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T , typename IDX , typename Group , typename Out &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::GroupedSizeFactors::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Group *&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Out *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute per-cell size factors based on user-supplied groupings. The reference group is automatically determined from the pseudo-cell with the highest sum of root-counts, inspired by the <code>calcNormFactors</code> function from the <b>edgeR</b> R package. This approach favors higher-coverage libraries with decent transcriptomic complexity.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Numeric data type of the input matrix. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer index type of the input matrix. </td></tr>
    <tr><td class="paramname">Group</td><td>Integer type for the groupings. </td></tr>
    <tr><td class="paramname">Out</td><td>Numeric data type of the output vector.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Matrix containing non-negative expression data, usually counts. Rows should be genes and columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">group</td><td>Pointer to an array of group identifiers, of length equal to the number of columns in <code>mat</code>. Values should be integers in <picture><source srcset="form_0_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$[0, N)$" src="form_0.png"/></picture> where <picture><source srcset="form_1_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$N$" src="form_1.png"/></picture> is the total number of groups. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array to use to store the output size factors. This should be of length equal to the number of columns in <code>mat</code>. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a2dc18c86bde54ed4bda775d5f2bbac16" name="a2dc18c86bde54ed4bda775d5f2bbac16"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2dc18c86bde54ed4bda775d5f2bbac16">&#9670;&#160;</a></span>run() <span class="overload">[2/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T , typename IDX , typename Group , typename Out &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::GroupedSizeFactors::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Group *&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>reference</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Out *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute per-cell size factors based on user-supplied groupings and a user-specified reference group.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Numeric data type of the input matrix. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer index type of the input matrix. </td></tr>
    <tr><td class="paramname">Group</td><td>Integer type for the groupings. </td></tr>
    <tr><td class="paramname">Out</td><td>Numeric data type of the output vector.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Matrix containing non-negative expression data, usually counts. Rows should be genes and columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">group</td><td>Pointer to an array of group identifiers, of length equal to the number of columns in <code>mat</code>. Values should be integers in <picture><source srcset="form_0_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$[0, N)$" src="form_0.png"/></picture> where <picture><source srcset="form_1_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$N$" src="form_1.png"/></picture> is the total number of groups. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">reference</td><td>Identifier of the group to use as the reference. This should be an integer in <picture><source srcset="form_0_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$[0, N)$" src="form_0.png"/></picture>. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array to use to store the output size factors. This should be of length equal to the number of columns in <code>mat</code>. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="abda7647acd8947583d91e9b86d7b5461" name="abda7647acd8947583d91e9b86d7b5461"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abda7647acd8947583d91e9b86d7b5461">&#9670;&#160;</a></span>run() <span class="overload">[3/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Out  = double, typename T , typename IDX , typename Group &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a>&lt; Out &gt; scran::GroupedSizeFactors::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Group *&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute per-cell size factors based on user-supplied groupings. The reference sample is automatically chosen, see <code><a class="el" href="classscran_1_1GroupedSizeFactors.html#afec8d5ba32a045c587f40c832973de2e">run()</a></code> for details.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Numeric data type of the input matrix. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer index type of the input matrix. </td></tr>
    <tr><td class="paramname">Group</td><td>Integer type for the groupings.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Matrix containing non-negative expression data, usually counts. Rows should be genes and columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">group</td><td>Pointer to an array of group identifiers, of length equal to the number of columns in <code>mat</code>. Values should be integers in <picture><source srcset="form_0_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$[0, N)$" src="form_0.png"/></picture> where <picture><source srcset="form_1_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$N$" src="form_1.png"/></picture> is the total number of groups.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html" title="Result of the size factor calculations.">Results</a></code> object is returned containing the size factors. </dd></dl>

</div>
</div>
<a id="a8959b16acbe263ab172cdffc8a3efb36" name="a8959b16acbe263ab172cdffc8a3efb36"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8959b16acbe263ab172cdffc8a3efb36">&#9670;&#160;</a></span>run() <span class="overload">[4/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Out  = double, typename T , typename IDX , typename Group &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a>&lt; Out &gt; scran::GroupedSizeFactors::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Group *&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>reference</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute per-cell size factors based on user-supplied groupings and a user-specified grouping.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Numeric data type of the input matrix. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer index type of the input matrix. </td></tr>
    <tr><td class="paramname">Group</td><td>Integer type for the groupings.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Matrix containing non-negative expression data, usually counts. Rows should be genes and columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">group</td><td>Pointer to an array of group identifiers, of length equal to the number of columns in <code>mat</code>. Values should be integers in <picture><source srcset="form_0_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$[0, N)$" src="form_0.png"/></picture> where <picture><source srcset="form_1_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$N$" src="form_1.png"/></picture> is the total number of groups. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">reference</td><td>Identifier of the group to use as the reference. This should be an integer in <picture><source srcset="form_0_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$[0, N)$" src="form_0.png"/></picture>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html" title="Result of the size factor calculations.">Results</a></code> object is returned containing the size factors. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/normalization/<a class="el" href="GroupedSizeFactors_8hpp_source.html">GroupedSizeFactors.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.7
</small></address>
</body>
</html>
