<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.7"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::ModelGeneVariances Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.7 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1ModelGeneVariances-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::ModelGeneVariances Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute and model the per-gene variances in log-expression data.  
 <a href="classscran_1_1ModelGeneVariances.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="ModelGeneVariances_8hpp_source.html">ModelGeneVariances.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVariances_1_1BlockResults.html">BlockResults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1ModelGeneVariances_1_1Results.html" title="Results of variance modelling without blocks.">Results</a> of variance modelling with blocks.  <a href="structscran_1_1ModelGeneVariances_1_1BlockResults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameters for variance modelling.  <a href="structscran_1_1ModelGeneVariances_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1ModelGeneVariances_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1ModelGeneVariances_1_1Results.html" title="Results of variance modelling without blocks.">Results</a> of variance modelling without blocks.  <a href="structscran_1_1ModelGeneVariances_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a132525a72af22ab9788ba1fc37b8c310"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a132525a72af22ab9788ba1fc37b8c310">set_span</a> (double s=<a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a1064a561d96d74640c24ec83e579d37c">FitVarianceTrend::Defaults::span</a>)</td></tr>
<tr class="separator:a132525a72af22ab9788ba1fc37b8c310"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97acbaa38d60b852777c4807c3ba2620"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a97acbaa38d60b852777c4807c3ba2620">set_minimum_mean</a> (double m=<a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a3fa5c15af12eb7f708af93dfde52f8fc">FitVarianceTrend::Defaults::minimum_mean</a>)</td></tr>
<tr class="separator:a97acbaa38d60b852777c4807c3ba2620"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7199a9128ffe7c420b6198aa7d9a8ef6"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a7199a9128ffe7c420b6198aa7d9a8ef6">set_use_fixed_width</a> (bool u=<a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a8861c758b3077b2243fb93e1c4232d00">FitVarianceTrend::Defaults::use_fixed_width</a>)</td></tr>
<tr class="separator:a7199a9128ffe7c420b6198aa7d9a8ef6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a6aa29d2a8088a9663eb44799a53ca1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a5a6aa29d2a8088a9663eb44799a53ca1">set_fixed_width</a> (double f=<a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#af134fa2aa91836c7b9d7955e79f43084">FitVarianceTrend::Defaults::fixed_width</a>)</td></tr>
<tr class="separator:a5a6aa29d2a8088a9663eb44799a53ca1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7e0c8f5bf7fa1699448483ec26e52951"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a7e0c8f5bf7fa1699448483ec26e52951">set_minimum_window_count</a> (int c=<a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a0f17da67b56f5b46f608661454ab1beb">FitVarianceTrend::Defaults::minimum_window_count</a>)</td></tr>
<tr class="separator:a7e0c8f5bf7fa1699448483ec26e52951"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a443daf250ab0cbf48dfaa3816442a45e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a443daf250ab0cbf48dfaa3816442a45e">set_block_weight_policy</a> (<a class="el" href="namespacescran.html#abcd041d625290609e221552910ae2c34">WeightPolicy</a> w=<a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#aa56d09d541aa8cb6a62aa75d910470fd">Defaults::block_weight_policy</a>)</td></tr>
<tr class="separator:a443daf250ab0cbf48dfaa3816442a45e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3fe070bda91379b20db1f516cffb0a19"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a3fe070bda91379b20db1f516cffb0a19">set_variable_block_weight_parameters</a> (<a class="el" href="structscran_1_1VariableBlockWeightParameters.html">VariableBlockWeightParameters</a> v=<a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#a6eeeca2eaa558e3ed0e4452ac625e063">Defaults::variable_block_weight_parameters</a>)</td></tr>
<tr class="separator:a3fe070bda91379b20db1f516cffb0a19"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b9ca0e93dbb98159df0f7dd340ed69b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a9b9ca0e93dbb98159df0f7dd340ed69b">set_compute_average</a> (bool a=<a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#acd15c6eb4576a70150096f05099f6d9b">Defaults::compute_average</a>)</td></tr>
<tr class="separator:a9b9ca0e93dbb98159df0f7dd340ed69b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd751bd1d661e1d9882f6e83fc0a0aa6"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#abd751bd1d661e1d9882f6e83fc0a0aa6">set_num_threads</a> (int n=<a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#a46f4dda7a0a51463872432d73cbe0823">Defaults::num_threads</a>)</td></tr>
<tr class="separator:abd751bd1d661e1d9882f6e83fc0a0aa6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab86e8b75d53cbc77a09c65096e012389"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:ab86e8b75d53cbc77a09c65096e012389"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#ab86e8b75d53cbc77a09c65096e012389">run</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, Stat_ *means, Stat_ *variances, Stat_ *fitted, Stat_ *residuals) const</td></tr>
<tr class="separator:ab86e8b75d53cbc77a09c65096e012389"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad50b110c52beb8a20e0bba25c3fb2ee9"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:ad50b110c52beb8a20e0bba25c3fb2ee9"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#ad50b110c52beb8a20e0bba25c3fb2ee9">run_blocked</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block, std::vector&lt; Stat_ * &gt; means, std::vector&lt; Stat_ * &gt; variances, std::vector&lt; Stat_ * &gt; fitted, std::vector&lt; Stat_ * &gt; residuals, Stat_ *ave_means, Stat_ *ave_variances, Stat_ *ave_fitted, Stat_ *ave_residuals) const</td></tr>
<tr class="separator:ad50b110c52beb8a20e0bba25c3fb2ee9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a57cde36a7e287a4efe045b58ff0095f6"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </td></tr>
<tr class="memitem:a57cde36a7e287a4efe045b58ff0095f6"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a57cde36a7e287a4efe045b58ff0095f6">run_blocked</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block, std::vector&lt; Stat_ * &gt; means, std::vector&lt; Stat_ * &gt; variances, std::vector&lt; Stat_ * &gt; fitted, std::vector&lt; Stat_ * &gt; residuals) const</td></tr>
<tr class="separator:a57cde36a7e287a4efe045b58ff0095f6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3fe3fb84975a5ab2374c110c0efc2731"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ &gt; </td></tr>
<tr class="memitem:a3fe3fb84975a5ab2374c110c0efc2731"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1ModelGeneVariances_1_1Results.html">Results</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#a3fe3fb84975a5ab2374c110c0efc2731">run</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat) const</td></tr>
<tr class="separator:a3fe3fb84975a5ab2374c110c0efc2731"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad89f68945b1ea031940b1424c823dd6"><td class="memTemplParams" colspan="2">template&lt;typename Value_ , typename Index_ , typename Block_ &gt; </td></tr>
<tr class="memitem:aad89f68945b1ea031940b1424c823dd6"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1ModelGeneVariances_1_1BlockResults.html">BlockResults</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1ModelGeneVariances.html#aad89f68945b1ea031940b1424c823dd6">run_blocked</a> (const tatami::Matrix&lt; Value_, Index_ &gt; *mat, const Block_ *block) const</td></tr>
<tr class="separator:aad89f68945b1ea031940b1424c823dd6"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Compute and model the per-gene variances in log-expression data. </p>
<p>This scans through a log-transformed normalized expression matrix (e.g., from <code><a class="el" href="classscran_1_1LogNormCounts.html" title="Compute log-normalized expression values.">LogNormCounts</a></code>) and computes per-feature means and variances. It then fits a trend to the variances with respect to the means using <code><a class="el" href="classscran_1_1FitVarianceTrend.html" title="Fit a mean-variance trend to log-count data.">FitVarianceTrend</a></code>. We assume that most genes at any given abundance are not highly variable, such that the fitted value of the trend is interpreted as the "uninteresting" variance - this is mostly attributed to technical variation like sequencing noise, but can also represent constitutive biological noise like transcriptional bursting. Under this assumption, the residual can be treated as a quantification of biologically interesting variation, and can be used to identify relevant features for downstream analyses. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a132525a72af22ab9788ba1fc37b8c310" name="a132525a72af22ab9788ba1fc37b8c310"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a132525a72af22ab9788ba1fc37b8c310">&#9670;&#160;</a></span>set_span()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_span </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>s</em> = <code><a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a1064a561d96d74640c24ec83e579d37c">FitVarianceTrend::Defaults::span</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">s</td><td>See <code><a class="el" href="classscran_1_1FitVarianceTrend.html#acd04112803e8a834b7105c4666b816dc">FitVarianceTrend::set_span()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="a97acbaa38d60b852777c4807c3ba2620" name="a97acbaa38d60b852777c4807c3ba2620"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a97acbaa38d60b852777c4807c3ba2620">&#9670;&#160;</a></span>set_minimum_mean()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_minimum_mean </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>m</em> = <code><a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a3fa5c15af12eb7f708af93dfde52f8fc">FitVarianceTrend::Defaults::minimum_mean</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>See <code><a class="el" href="classscran_1_1FitVarianceTrend.html#a71a308e561da1a904f60962f422edc3d">FitVarianceTrend::set_minimum_mean()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="a7199a9128ffe7c420b6198aa7d9a8ef6" name="a7199a9128ffe7c420b6198aa7d9a8ef6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7199a9128ffe7c420b6198aa7d9a8ef6">&#9670;&#160;</a></span>set_use_fixed_width()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_use_fixed_width </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>u</em> = <code><a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a8861c758b3077b2243fb93e1c4232d00">FitVarianceTrend::Defaults::use_fixed_width</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">u</td><td>See <code><a class="el" href="classscran_1_1FitVarianceTrend.html#ad1855398708d5426f3b76e5ff94c58e3">FitVarianceTrend::set_use_fixed_width()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="a5a6aa29d2a8088a9663eb44799a53ca1" name="a5a6aa29d2a8088a9663eb44799a53ca1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5a6aa29d2a8088a9663eb44799a53ca1">&#9670;&#160;</a></span>set_fixed_width()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_fixed_width </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>f</em> = <code><a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#af134fa2aa91836c7b9d7955e79f43084">FitVarianceTrend::Defaults::fixed_width</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>See <code><a class="el" href="classscran_1_1FitVarianceTrend.html#a9f336bf62a54fe747b022ad5b056edc8">FitVarianceTrend::set_fixed_width()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="a7e0c8f5bf7fa1699448483ec26e52951" name="a7e0c8f5bf7fa1699448483ec26e52951"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7e0c8f5bf7fa1699448483ec26e52951">&#9670;&#160;</a></span>set_minimum_window_count()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_minimum_window_count </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1FitVarianceTrend_1_1Defaults.html#a0f17da67b56f5b46f608661454ab1beb">FitVarianceTrend::Defaults::minimum_window_count</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>See <code><a class="el" href="classscran_1_1FitVarianceTrend.html#a1cd339b02701023a32cb18b0359dac6e">FitVarianceTrend::set_minimum_window_count()</a></code> for more details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="a443daf250ab0cbf48dfaa3816442a45e" name="a443daf250ab0cbf48dfaa3816442a45e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a443daf250ab0cbf48dfaa3816442a45e">&#9670;&#160;</a></span>set_block_weight_policy()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_block_weight_policy </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="namespacescran.html#abcd041d625290609e221552910ae2c34">WeightPolicy</a>&#160;</td>
          <td class="paramname"><em>w</em> = <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#aa56d09d541aa8cb6a62aa75d910470fd">Defaults::block_weight_policy</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">w</td><td>Weighting policy to use for averaging statistics across blocks.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> instance. </dd></dl>

</div>
</div>
<a id="a3fe070bda91379b20db1f516cffb0a19" name="a3fe070bda91379b20db1f516cffb0a19"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3fe070bda91379b20db1f516cffb0a19">&#9670;&#160;</a></span>set_variable_block_weight_parameters()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_variable_block_weight_parameters </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structscran_1_1VariableBlockWeightParameters.html">VariableBlockWeightParameters</a>&#160;</td>
          <td class="paramname"><em>v</em> = <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#a6eeeca2eaa558e3ed0e4452ac625e063">Defaults::variable_block_weight_parameters</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">v</td><td>Parameters for the variable block weights, see <code><a class="el" href="namespacescran.html#a128f43444ef77988f27ae5b91cef6b13">variable_block_weight()</a></code> for more details. Only used when the block weight policy is set to <code>WeightPolicy::VARIABLE</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> instance. </dd></dl>

</div>
</div>
<a id="a9b9ca0e93dbb98159df0f7dd340ed69b" name="a9b9ca0e93dbb98159df0f7dd340ed69b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9b9ca0e93dbb98159df0f7dd340ed69b">&#9670;&#160;</a></span>set_compute_average()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_compute_average </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>a</em> = <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#acd15c6eb4576a70150096f05099f6d9b">Defaults::compute_average</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>Whether to compute the average of each statistic across blocks. Note that this only affects the <code><a class="el" href="classscran_1_1ModelGeneVariances.html#ad50b110c52beb8a20e0bba25c3fb2ee9">run_blocked()</a></code> method that returns a <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1BlockResults.html" title="Results of variance modelling with blocks.">BlockResults</a></code> object. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="abd751bd1d661e1d9882f6e83fc0a0aa6" name="abd751bd1d661e1d9882f6e83fc0a0aa6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abd751bd1d661e1d9882f6e83fc0a0aa6">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1ModelGeneVariances.html">ModelGeneVariances</a> &amp; scran::ModelGeneVariances::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1Defaults.html#a46f4dda7a0a51463872432d73cbe0823">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1ModelGeneVariances.html" title="Compute and model the per-gene variances in log-expression data.">ModelGeneVariances</a></code> object. </dd></dl>

</div>
</div>
<a id="ab86e8b75d53cbc77a09c65096e012389" name="ab86e8b75d53cbc77a09c65096e012389"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab86e8b75d53cbc77a09c65096e012389">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Stat_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVariances::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix. This returns the mean and variance for each feature, as well as the fitted value and residuals from the mean-variance trend fitted across features.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the mean of each feature. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the variance of each feature. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the fitted value of the trend. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Pointer to an output array of length equal to the number of rows in <code>mat</code>, used to store the residual from the trend for each feature. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ad50b110c52beb8a20e0bba25c3fb2ee9" name="ad50b110c52beb8a20e0bba25c3fb2ee9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad50b110c52beb8a20e0bba25c3fb2ee9">&#9670;&#160;</a></span>run_blocked() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVariances::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>residuals</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Stat_ *&#160;</td>
          <td class="paramname"><em>ave_residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking. The mean and variance of each gene is computed separately for all cells in each block, and a separate trend is fitted to each block to obtain residuals. This ensures that sample and batch effects do not confound the variance estimates.</p>
<p>We also compute the average of each statistic across blocks, using the weighting strategy described in <code>weight_block()</code>. The average residual is particularly useful for feature selection with <code>ChooseHVGs</code>.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type to hold the block IDs. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing a 0-based block ID for each cell - see <code><a class="el" href="namespacescran.html#a442c2327430d3e4c425beebfe4dcd768">tabulate_ids()</a></code> for more details. This can also be a <code>nullptr</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the mean of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the variance of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the fitted value of the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the residual from the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_means</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average mean across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_variances</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average variance across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_fitted</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average fitted value across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">ave_residuals</td><td>Pointer to an array of length equal to the number of rows in <code>mat</code>, storing the average residual across blocks for each gene. If <code>nullptr</code>, the average calculation is skipped. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a57cde36a7e287a4efe045b58ff0095f6" name="a57cde36a7e287a4efe045b58ff0095f6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a57cde36a7e287a4efe045b58ff0095f6">&#9670;&#160;</a></span>run_blocked() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ , typename Stat_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::ModelGeneVariances::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>means</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>variances</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>fitted</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; Stat_ * &gt;&#160;</td>
          <td class="paramname"><em>residuals</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking. This overload omits the calculation of the averaged statistics across blocks.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type to hold the block IDs. </td></tr>
    <tr><td class="paramname">Stat_</td><td>Floating-point type for the output statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing block identifiers - see <code><a class="el" href="namespacescran.html#a442c2327430d3e4c425beebfe4dcd768">tabulate_ids()</a></code> for more details. This can also be a <code>nullptr</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">means</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the mean of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">variances</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the variance of each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">fitted</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the fitted value of the trend for each feature in the corresponding block of cells. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">residuals</td><td>Vector of length equal to the number of blocks, containing pointers to output arrays of length equal to the number of rows in <code>mat</code>. Each vector stores the residual from the trend for each feature in the corresponding block of cells. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a3fe3fb84975a5ab2374c110c0efc2731" name="a3fe3fb84975a5ab2374c110c0efc2731"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3fe3fb84975a5ab2374c110c0efc2731">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1ModelGeneVariances_1_1Results.html">Results</a> scran::ModelGeneVariances::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1Results.html" title="Results of variance modelling without blocks.">Results</a></code> object containing the results of the variance modelling. </dd></dl>

</div>
</div>
<a id="aad89f68945b1ea031940b1424c823dd6" name="aad89f68945b1ea031940b1424c823dd6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aad89f68945b1ea031940b1424c823dd6">&#9670;&#160;</a></span>run_blocked() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Value_ , typename Index_ , typename Block_ &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1ModelGeneVariances_1_1BlockResults.html">BlockResults</a> scran::ModelGeneVariances::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; Value_, Index_ &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block_ *&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Compute and model the per-feature variances from a log-expression matrix with blocking, see <code><a class="el" href="classscran_1_1ModelGeneVariances.html#ad50b110c52beb8a20e0bba25c3fb2ee9">run_blocked()</a></code> for details.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Value_</td><td>Data type of the matrix. </td></tr>
    <tr><td class="paramname">Index_</td><td>Integer type for the row/column indices. </td></tr>
    <tr><td class="paramname">Block_</td><td>Integer type, to hold the block IDs.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Pointer to a feature-by-cells <b>tatami</b> matrix containing log-expression values. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of block identifiers, see <code><a class="el" href="classscran_1_1ModelGeneVariances.html#ad50b110c52beb8a20e0bba25c3fb2ee9">run_blocked()</a></code> for details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1ModelGeneVariances_1_1BlockResults.html" title="Results of variance modelling with blocks.">BlockResults</a></code> object containing the results of the variance modelling in each block. An average for each statistic is also computed by default unless <code><a class="el" href="classscran_1_1ModelGeneVariances.html#a9b9ca0e93dbb98159df0f7dd340ed69b">set_compute_average()</a></code> is set to <code>false</code>. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/feature_selection/<a class="el" href="ModelGeneVariances_8hpp_source.html">ModelGeneVariances.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.7
</small></address>
</body>
</html>
