<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.6"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::SummarizeEffects Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.6 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1SummarizeEffects-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::SummarizeEffects Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Summarize pairwise effects into summary statistics per group.  
 <a href="classscran_1_1SummarizeEffects.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="SummarizeEffects_8hpp_source.html">SummarizeEffects.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter settings.  <a href="structscran_1_1SummarizeEffects_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a579d507532421a29a72a4ead040f2611"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a579d507532421a29a72a4ead040f2611">set_num_threads</a> (int n=<a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#a961f7e01b98c5c365773a20e32cc2ac3">Defaults::num_threads</a>)</td></tr>
<tr class="separator:a579d507532421a29a72a4ead040f2611"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1dce14dd1e4554c724221e8525952aac"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a1dce14dd1e4554c724221e8525952aac">set_compute_min</a> (bool c=<a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#aef011af1e9bbcd7e606b4e16307e8212">Defaults::compute_min</a>)</td></tr>
<tr class="separator:a1dce14dd1e4554c724221e8525952aac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a795e9645291e0ff963f47b98ee342640"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a795e9645291e0ff963f47b98ee342640">set_compute_mean</a> (bool c=<a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#af17ebeef3fbefc11fef6e3732c524f8c">Defaults::compute_mean</a>)</td></tr>
<tr class="separator:a795e9645291e0ff963f47b98ee342640"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a40620cf6b6bdcf57861ee6d428e9cee2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a40620cf6b6bdcf57861ee6d428e9cee2">set_compute_median</a> (bool c=<a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#a27953b061082adad335171f7d17cb023">Defaults::compute_median</a>)</td></tr>
<tr class="separator:a40620cf6b6bdcf57861ee6d428e9cee2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4078a63f650cbb7e538347e9c54d0ec"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#ac4078a63f650cbb7e538347e9c54d0ec">set_compute_max</a> (bool c=<a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#a55d321e03e94bbaca62d75eb1cc10c81">Defaults::compute_max</a>)</td></tr>
<tr class="separator:ac4078a63f650cbb7e538347e9c54d0ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1fcdfec45f1bc999eee3b749727388ca"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a1fcdfec45f1bc999eee3b749727388ca">set_compute_min_rank</a> (bool c=<a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#af3511c0b77e50addde9d2b9fc782f39f">Defaults::compute_min_rank</a>)</td></tr>
<tr class="separator:a1fcdfec45f1bc999eee3b749727388ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a13302827e97fd2b05f4b421a8c2a13d1"><td class="memTemplParams" colspan="2">template&lt;typename Stat &gt; </td></tr>
<tr class="memitem:a13302827e97fd2b05f4b421a8c2a13d1"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a13302827e97fd2b05f4b421a8c2a13d1">run</a> (size_t ngenes, size_t ngroups, const Stat *effects, std::vector&lt; std::vector&lt; Stat * &gt; &gt; summaries) const</td></tr>
<tr class="separator:a13302827e97fd2b05f4b421a8c2a13d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a422e9dfd8e19edf2716595d4b0d87a1d"><td class="memTemplParams" colspan="2">template&lt;typename Stat &gt; </td></tr>
<tr class="memitem:a422e9dfd8e19edf2716595d4b0d87a1d"><td class="memTemplItemLeft" align="right" valign="top">std::vector&lt; std::vector&lt; std::vector&lt; Stat &gt; &gt; &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SummarizeEffects.html#a422e9dfd8e19edf2716595d4b0d87a1d">run</a> (size_t ngenes, size_t ngroups, const Stat *effects) const</td></tr>
<tr class="separator:a422e9dfd8e19edf2716595d4b0d87a1d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Summarize pairwise effects into summary statistics per group. </p>
<p>This class computes the statistics that are used for marker detection in <code><a class="el" href="classscran_1_1ScoreMarkers.html" title="Score each gene as a candidate marker for each group of cells.">ScoreMarkers</a></code>. Briefly, given <code>n</code> groups, each group is involved in <code>n - 1</code> pairwise comparisons and thus has <code>n - 1</code> effect sizes, as computed by <code><a class="el" href="classscran_1_1PairwiseEffects.html" title="Compute pairwise effect size between groups of cells.">PairwiseEffects</a></code>. For each group, we compute summary statistics - e.g., the minimum, median, mean - of the effect sizes across all of that group's comparisons. Users can then sort by any of these summaries to obtain a ranking of potential marker genes for each group.</p>
<p>The choice of summary statistic dictates the interpretation of the ranking. Given a group X:</p>
<ul>
<li>A large mean effect size indicates that the gene is upregulated in X compared to the average of the other groups. A small value indicates that the gene is downregulated in X instead. This is a good general-purpose summary statistic for ranking, usually by decreasing size to obtain upregulated markers in X.</li>
<li>A large median effect size indicates that the gene is upregulated in X compared to most (&gt;50%) other groups. A small value indicates that the gene is downregulated in X instead. This is also a good general-purpose summary, with the advantage of being more robust to outlier effects compared to the mean. However, it also has the disadvantage of being less sensitive to strong effects in a minority of comparisons.</li>
<li>A large minimum effect size indicates that the gene is upregulated in X compared to all other groups. A small value indicates that the gene is downregulated in X compared to at least one other group. For upregulation, this is the most stringent summary as markers will only have extreme values if they are <em>uniquely</em> upregulated in X compared to every other group. However, it may not be effective if X is closely related to any of the groups.</li>
<li>A large maximum effect size indicates that the gene is upregulated in X compared to at least one other group. A small value indicates that the gene is downregulated in X compared to all other groups. For downregulation, this is the most stringent summary as markers will only have extreme values if they are <em>uniquely</em> downregulated in X compared to every other group. However, it may not be effective if X is closely related to any of the groups.</li>
<li>The "minimum rank" (a.k.a. min-rank) is defined by ranking genes based on decreasing effect size <em>within</em> each comparison, and then taking the smallest rank <em>across</em> comparisons. A minimum rank of 1 means that the gene is the top upregulated gene in at least one comparison to another group. More generally, a minimum rank of T indicates that the gene is the T-th upregulated gene in at least one comparison. Applying a threshold on the minimum rank is useful for obtaining a set of genes that, in combination, are guaranteed to distinguish X from every other group.</li>
</ul>
<p>The exact definition of "large" and "small" depends on the choice of effect size from <code><a class="el" href="classscran_1_1PairwiseEffects.html" title="Compute pairwise effect size between groups of cells.">PairwiseEffects</a></code>. For Cohen's d, LFC and delta-detected, the value must be positive to be considered "large", and negative to be considered "small". For the AUC, a value greater than 0.5 is considered "large" and less than 0.5 is considered "small".</p>
<p>The interpretation above is also contingent on the log-fold change threshold used in <code><a class="el" href="classscran_1_1PairwiseEffects.html" title="Compute pairwise effect size between groups of cells.">PairwiseEffects</a></code>. For positive thresholds, small effects cannot be unambiguously interpreted as downregulation, as the effect is already adjusted to account for the threshold. As a result, only large effects can be interpreted as evidence for upregulation.</p>
<p><code>NaN</code> effect sizes are allowed, e.g., if two groups do not exist in the same block for a blocked analysis in <code><a class="el" href="classscran_1_1PairwiseEffects.html" title="Compute pairwise effect size between groups of cells.">PairwiseEffects</a></code>. This class will ignore <code>NaN</code> values when computing each summary. If all effects are <code>NaN</code> for a particular group, the summary statistic will also be <code>NaN</code>.</p>
<p>All choices of summary statistics are enumerated by <code><a class="el" href="namespacescran_1_1differential__analysis.html#a9651cedd1ef5f3c2631ff3d7af955b23">differential_analysis::summary</a></code>. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a579d507532421a29a72a4ead040f2611" name="a579d507532421a29a72a4ead040f2611"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a579d507532421a29a72a4ead040f2611">&#9670;&#160;</a></span>set_num_threads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp; scran::SummarizeEffects::set_num_threads </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#a961f7e01b98c5c365773a20e32cc2ac3">Defaults::num_threads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of threads to use. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group.">SummarizeEffects</a></code> object. </dd></dl>

</div>
</div>
<a id="a1dce14dd1e4554c724221e8525952aac" name="a1dce14dd1e4554c724221e8525952aac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1dce14dd1e4554c724221e8525952aac">&#9670;&#160;</a></span>set_compute_min()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp; scran::SummarizeEffects::set_compute_min </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#aef011af1e9bbcd7e606b4e16307e8212">Defaults::compute_min</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to report the minimum of the pairwise effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group.">SummarizeEffects</a></code> object.</dd></dl>
<p>This has no effect on the <code><a class="el" href="classscran_1_1SummarizeEffects.html#a13302827e97fd2b05f4b421a8c2a13d1">run()</a></code> overload that accepts a <code>summaries</code> vector. For this method, the minimum is calculated if <code>summaries[differential_analysis::MIN]</code> is of non-zero length. </p>

</div>
</div>
<a id="a795e9645291e0ff963f47b98ee342640" name="a795e9645291e0ff963f47b98ee342640"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a795e9645291e0ff963f47b98ee342640">&#9670;&#160;</a></span>set_compute_mean()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp; scran::SummarizeEffects::set_compute_mean </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#af17ebeef3fbefc11fef6e3732c524f8c">Defaults::compute_mean</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to report the mean of the pairwise effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group.">SummarizeEffects</a></code> object.</dd></dl>
<p>This has no effect on the <code><a class="el" href="classscran_1_1SummarizeEffects.html#a13302827e97fd2b05f4b421a8c2a13d1">run()</a></code> overload that accepts a <code>summaries</code> vector. For this method, the minimum is calculated if <code>summaries[differential_analysis::MEAN]</code> is of non-zero length. </p>

</div>
</div>
<a id="a40620cf6b6bdcf57861ee6d428e9cee2" name="a40620cf6b6bdcf57861ee6d428e9cee2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a40620cf6b6bdcf57861ee6d428e9cee2">&#9670;&#160;</a></span>set_compute_median()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp; scran::SummarizeEffects::set_compute_median </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#a27953b061082adad335171f7d17cb023">Defaults::compute_median</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to report the median of the pairwise effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group.">SummarizeEffects</a></code> object.</dd></dl>
<p>This has no effect on the <code><a class="el" href="classscran_1_1SummarizeEffects.html#a13302827e97fd2b05f4b421a8c2a13d1">run()</a></code> overload that accepts a <code>summaries</code> vector. For this method, the minimum is calculated if <code>summaries[differential_analysis::MEDIAN]</code> is of non-zero length. </p>

</div>
</div>
<a id="ac4078a63f650cbb7e538347e9c54d0ec" name="ac4078a63f650cbb7e538347e9c54d0ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac4078a63f650cbb7e538347e9c54d0ec">&#9670;&#160;</a></span>set_compute_max()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp; scran::SummarizeEffects::set_compute_max </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#a55d321e03e94bbaca62d75eb1cc10c81">Defaults::compute_max</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to report the maximum of the pairwise effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group.">SummarizeEffects</a></code> object.</dd></dl>
<p>This has no effect on the <code><a class="el" href="classscran_1_1SummarizeEffects.html#a13302827e97fd2b05f4b421a8c2a13d1">run()</a></code> overload that accepts a <code>summaries</code> vector. For this method, the minimum is calculated if <code>summaries[differential_analysis::MAX]</code> is of non-zero length. </p>

</div>
</div>
<a id="a1fcdfec45f1bc999eee3b749727388ca" name="a1fcdfec45f1bc999eee3b749727388ca"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1fcdfec45f1bc999eee3b749727388ca">&#9670;&#160;</a></span>set_compute_min_rank()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SummarizeEffects.html">SummarizeEffects</a> &amp; scran::SummarizeEffects::set_compute_min_rank </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1SummarizeEffects_1_1Defaults.html#af3511c0b77e50addde9d2b9fc782f39f">Defaults::compute_min_rank</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to report the min-rank of the pairwise effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group.">SummarizeEffects</a></code> object.</dd></dl>
<p>This has no effect on the <code><a class="el" href="classscran_1_1SummarizeEffects.html#a13302827e97fd2b05f4b421a8c2a13d1">run()</a></code> overload that accepts a <code>summaries</code> vector. For this method, the minimum is calculated if <code>summaries[differential_analysis::MIN_RANK]</code> is of non-zero length. </p>

</div>
</div>
<a id="a13302827e97fd2b05f4b421a8c2a13d1" name="a13302827e97fd2b05f4b421a8c2a13d1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a13302827e97fd2b05f4b421a8c2a13d1">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Stat &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::SummarizeEffects::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ngenes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ngroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Stat *&#160;</td>
          <td class="paramname"><em>effects</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::vector&lt; std::vector&lt; Stat * &gt; &gt;&#160;</td>
          <td class="paramname"><em>summaries</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Summarize the effect sizes for the pairwise comparisons to obtain a set of summary statistics for each gene in each group.</p>
<p>If <code>summaries</code> is of length 0, no summaries are computed. If any of the inner vectors of <code>summaries</code> are of length 0, the corresponding summary statistic is not computed.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Stat</td><td>Floating point type for the statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">ngenes</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ngroups</td><td>Number of groups. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">effects</td><td>Pointer to a 3-dimensional array containing the pairwise statistics, see <code><a class="el" href="structscran_1_1PairwiseEffects_1_1Results.html" title="Pairwise effect size results.">PairwiseEffects::Results</a></code> for details. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">summaries</td><td>Vector of vector of pointers to arrays of length equal to the number of genes. The vector should be of length equal to <code>differential_analysis::n_summaries</code> (see <code><a class="el" href="namespacescran_1_1differential__analysis.html#a9651cedd1ef5f3c2631ff3d7af955b23">differential_analysis::summary</a></code>). Each inner vector corresponds to a summary statistic - i.e., minimum, mean, median, maximum and min-rank - and should be of length equal to the number of groups. Each pointer corresponds to a group, and points to an array that is used to store the associated summary statistic across all genes for that group. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a422e9dfd8e19edf2716595d4b0d87a1d" name="a422e9dfd8e19edf2716595d4b0d87a1d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a422e9dfd8e19edf2716595d4b0d87a1d">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Stat &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; std::vector&lt; std::vector&lt; Stat &gt; &gt; &gt; scran::SummarizeEffects::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ngenes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>ngroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Stat *&#160;</td>
          <td class="paramname"><em>effects</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Summarize the effect sizes for the pairwise comparisons to obtain a set of summary statistics for each gene in each group.</p>
<p>If <code>summaries</code> is of length 0, no summaries are computed. If any of the inner vectors of <code>summaries</code> are of length 0, the corresponding summary statistic is not computed.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Stat</td><td>Floating point type for the statistics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">ngenes</td><td>Number of genes. </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">ngroups</td><td>Number of groups. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">effects</td><td>Pointer to a 3-dimensional array containing the pairwise statistics, see <code><a class="el" href="structscran_1_1PairwiseEffects_1_1Results.html" title="Pairwise effect size results.">PairwiseEffects::Results</a></code> for details.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A vector of vectors of vectors containing summary effects for each gene in each group. The vector is of length equal to <code>differential_analysis::n_summaries</code> (see <code><a class="el" href="namespacescran_1_1differential__analysis.html#a9651cedd1ef5f3c2631ff3d7af955b23">differential_analysis::summary</a></code>). Each inner vector corresponds to a summary statistic - i.e., minimum, mean, median, maximum and min-rank - and is of length equal to the number of groups. Each pointer corresponds to a group, and points to an array containing the associated summary statistic across all genes for that group. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/differential_analysis/<a class="el" href="SummarizeEffects_8hpp_source.html">SummarizeEffects.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.6
</small></address>
</body>
</html>
