<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.5"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::SuggestRnaQcFilters Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.5 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacescran.html">scran</a></li><li class="navelem"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1SuggestRnaQcFilters-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::SuggestRnaQcFilters Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Create filters to identify low-quality cells from RNA-derived QC metrics.  
 <a href="classscran_1_1SuggestRnaQcFilters.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="SuggestRnaQcFilters_8hpp_source.html">SuggestRnaQcFilters.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameters.  <a href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html" title="Thresholds to define outliers on each metric.">Thresholds</a> to define outliers on each metric.  <a href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ac98f91f26febaaf787723642900e217a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#ac98f91f26febaaf787723642900e217a">set_detected_num_mads</a> (double n=<a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a>)</td></tr>
<tr class="separator:ac98f91f26febaaf787723642900e217a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abfd616df329cc6b9a1171e3bf8969f8d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#abfd616df329cc6b9a1171e3bf8969f8d">set_sums_num_mads</a> (double n=<a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a>)</td></tr>
<tr class="separator:abfd616df329cc6b9a1171e3bf8969f8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac7dbea3869bc4b9efa14296d206badf1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#ac7dbea3869bc4b9efa14296d206badf1">set_subset_num_mads</a> (double n=<a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a>)</td></tr>
<tr class="separator:ac7dbea3869bc4b9efa14296d206badf1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a96499e2bc78fa089da78ffed48f964da"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#a96499e2bc78fa089da78ffed48f964da">set_num_mads</a> (double n=<a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a>)</td></tr>
<tr class="separator:a96499e2bc78fa089da78ffed48f964da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4fe13f22af3b265a86a1bb4591b698d8"><td class="memTemplParams" colspan="2">template&lt;typename Float , typename Integer &gt; </td></tr>
<tr class="memitem:a4fe13f22af3b265a86a1bb4591b698d8"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#a4fe13f22af3b265a86a1bb4591b698d8">run</a> (size_t n, const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Buffers.html">PerCellRnaQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;buffers) const</td></tr>
<tr class="separator:a4fe13f22af3b265a86a1bb4591b698d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae8478df615615eb407890c9c7f4e1bd3"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#ae8478df615615eb407890c9c7f4e1bd3">run</a> (const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Results.html">PerCellRnaQcMetrics::Results</a> &amp;metrics) const</td></tr>
<tr class="separator:ae8478df615615eb407890c9c7f4e1bd3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8d968ffd2e2140c01df22655d6a7dbac"><td class="memTemplParams" colspan="2">template&lt;typename Block , typename Float , typename Integer &gt; </td></tr>
<tr class="memitem:a8d968ffd2e2140c01df22655d6a7dbac"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#a8d968ffd2e2140c01df22655d6a7dbac">run_blocked</a> (size_t n, const Block *block, const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Buffers.html">PerCellRnaQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;buffers) const</td></tr>
<tr class="separator:a8d968ffd2e2140c01df22655d6a7dbac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac25c4bceca23127580b27221d3a87e8d"><td class="memTemplParams" colspan="2">template&lt;typename Block &gt; </td></tr>
<tr class="memitem:ac25c4bceca23127580b27221d3a87e8d"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#ac25c4bceca23127580b27221d3a87e8d">run_blocked</a> (const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Results.html">PerCellRnaQcMetrics::Results</a> &amp;metrics, const Block *block) const</td></tr>
<tr class="separator:ac25c4bceca23127580b27221d3a87e8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Create filters to identify low-quality cells from RNA-derived QC metrics. </p>
<p >Use an outlier-based approach on common QC metrics on the RNA data (see the <code><a class="el" href="classscran_1_1PerCellRnaQcMetrics.html" title="Compute typical per-cell quality control metrics from an RNA count matrix.">PerCellRnaQcMetrics</a></code> class) to identify low-quality cells. Specifically, low-quality cells are defined as those with:</p>
<ul>
<li>Low total counts, indicating that library preparation or sequencing depth was suboptimal.</li>
<li>Low numbers of detected features, a slightly different flavor of the above reasoning.</li>
<li>High proportions of counts in the mitochondrial (or spike-in) subsets, representing cell damage.</li>
</ul>
<p >Outliers are defined on each metric by counting the number of MADs from the median value across all cells. This assumes that most cells in the experiment are of high (or at least acceptable) quality; any anomalies are indicative of low-quality cells that should be filtered out. See the <code><a class="el" href="structscran_1_1ChooseOutlierFilters.html" title="Define outlier filters from the median and MAD.">ChooseOutlierFilters</a></code> class for implementation details.</p>
<p >For the total counts and number of detected features, the outliers are defined after log-transformation of the metrics. This improves resolution at low values and ensures that the defined threshold is not negative. Note that all thresholds are still reported on the original scale, so no further exponentiation is required. </p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="ac98f91f26febaaf787723642900e217a" name="ac98f91f26febaaf787723642900e217a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac98f91f26febaaf787723642900e217a">&#9670;&#160;</a></span>set_detected_num_mads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp; scran::SuggestRnaQcFilters::set_detected_num_mads </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of MADs below the median, to define the threshold for outliers in the number of detected features. This should be non-negative.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1SuggestRnaQcFilters.html" title="Create filters to identify low-quality cells from RNA-derived QC metrics.">SuggestRnaQcFilters</a></code> object. </dd></dl>

</div>
</div>
<a id="abfd616df329cc6b9a1171e3bf8969f8d" name="abfd616df329cc6b9a1171e3bf8969f8d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abfd616df329cc6b9a1171e3bf8969f8d">&#9670;&#160;</a></span>set_sums_num_mads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp; scran::SuggestRnaQcFilters::set_sums_num_mads </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of MADs below the median, to define the threshold for outliers in the total count per cell. This should be non-negative.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1SuggestRnaQcFilters.html" title="Create filters to identify low-quality cells from RNA-derived QC metrics.">SuggestRnaQcFilters</a></code> object. </dd></dl>

</div>
</div>
<a id="ac7dbea3869bc4b9efa14296d206badf1" name="ac7dbea3869bc4b9efa14296d206badf1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac7dbea3869bc4b9efa14296d206badf1">&#9670;&#160;</a></span>set_subset_num_mads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp; scran::SuggestRnaQcFilters::set_subset_num_mads </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of MADs above the median, to define the threshold for outliers in the subset proportions. This should be non-negative.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1SuggestRnaQcFilters.html" title="Create filters to identify low-quality cells from RNA-derived QC metrics.">SuggestRnaQcFilters</a></code> object. </dd></dl>

</div>
</div>
<a id="a96499e2bc78fa089da78ffed48f964da" name="a96499e2bc78fa089da78ffed48f964da"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a96499e2bc78fa089da78ffed48f964da">&#9670;&#160;</a></span>set_num_mads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1SuggestRnaQcFilters.html">SuggestRnaQcFilters</a> &amp; scran::SuggestRnaQcFilters::set_num_mads </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>n</em> = <code><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Defaults.html#a89a74362f4c4b8bc7fab9ce2bd54aa6c">Defaults::num_mads</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n</td><td>Number of MADs from the median, overriding previous calls to <code><a class="el" href="classscran_1_1SuggestRnaQcFilters.html#abfd616df329cc6b9a1171e3bf8969f8d">set_sums_num_mads()</a></code> and counterparts. This should be non-negative.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to this <code><a class="el" href="classscran_1_1SuggestRnaQcFilters.html" title="Create filters to identify low-quality cells from RNA-derived QC metrics.">SuggestRnaQcFilters</a></code> object. </dd></dl>

</div>
</div>
<a id="a4fe13f22af3b265a86a1bb4591b698d8" name="a4fe13f22af3b265a86a1bb4591b698d8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4fe13f22af3b265a86a1bb4591b698d8">&#9670;&#160;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Float , typename Integer &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestRnaQcFilters::run </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Buffers.html">PerCellRnaQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;&#160;</td>
          <td class="paramname"><em>buffers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Float</td><td>Floating point type for the metrics. </td></tr>
    <tr><td class="paramname">Integer</td><td>Integer for the metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">buffers</td><td>Pointers to arrays of length <code>n</code>, containing the per-cell RNA-derived metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric. </dd></dl>

</div>
</div>
<a id="ae8478df615615eb407890c9c7f4e1bd3" name="ae8478df615615eb407890c9c7f4e1bd3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae8478df615615eb407890c9c7f4e1bd3">&#9670;&#160;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestRnaQcFilters::run </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Results.html">PerCellRnaQcMetrics::Results</a> &amp;&#160;</td>
          <td class="paramname"><em>metrics</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">metrics</td><td>Collection of arrays of length equal to the number of cells, containing the per-cell RNA-derived metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric. </dd></dl>

</div>
</div>
<a id="a8d968ffd2e2140c01df22655d6a7dbac" name="a8d968ffd2e2140c01df22655d6a7dbac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8d968ffd2e2140c01df22655d6a7dbac">&#9670;&#160;</a></span>run_blocked() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Block , typename Float , typename Integer &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestRnaQcFilters::run_blocked </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block *&#160;</td>
          <td class="paramname"><em>block</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Buffers.html">PerCellRnaQcMetrics::Buffers</a>&lt; Float, Integer &gt; &amp;&#160;</td>
          <td class="paramname"><em>buffers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Block</td><td>Integer type for the block assignments. </td></tr>
    <tr><td class="paramname">Float</td><td>Floating point type for the metrics. </td></tr>
    <tr><td class="paramname">Integer</td><td>Integer for the metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">n</td><td>Number of cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length <code>n</code>, containing the block assignments for each cell. This may be <code>NULL</code>, in which case all cells are assumed to belong to the same block. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">buffers</td><td>Pointers to arrays of length <code>n</code>, containing the per-cell RNA-derived metrics.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric in each block. </dd></dl>

</div>
</div>
<a id="ac25c4bceca23127580b27221d3a87e8d" name="ac25c4bceca23127580b27221d3a87e8d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac25c4bceca23127580b27221d3a87e8d">&#9670;&#160;</a></span>run_blocked() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Block &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1SuggestRnaQcFilters_1_1Thresholds.html">Thresholds</a> scran::SuggestRnaQcFilters::run_blocked </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structscran_1_1PerCellRnaQcMetrics_1_1Results.html">PerCellRnaQcMetrics::Results</a> &amp;&#160;</td>
          <td class="paramname"><em>metrics</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Block *&#160;</td>
          <td class="paramname"><em>block</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts. </p><dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">Block</td><td>Integer type for the block assignments.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">metrics</td><td>Collection of arrays of length equal to the number of cells, containing the per-cell RNA-derived metrics. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">block</td><td>Pointer to an array of length equal to the number of cells, containing the block assignments for each cell. This may be <code>NULL</code>, in which case all cells are assumed to belong to the same block.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Filtering thresholds for each metric in each block. </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/quality_control/<a class="el" href="SuggestRnaQcFilters_8hpp_source.html">SuggestRnaQcFilters.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.5
</small></address>
</body>
</html>
