<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran::GroupedSizeFactors Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><b>scran</b></li><li class="navelem"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="classscran_1_1GroupedSizeFactors-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">scran::GroupedSizeFactors Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Compute grouped size factors to handle composition bias.  
 <a href="classscran_1_1GroupedSizeFactors.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="GroupedSizeFactors_8hpp_source.html">GroupedSizeFactors.hpp</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html">Defaults</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter settings.  <a href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Result of the size factor calculations.  <a href="structscran_1_1GroupedSizeFactors_1_1Results.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a907245f06f970a65c5a7c1a4a96af0fc"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a907245f06f970a65c5a7c1a4a96af0fc">set_center</a> (bool c=<a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#acbe5c154e3b67420a1f2ef84ef304310">Defaults::center</a>)</td></tr>
<tr class="separator:a907245f06f970a65c5a7c1a4a96af0fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0d4fdd8307e4d1a40709742eac394b76"><td class="memTemplParams" colspan="2">template&lt;typename T , typename IDX , typename Group , typename Out &gt; </td></tr>
<tr class="memitem:a0d4fdd8307e4d1a40709742eac394b76"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a0d4fdd8307e4d1a40709742eac394b76">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Group *group, Out *output)</td></tr>
<tr class="separator:a0d4fdd8307e4d1a40709742eac394b76"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3d64a2f29004bdb73bf03a93a4007ad9"><td class="memTemplParams" colspan="2">template&lt;typename Out  = double, typename T , typename IDX , typename Group &gt; </td></tr>
<tr class="memitem:a3d64a2f29004bdb73bf03a93a4007ad9"><td class="memTemplItemLeft" align="right" valign="top"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a>&lt; Out &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classscran_1_1GroupedSizeFactors.html#a3d64a2f29004bdb73bf03a93a4007ad9">run</a> (const tatami::Matrix&lt; T, IDX &gt; *mat, const Group *group)</td></tr>
<tr class="separator:a3d64a2f29004bdb73bf03a93a4007ad9"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Compute grouped size factors to handle composition bias. </p>
<p >This implements the grouping approach described in Lun et al. (2016) whereby groups/clusters of cells are used to construct pseudo-cells. These pseudo-cells are normalized against each other using median-based size factors (see <code><a class="el" href="classscran_1_1MedianSizeFactors.html" title="Compute median-based size factors to handle composition bias.">MedianSizeFactors</a></code>) to obtain group-specific scaling factors. Each cell is then normalized against its pseudo-cell using the library size; each cell's size factor is defined as product of its library size-based factor and the median-based factor for its group.</p>
<p >This strategy leverages the reduced sparsity of the pseudo-cells to obtain sensible median-based size factors for removing composition biases, while still generating per-cell factors for computing a normalized single-cell expression matrix in <code><a class="el" href="classscran_1_1LogNormCounts.html" title="Compute log-normalized expression values.">LogNormCounts</a></code>. The assumption is that there are no composition biases within each group; thus, the supplied groupings should correspond to subpopulations in the data, typically generated by clustering.</p>
<dl class="section see"><dt>See also</dt><dd>Lun ATL, Bach K and Marioni JC (2016). Pooling across cells to normalize single-cell RNA sequencing data with many zero counts. <em>Genome Biol.</em> 17:75 </dd></dl>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a3d64a2f29004bdb73bf03a93a4007ad9" name="a3d64a2f29004bdb73bf03a93a4007ad9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3d64a2f29004bdb73bf03a93a4007ad9">&#9670;&nbsp;</a></span>run() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename Out  = double, typename T , typename IDX , typename Group &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html">Results</a>&lt; Out &gt; scran::GroupedSizeFactors::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Group *&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Compute per-cell size factors based on user-supplied groupings.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Numeric data type of the input matrix. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer index type of the input matrix. </td></tr>
    <tr><td class="paramname">Group</td><td>Integer type for the groupings.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Matrix containing non-negative expression data, usually counts. Rows should be genes and columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">group</td><td>Pointer to an array of group identifiers, of length equal to the number of columns in <code>mat</code>. Values should be integers in <img class="formulaInl" alt="$[0, N)$" src="form_0.png"/> where <img class="formulaInl" alt="$N$" src="form_1.png"/> is the total number of groups.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Results.html" title="Result of the size factor calculations.">Results</a></code> object is returned containing the size factors. </dd></dl>

</div>
</div>
<a id="a0d4fdd8307e4d1a40709742eac394b76" name="a0d4fdd8307e4d1a40709742eac394b76"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0d4fdd8307e4d1a40709742eac394b76">&#9670;&nbsp;</a></span>run() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename T , typename IDX , typename Group , typename Out &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void scran::GroupedSizeFactors::run </td>
          <td>(</td>
          <td class="paramtype">const tatami::Matrix&lt; T, IDX &gt; *&#160;</td>
          <td class="paramname"><em>mat</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const Group *&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Out *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p >Compute per-cell size factors based on user-supplied groupings.</p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">T</td><td>Numeric data type of the input matrix. </td></tr>
    <tr><td class="paramname">IDX</td><td>Integer index type of the input matrix. </td></tr>
    <tr><td class="paramname">Group</td><td>Integer type for the groupings. </td></tr>
    <tr><td class="paramname">Out</td><td>Numeric data type of the output vector.</td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">mat</td><td>Matrix containing non-negative expression data, usually counts. Rows should be genes and columns should be cells. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">group</td><td>Pointer to an array of group identifiers, of length equal to the number of columns in <code>mat</code>. Values should be integers in <img class="formulaInl" alt="$[0, N)$" src="form_0.png"/> where <img class="formulaInl" alt="$N$" src="form_1.png"/> is the total number of groups. </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">output</td><td>Pointer to an array to use to store the output size factors. This should be of length equal to the number of columns in <code>mat</code>.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><code>output</code> is filled with the size factors for all cells in <code>mat</code>. </dd></dl>

</div>
</div>
<a id="a907245f06f970a65c5a7c1a4a96af0fc" name="a907245f06f970a65c5a7c1a4a96af0fc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a907245f06f970a65c5a7c1a4a96af0fc">&#9670;&nbsp;</a></span>set_center()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classscran_1_1GroupedSizeFactors.html">GroupedSizeFactors</a> &amp; scran::GroupedSizeFactors::set_center </td>
          <td>(</td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>c</em> = <code><a class="el" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html#acbe5c154e3b67420a1f2ef84ef304310">Defaults::center</a></code></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>Whether to center the size factors to have a mean of unity. This is usually desirable for interpretation of relative values.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A reference to this <code><a class="el" href="classscran_1_1MedianSizeFactors.html" title="Compute median-based size factors to handle composition bias.">MedianSizeFactors</a></code> object.</dd></dl>
<p>For more control over centering, this can be set to <code>false</code> and the resulting size factors can be passed to <code><a class="el" href="classscran_1_1CenterSizeFactors.html" title="Center size factors prior to scaling normalization.">CenterSizeFactors</a></code>. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>scran/normalization/<a class="el" href="GroupedSizeFactors_8hpp_source.html">GroupedSizeFactors.hpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
