<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.5"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: Class Hierarchy</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.5 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div class="header">
  <div class="headertitle"><div class="title">Class Hierarchy</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock">
<p><a href="hierarchy.html">Go to the textual class hierarchy</a></p>
</div><table border="0" cellspacing="10" cellpadding="0">
<tr><td><img src="inherit_graph_0.png" border="0" usemap="#ascran_1_1AggregateAcrossCells" alt=""/>
<map name="ascran_1_1AggregateAcrossCells" id="ascran_1_1AggregateAcrossCells">
<area shape="rect" href="classscran_1_1AggregateAcrossCells.html" title="Aggregate expression values across cells." coords="5,5,195,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_1.png" border="0" usemap="#ascran_1_1AggregateAcrossCells_1_1Combinations_3_01Factor_01_4" alt=""/>
<map name="ascran_1_1AggregateAcrossCells_1_1Combinations_3_01Factor_01_4" id="ascran_1_1AggregateAcrossCells_1_1Combinations_3_01Factor_01_4">
<area shape="rect" href="structscran_1_1AggregateAcrossCells_1_1Combinations.html" title="Unique combinations of factors." coords="5,5,195,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_2.png" border="0" usemap="#ascran_1_1AggregateAcrossCells_1_1Defaults" alt=""/>
<map name="ascran_1_1AggregateAcrossCells_1_1Defaults" id="ascran_1_1AggregateAcrossCells_1_1Defaults">
<area shape="rect" href="structscran_1_1AggregateAcrossCells_1_1Defaults.html" title="Default parameters for aggregation." coords="5,5,195,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_3.png" border="0" usemap="#ascran_1_1AggregateAcrossCells_1_1Results_3_01Sum_00_01Detected_01_4" alt=""/>
<map name="ascran_1_1AggregateAcrossCells_1_1Results_3_01Sum_00_01Detected_01_4" id="ascran_1_1AggregateAcrossCells_1_1Results_3_01Sum_00_01Detected_01_4">
<area shape="rect" href="structscran_1_1AggregateAcrossCells_1_1Results.html" title="Container for the aggregation results." coords="5,5,195,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_4.png" border="0" usemap="#ascran_1_1BlockedPCA" alt=""/>
<map name="ascran_1_1BlockedPCA" id="ascran_1_1BlockedPCA">
<area shape="rect" href="classscran_1_1BlockedPCA.html" title="Compute PCA after blocking on an uninteresting factor." coords="5,5,140,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_5.png" border="0" usemap="#ascran_1_1BlockedPCA_1_1Defaults" alt=""/>
<map name="ascran_1_1BlockedPCA_1_1Defaults" id="ascran_1_1BlockedPCA_1_1Defaults">
<area shape="rect" href="structscran_1_1BlockedPCA_1_1Defaults.html" title="Default parameter settings." coords="5,5,145,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_6.png" border="0" usemap="#ascran_1_1BlockedPCA_1_1Results" alt=""/>
<map name="ascran_1_1BlockedPCA_1_1Results" id="ascran_1_1BlockedPCA_1_1Results">
<area shape="rect" href="structscran_1_1BlockedPCA_1_1Results.html" title="Container for the PCA results." coords="5,5,145,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_7.png" border="0" usemap="#ascran_1_1CenterSizeFactors" alt=""/>
<map name="ascran_1_1CenterSizeFactors" id="ascran_1_1CenterSizeFactors">
<area shape="rect" href="classscran_1_1CenterSizeFactors.html" title="Center size factors prior to scaling normalization." coords="5,5,172,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_8.png" border="0" usemap="#ascran_1_1CenterSizeFactors_1_1Defaults" alt=""/>
<map name="ascran_1_1CenterSizeFactors_1_1Defaults" id="ascran_1_1CenterSizeFactors_1_1Defaults">
<area shape="rect" href="structscran_1_1CenterSizeFactors_1_1Defaults.html" title="Default parameter settings." coords="5,5,172,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_9.png" border="0" usemap="#ascran_1_1ChooseHVGs" alt=""/>
<map name="ascran_1_1ChooseHVGs" id="ascran_1_1ChooseHVGs">
<area shape="rect" href="classscran_1_1ChooseHVGs.html" title="Choose highly variable genes for downstream analyses." coords="5,5,145,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_10.png" border="0" usemap="#ascran_1_1ChooseHVGs_1_1Defaults" alt=""/>
<map name="ascran_1_1ChooseHVGs_1_1Defaults" id="ascran_1_1ChooseHVGs_1_1Defaults">
<area shape="rect" href="structscran_1_1ChooseHVGs_1_1Defaults.html" title="Default paramater settings." coords="5,5,152,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_11.png" border="0" usemap="#ascran_1_1ClusterSNNGraphLeiden" alt=""/>
<map name="ascran_1_1ClusterSNNGraphLeiden" id="ascran_1_1ClusterSNNGraphLeiden">
<area shape="rect" href="classscran_1_1ClusterSNNGraphLeiden.html" title="Leiden clustering on a shared nearest&#45;neighbor graph." coords="5,5,209,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_12.png" border="0" usemap="#ascran_1_1ClusterSNNGraphLeiden_1_1Defaults" alt=""/>
<map name="ascran_1_1ClusterSNNGraphLeiden_1_1Defaults" id="ascran_1_1ClusterSNNGraphLeiden_1_1Defaults">
<area shape="rect" href="structscran_1_1ClusterSNNGraphLeiden_1_1Defaults.html" title="Default parameter settings." coords="5,5,209,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_13.png" border="0" usemap="#ascran_1_1ClusterSNNGraphLeiden_1_1Results" alt=""/>
<map name="ascran_1_1ClusterSNNGraphLeiden_1_1Results" id="ascran_1_1ClusterSNNGraphLeiden_1_1Results">
<area shape="rect" href="structscran_1_1ClusterSNNGraphLeiden_1_1Results.html" title="Result of the igraph leiden community detection algorithm." coords="5,5,209,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_14.png" border="0" usemap="#ascran_1_1ClusterSNNGraphMultiLevel" alt=""/>
<map name="ascran_1_1ClusterSNNGraphMultiLevel" id="ascran_1_1ClusterSNNGraphMultiLevel">
<area shape="rect" href="classscran_1_1ClusterSNNGraphMultiLevel.html" title="Multi&#45;level clustering on a shared nearest&#45;neighbor graph." coords="5,5,199,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_15.png" border="0" usemap="#ascran_1_1ClusterSNNGraphMultiLevel_1_1Defaults" alt=""/>
<map name="ascran_1_1ClusterSNNGraphMultiLevel_1_1Defaults" id="ascran_1_1ClusterSNNGraphMultiLevel_1_1Defaults">
<area shape="rect" href="structscran_1_1ClusterSNNGraphMultiLevel_1_1Defaults.html" title="Default parameter settings." coords="5,5,199,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_16.png" border="0" usemap="#ascran_1_1ClusterSNNGraphMultiLevel_1_1Results" alt=""/>
<map name="ascran_1_1ClusterSNNGraphMultiLevel_1_1Results" id="ascran_1_1ClusterSNNGraphMultiLevel_1_1Results">
<area shape="rect" href="structscran_1_1ClusterSNNGraphMultiLevel_1_1Results.html" title="Result of the igraph multi&#45;level community detection algorithm." coords="5,5,199,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_17.png" border="0" usemap="#ascran_1_1ClusterSNNGraphWalktrap" alt=""/>
<map name="ascran_1_1ClusterSNNGraphWalktrap" id="ascran_1_1ClusterSNNGraphWalktrap">
<area shape="rect" href="classscran_1_1ClusterSNNGraphWalktrap.html" title="Walktrap clustering on a shared nearest&#45;neighbor graph." coords="5,5,221,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_18.png" border="0" usemap="#ascran_1_1ClusterSNNGraphWalktrap_1_1Defaults" alt=""/>
<map name="ascran_1_1ClusterSNNGraphWalktrap_1_1Defaults" id="ascran_1_1ClusterSNNGraphWalktrap_1_1Defaults">
<area shape="rect" href="structscran_1_1ClusterSNNGraphWalktrap_1_1Defaults.html" title="Default parameter settings." coords="5,5,221,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_19.png" border="0" usemap="#ascran_1_1ClusterSNNGraphWalktrap_1_1Results" alt=""/>
<map name="ascran_1_1ClusterSNNGraphWalktrap_1_1Results" id="ascran_1_1ClusterSNNGraphWalktrap_1_1Results">
<area shape="rect" href="structscran_1_1ClusterSNNGraphWalktrap_1_1Results.html" title="Result of the igraph Walktrap community detection algorithm." coords="5,5,221,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_20.png" border="0" usemap="#ascran_1_1DownsampleByNeighbors" alt=""/>
<map name="ascran_1_1DownsampleByNeighbors" id="ascran_1_1DownsampleByNeighbors">
<area shape="rect" href="classscran_1_1DownsampleByNeighbors.html" title="Downsample a dataset based on its neighbors." coords="5,5,215,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_21.png" border="0" usemap="#ascran_1_1DownsampleByNeighbors_1_1Defaults" alt=""/>
<map name="ascran_1_1DownsampleByNeighbors_1_1Defaults" id="ascran_1_1DownsampleByNeighbors_1_1Defaults">
<area shape="rect" href="structscran_1_1DownsampleByNeighbors_1_1Defaults.html" title="Default parameter settings." coords="5,5,215,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_22.png" border="0" usemap="#ascran_1_1FilterCells" alt=""/>
<map name="ascran_1_1FilterCells" id="ascran_1_1FilterCells">
<area shape="rect" href="classscran_1_1FilterCells.html" title="Filter out low&#45;quality cells." coords="5,5,124,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_23.png" border="0" usemap="#ascran_1_1FitTrendVar" alt=""/>
<map name="ascran_1_1FitTrendVar" id="ascran_1_1FitTrendVar">
<area shape="rect" href="classscran_1_1FitTrendVar.html" title="Fit a mean&#45;variance trend to log&#45;count data." coords="5,5,133,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_24.png" border="0" usemap="#ascran_1_1FitTrendVar_1_1Results" alt=""/>
<map name="ascran_1_1FitTrendVar_1_1Results" id="ascran_1_1FitTrendVar_1_1Results">
<area shape="rect" href="structscran_1_1FitTrendVar_1_1Results.html" title="Results of the trend fit." coords="5,5,133,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_25.png" border="0" usemap="#ascran_1_1GroupedSizeFactors" alt=""/>
<map name="ascran_1_1GroupedSizeFactors" id="ascran_1_1GroupedSizeFactors">
<area shape="rect" href="classscran_1_1GroupedSizeFactors.html" title="Compute grouped size factors to handle composition bias." coords="5,5,184,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_26.png" border="0" usemap="#ascran_1_1GroupedSizeFactors_1_1Defaults" alt=""/>
<map name="ascran_1_1GroupedSizeFactors_1_1Defaults" id="ascran_1_1GroupedSizeFactors_1_1Defaults">
<area shape="rect" href="structscran_1_1GroupedSizeFactors_1_1Defaults.html" title="Default parameter settings." coords="5,5,184,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_27.png" border="0" usemap="#ascran_1_1GroupedSizeFactors_1_1Results_3_01Out_01_4" alt=""/>
<map name="ascran_1_1GroupedSizeFactors_1_1Results_3_01Out_01_4" id="ascran_1_1GroupedSizeFactors_1_1Results_3_01Out_01_4">
<area shape="rect" href="structscran_1_1GroupedSizeFactors_1_1Results.html" title="Result of the size factor calculations." coords="5,5,184,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_28.png" border="0" usemap="#ascran_1_1IsOutlier" alt=""/>
<map name="ascran_1_1IsOutlier" id="ascran_1_1IsOutlier">
<area shape="rect" href="classscran_1_1IsOutlier.html" title="Define outlier values based on MADs." coords="5,5,115,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_29.png" border="0" usemap="#ascran_1_1IsOutlier_1_1Results_3_01X_01_4" alt=""/>
<map name="ascran_1_1IsOutlier_1_1Results_3_01X_01_4" id="ascran_1_1IsOutlier_1_1Results_3_01X_01_4">
<area shape="rect" href="structscran_1_1IsOutlier_1_1Results.html" title="Results of the outlier identification." coords="5,5,196,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_30.png" border="0" usemap="#ascran_1_1IsOutlier_1_1Thresholds" alt=""/>
<map name="ascran_1_1IsOutlier_1_1Thresholds" id="ascran_1_1IsOutlier_1_1Thresholds">
<area shape="rect" href="structscran_1_1IsOutlier_1_1Thresholds.html" title="Thresholds to define outliers on each metric." coords="5,5,184,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_31.png" border="0" usemap="#ascran_1_1LogNormCounts" alt=""/>
<map name="ascran_1_1LogNormCounts" id="ascran_1_1LogNormCounts">
<area shape="rect" href="classscran_1_1LogNormCounts.html" title="Compute log&#45;normalized expression values." coords="5,5,160,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_32.png" border="0" usemap="#ascran_1_1LogNormCounts_1_1Defaults" alt=""/>
<map name="ascran_1_1LogNormCounts_1_1Defaults" id="ascran_1_1LogNormCounts_1_1Defaults">
<area shape="rect" href="structscran_1_1LogNormCounts_1_1Defaults.html" title="Default parameter settings." coords="5,5,160,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_33.png" border="0" usemap="#ascran_1_1MedianSizeFactors" alt=""/>
<map name="ascran_1_1MedianSizeFactors" id="ascran_1_1MedianSizeFactors">
<area shape="rect" href="classscran_1_1MedianSizeFactors.html" title="Compute median&#45;based size factors to handle composition bias." coords="5,5,176,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_34.png" border="0" usemap="#ascran_1_1MedianSizeFactors_1_1Defaults" alt=""/>
<map name="ascran_1_1MedianSizeFactors_1_1Defaults" id="ascran_1_1MedianSizeFactors_1_1Defaults">
<area shape="rect" href="structscran_1_1MedianSizeFactors_1_1Defaults.html" title="Default parameter settings." coords="5,5,176,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_35.png" border="0" usemap="#ascran_1_1MedianSizeFactors_1_1Results_3_01Out_01_4" alt=""/>
<map name="ascran_1_1MedianSizeFactors_1_1Results_3_01Out_01_4" id="ascran_1_1MedianSizeFactors_1_1Results_3_01Out_01_4">
<area shape="rect" href="structscran_1_1MedianSizeFactors_1_1Results.html" title="Result of the size factor calculation." coords="5,5,176,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_36.png" border="0" usemap="#ascran_1_1ModelGeneVar" alt=""/>
<map name="ascran_1_1ModelGeneVar" id="ascran_1_1ModelGeneVar">
<area shape="rect" href="classscran_1_1ModelGeneVar.html" title="Compute and model the per&#45;gene variances in log&#45;expression data." coords="5,5,152,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_37.png" border="0" usemap="#ascran_1_1ModelGeneVar_1_1Defaults" alt=""/>
<map name="ascran_1_1ModelGeneVar_1_1Defaults" id="ascran_1_1ModelGeneVar_1_1Defaults">
<area shape="rect" href="structscran_1_1ModelGeneVar_1_1Defaults.html" title="Default parameters for variance modelling." coords="5,5,152,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_38.png" border="0" usemap="#ascran_1_1ModelGeneVar_1_1Results" alt=""/>
<map name="ascran_1_1ModelGeneVar_1_1Results" id="ascran_1_1ModelGeneVar_1_1Results">
<area shape="rect" href="structscran_1_1ModelGeneVar_1_1Results.html" title="Results of the variance modelling." coords="5,5,152,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_39.png" border="0" usemap="#ascran_1_1MultiBatchPCA" alt=""/>
<map name="ascran_1_1MultiBatchPCA" id="ascran_1_1MultiBatchPCA">
<area shape="rect" href="classscran_1_1MultiBatchPCA.html" title="Compute PCA after adjusting for differences between batch sizes." coords="5,5,155,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_40.png" border="0" usemap="#ascran_1_1MultiBatchPCA_1_1Defaults" alt=""/>
<map name="ascran_1_1MultiBatchPCA_1_1Defaults" id="ascran_1_1MultiBatchPCA_1_1Defaults">
<area shape="rect" href="structscran_1_1MultiBatchPCA_1_1Defaults.html" title="Default parameter settings." coords="5,5,155,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_41.png" border="0" usemap="#ascran_1_1MultiBatchPCA_1_1Results" alt=""/>
<map name="ascran_1_1MultiBatchPCA_1_1Results" id="ascran_1_1MultiBatchPCA_1_1Results">
<area shape="rect" href="structscran_1_1MultiBatchPCA_1_1Results.html" title="Container for the PCA results." coords="5,5,155,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_42.png" border="0" usemap="#ascran_1_1PairwiseEffects" alt=""/>
<map name="ascran_1_1PairwiseEffects" id="ascran_1_1PairwiseEffects">
<area shape="rect" href="classscran_1_1PairwiseEffects.html" title="Compute pairwise effect size between groups of cells." coords="5,5,156,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_43.png" border="0" usemap="#ascran_1_1PairwiseEffects_1_1Defaults" alt=""/>
<map name="ascran_1_1PairwiseEffects_1_1Defaults" id="ascran_1_1PairwiseEffects_1_1Defaults">
<area shape="rect" href="structscran_1_1PairwiseEffects_1_1Defaults.html" title="Default parameter settings." coords="5,5,156,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_44.png" border="0" usemap="#ascran_1_1PairwiseEffects_1_1Results_3_01Stat_01_4" alt=""/>
<map name="ascran_1_1PairwiseEffects_1_1Results_3_01Stat_01_4" id="ascran_1_1PairwiseEffects_1_1Results_3_01Stat_01_4">
<area shape="rect" href="structscran_1_1PairwiseEffects_1_1Results.html" title="Pairwise effect size results." coords="5,5,156,45"/>
<area shape="rect" href="structscran_1_1PairwiseEffects_1_1ResultsWithMeans.html" title="Pairwise effect size results, with per&#45;group means." coords="204,5,387,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_45.png" border="0" usemap="#ascran_1_1PerCellAdtQcFilters" alt=""/>
<map name="ascran_1_1PerCellAdtQcFilters" id="ascran_1_1PerCellAdtQcFilters">
<area shape="rect" href="classscran_1_1PerCellAdtQcFilters.html" title="Create filters to identify low&#45;quality cells from ADT data." coords="5,5,180,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_46.png" border="0" usemap="#ascran_1_1PerCellAdtQcFilters_1_1Defaults" alt=""/>
<map name="ascran_1_1PerCellAdtQcFilters_1_1Defaults" id="ascran_1_1PerCellAdtQcFilters_1_1Defaults">
<area shape="rect" href="structscran_1_1PerCellAdtQcFilters_1_1Defaults.html" title="Default parameter settings." coords="5,5,180,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_47.png" border="0" usemap="#ascran_1_1PerCellAdtQcFilters_1_1Results_3_01X_01_4" alt=""/>
<map name="ascran_1_1PerCellAdtQcFilters_1_1Results_3_01X_01_4" id="ascran_1_1PerCellAdtQcFilters_1_1Results_3_01X_01_4">
<area shape="rect" href="structscran_1_1PerCellAdtQcFilters_1_1Results.html" title="Results of the QC filtering." coords="5,5,180,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_48.png" border="0" usemap="#ascran_1_1PerCellAdtQcFilters_1_1Thresholds" alt=""/>
<map name="ascran_1_1PerCellAdtQcFilters_1_1Thresholds" id="ascran_1_1PerCellAdtQcFilters_1_1Thresholds">
<area shape="rect" href="structscran_1_1PerCellAdtQcFilters_1_1Thresholds.html" title="Thresholds to define outliers on each metric." coords="5,5,180,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_49.png" border="0" usemap="#ascran_1_1PerCellAdtQcMetrics" alt=""/>
<map name="ascran_1_1PerCellAdtQcMetrics" id="ascran_1_1PerCellAdtQcMetrics">
<area shape="rect" href="classscran_1_1PerCellAdtQcMetrics.html" title="Compute per&#45;cell quality control metrics from an ADT count matrix." coords="5,5,188,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_50.png" border="0" usemap="#ascran_1_1PerCellAdtQcMetrics_1_1Defaults" alt=""/>
<map name="ascran_1_1PerCellAdtQcMetrics_1_1Defaults" id="ascran_1_1PerCellAdtQcMetrics_1_1Defaults">
<area shape="rect" href="structscran_1_1PerCellAdtQcMetrics_1_1Defaults.html" title="Default parameters for aggregation." coords="5,5,188,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_51.png" border="0" usemap="#ascran_1_1PerCellAdtQcMetrics_1_1Results" alt=""/>
<map name="ascran_1_1PerCellAdtQcMetrics_1_1Results" id="ascran_1_1PerCellAdtQcMetrics_1_1Results">
<area shape="rect" href="structscran_1_1PerCellAdtQcMetrics_1_1Results.html" title="Result store for QC metric calculations." coords="5,5,188,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_52.png" border="0" usemap="#ascran_1_1PerCellRnaQcFilters" alt=""/>
<map name="ascran_1_1PerCellRnaQcFilters" id="ascran_1_1PerCellRnaQcFilters">
<area shape="rect" href="classscran_1_1PerCellRnaQcFilters.html" title="Create filters to identify low&#45;quality cells from RNA&#45;derived metrics." coords="5,5,184,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_53.png" border="0" usemap="#ascran_1_1PerCellRnaQcFilters_1_1Results_3_01X_01_4" alt=""/>
<map name="ascran_1_1PerCellRnaQcFilters_1_1Results_3_01X_01_4" id="ascran_1_1PerCellRnaQcFilters_1_1Results_3_01X_01_4">
<area shape="rect" href="structscran_1_1PerCellRnaQcFilters_1_1Results.html" title="Results of the QC filtering." coords="5,5,184,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_54.png" border="0" usemap="#ascran_1_1PerCellRnaQcFilters_1_1Thresholds" alt=""/>
<map name="ascran_1_1PerCellRnaQcFilters_1_1Thresholds" id="ascran_1_1PerCellRnaQcFilters_1_1Thresholds">
<area shape="rect" href="structscran_1_1PerCellRnaQcFilters_1_1Thresholds.html" title="Thresholds to define outliers on each metric." coords="5,5,184,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_55.png" border="0" usemap="#ascran_1_1PerCellRnaQcMetrics" alt=""/>
<map name="ascran_1_1PerCellRnaQcMetrics" id="ascran_1_1PerCellRnaQcMetrics">
<area shape="rect" href="classscran_1_1PerCellRnaQcMetrics.html" title="Compute typical per&#45;cell quality control metrics from an RNA count matrix." coords="5,5,191,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_56.png" border="0" usemap="#ascran_1_1PerCellRnaQcMetrics_1_1Defaults" alt=""/>
<map name="ascran_1_1PerCellRnaQcMetrics_1_1Defaults" id="ascran_1_1PerCellRnaQcMetrics_1_1Defaults">
<area shape="rect" href="structscran_1_1PerCellRnaQcMetrics_1_1Defaults.html" title="Default parameter settings." coords="5,5,191,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_57.png" border="0" usemap="#ascran_1_1PerCellRnaQcMetrics_1_1Results" alt=""/>
<map name="ascran_1_1PerCellRnaQcMetrics_1_1Results" id="ascran_1_1PerCellRnaQcMetrics_1_1Results">
<area shape="rect" href="structscran_1_1PerCellRnaQcMetrics_1_1Results.html" title="Result store for QC metric calculations." coords="5,5,191,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_58.png" border="0" usemap="#ascran_1_1RunPCA" alt=""/>
<map name="ascran_1_1RunPCA" id="ascran_1_1RunPCA">
<area shape="rect" href="classscran_1_1RunPCA.html" title="Perform PCA on a gene&#45;cell matrix." coords="5,5,116,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_59.png" border="0" usemap="#ascran_1_1RunPCA_1_1Defaults" alt=""/>
<map name="ascran_1_1RunPCA_1_1Defaults" id="ascran_1_1RunPCA_1_1Defaults">
<area shape="rect" href="structscran_1_1RunPCA_1_1Defaults.html" title="Default parameter settings." coords="5,5,169,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_60.png" border="0" usemap="#ascran_1_1RunPCA_1_1Results" alt=""/>
<map name="ascran_1_1RunPCA_1_1Results" id="ascran_1_1RunPCA_1_1Results">
<area shape="rect" href="structscran_1_1RunPCA_1_1Results.html" title="Container for the PCA results." coords="5,5,165,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_61.png" border="0" usemap="#ascran_1_1ScaleByNeighbors" alt=""/>
<map name="ascran_1_1ScaleByNeighbors" id="ascran_1_1ScaleByNeighbors">
<area shape="rect" href="classscran_1_1ScaleByNeighbors.html" title="Scale multi&#45;modal embeddings to adjust for differences in variance." coords="5,5,172,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_62.png" border="0" usemap="#ascran_1_1ScaleByNeighbors_1_1Defaults" alt=""/>
<map name="ascran_1_1ScaleByNeighbors_1_1Defaults" id="ascran_1_1ScaleByNeighbors_1_1Defaults">
<area shape="rect" href="structscran_1_1ScaleByNeighbors_1_1Defaults.html" title="Default parameter settings." coords="5,5,172,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_63.png" border="0" usemap="#ascran_1_1ScoreMarkers" alt=""/>
<map name="ascran_1_1ScoreMarkers" id="ascran_1_1ScoreMarkers">
<area shape="rect" href="classscran_1_1ScoreMarkers.html" title="Score each gene as a candidate marker for each group of cells." coords="5,5,147,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_64.png" border="0" usemap="#ascran_1_1ScoreMarkers_1_1Defaults" alt=""/>
<map name="ascran_1_1ScoreMarkers_1_1Defaults" id="ascran_1_1ScoreMarkers_1_1Defaults">
<area shape="rect" href="structscran_1_1ScoreMarkers_1_1Defaults.html" title="Default parameter settings." coords="5,5,147,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_65.png" border="0" usemap="#ascran_1_1ScoreMarkers_1_1Results_3_01Stat_01_4" alt=""/>
<map name="ascran_1_1ScoreMarkers_1_1Results_3_01Stat_01_4" id="ascran_1_1ScoreMarkers_1_1Results_3_01Stat_01_4">
<area shape="rect" href="structscran_1_1ScoreMarkers_1_1Results.html" title="Results of the marker scoring." coords="5,5,147,45"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_66.png" border="0" usemap="#ascran_1_1SummarizeEffects" alt=""/>
<map name="ascran_1_1SummarizeEffects" id="ascran_1_1SummarizeEffects">
<area shape="rect" href="classscran_1_1SummarizeEffects.html" title="Summarize pairwise effects into summary statistics per group." coords="5,5,171,31"/>
</map>
</td></tr>
<tr><td><img src="inherit_graph_67.png" border="0" usemap="#ascran_1_1SummarizeEffects_1_1Defaults" alt=""/>
<map name="ascran_1_1SummarizeEffects_1_1Defaults" id="ascran_1_1SummarizeEffects_1_1Defaults">
<area shape="rect" href="structscran_1_1SummarizeEffects_1_1Defaults.html" title="Default parameter settings." coords="5,5,171,45"/>
</map>
</td></tr>
</table>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.5
</small></address>
</body>
</html>
