<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.6"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran: scran/utils/macros.hpp File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran
   </div>
   <div id="projectbrief">C++ library for basic single-cell RNA-seq analyses</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.6 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_bdabb9492db2368b3caf0f4c1df6a160.html">scran</a></li><li class="navelem"><a class="el" href="dir_9b9de28a72409b9bd6e78458203d0167.html">utils</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="headertitle"><div class="title">macros.hpp File Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Set common macros used through <b>libscran</b>.  
<a href="#details">More...</a></p>
<div class="textblock"><div class="dynheader">
This graph shows which files directly or indirectly include this file:</div>
<div class="dyncontent">
<div class="center"><img src="macros_8hpp__dep__incl.png" border="0" usemap="#ascran_2utils_2macros_8hppdep" alt=""/></div>
<map name="ascran_2utils_2macros_8hppdep" id="ascran_2utils_2macros_8hppdep">
<area shape="rect" title="Set common macros used through libscran." alt="" coords="1688,5,1841,31"/>
<area shape="rect" href="AggregateAcrossCells_8hpp.html" title="Aggregate expression values across cells." alt="" coords="17,181,205,221"/>
<area shape="rect" href="GroupedSizeFactors_8hpp.html" title="Compute size factors for groups of cells." alt="" coords="144,269,312,309"/>
<area shape="rect" href="scran_8hpp_source.html" title=" " alt="" coords="1619,445,1734,471"/>
<area shape="rect" href="DownsampleByNeighbors_8hpp.html" title="Downsample a dataset based on its neighbors." alt="" coords="3317,86,3521,126"/>
<area shape="rect" href="Factory_8hpp_source.html" title=" " alt="" coords="29,93,225,119"/>
<area shape="rect" href="HypergeometricTail_8hpp.html" title="Compute hypergeometric tail probabilities." alt="" coords="437,357,657,397"/>
<area shape="rect" href="ScoreFeatureSet_8hpp.html" title="Compute per&#45;cell scores for a feature set." alt="" coords="681,357,888,397"/>
<area shape="rect" href="BuildSNNGraph_8hpp_source.html" title=" " alt="" coords="819,189,1054,214"/>
<area shape="rect" href="ClusterSNNGraph_8hpp.html" title="Identify clusters of cells from a shared nearest&#45;neighbor graph." alt="" coords="985,269,1144,309"/>
<area shape="rect" href="ClusterKmeans_8hpp.html" title="Identify clusters of cells with k&#45;means." alt="" coords="1014,357,1173,397"/>
<area shape="rect" href="igraph__utils_8hpp.html" title="Utilities for manipulating igraph data structures." alt="" coords="961,86,1114,126"/>
<area shape="rect" href="PairwiseEffects_8hpp.html" title="Compute pairwise effect sizes between groups of cells." alt="" coords="2502,181,2696,221"/>
<area shape="rect" href="ScoreMarkers_8hpp.html" title="Compute marker scores for each gene in each group of cells." alt="" coords="2954,181,3140,221"/>
<area shape="rect" href="SummarizeEffects_8hpp.html" title="Summarize the effect sizes from pairwise comparisons." alt="" coords="2721,181,2930,221"/>
<area shape="rect" href="simple__diff_8hpp_source.html" title=" " alt="" coords="2606,86,2773,126"/>
<area shape="rect" href="summarize__comparisons_8hpp.html" title="Utilities for effect summarization." alt="" coords="2849,79,2992,133"/>
<area shape="rect" href="BlockedPCA_8hpp.html" title="Compute PCA after blocking on an uninteresting factor." alt="" coords="1198,357,1382,397"/>
<area shape="rect" href="RunPCA_8hpp.html" title="Perform PCA on a gene&#45;by&#45;cell matrix." alt="" coords="1291,269,1451,309"/>
<area shape="rect" href="ScaleByNeighbors_8hpp.html" title="Scale multi&#45;modal embeddings based on their relative variance." alt="" coords="3545,86,3762,126"/>
<area shape="rect" href="ChooseHVGs_8hpp.html" title="Choose highly variable genes for downstream analyses." alt="" coords="1456,357,1614,397"/>
<area shape="rect" href="FitTrendVar_8hpp.html" title="Fit a mean&#45;variance trend to log&#45;count data." alt="" coords="1494,86,1651,126"/>
<area shape="rect" href="ModelGeneVar_8hpp.html" title="Model the per&#45;gene variance from log&#45;count data." alt="" coords="1544,181,1702,221"/>
<area shape="rect" href="CenterSizeFactors_8hpp.html" title="Center size factors prior to normalization." alt="" coords="386,86,543,126"/>
<area shape="rect" href="LogNormCounts_8hpp.html" title="Compute log&#45;normalized expression values." alt="" coords="538,181,684,221"/>
<area shape="rect" href="MedianSizeFactors_8hpp.html" title="Compute median&#45;based size factors." alt="" coords="301,181,462,221"/>
<area shape="rect" href="ChooseOutlierFilters_8hpp.html" title="Define outlier filters using a MAD&#45;based approach." alt="" coords="1728,269,1896,309"/>
<area shape="rect" href="SuggestAdtQcFilters_8hpp.html" title="Create filters to identify low&#45;quality cells from ADT&#45;derived QC metrics." alt="" coords="1950,357,2120,397"/>
<area shape="rect" href="SuggestCrisprQcFilters_8hpp.html" title="Create filters to identify low&#45;quality cells from CRISPR&#45;derived QC metrics." alt="" coords="1740,357,1925,397"/>
<area shape="rect" href="SuggestRnaQcFilters_8hpp.html" title="Create filters to identify low&#45;quality cells from RNA&#45;derived QC metrics." alt="" coords="2145,357,2320,397"/>
<area shape="rect" href="ComputeMedianMad_8hpp.html" title="Compute the median and MAD from an array of values." alt="" coords="1828,181,1997,221"/>
<area shape="rect" href="FilterCells_8hpp.html" title="Filter out low&#45;quality cells from the count matrix." alt="" coords="2902,269,3043,309"/>
<area shape="rect" href="PerCellAdtQcMetrics_8hpp.html" title="Compute per&#45;cell quality control metrics from an ADT count matrix." alt="" coords="2124,269,2296,309"/>
<area shape="rect" href="PerCellCrisprQcMetrics_8hpp.html" title="Compute per&#45;cell quality control metrics from a CRISPR guide count matrix." alt="" coords="2320,269,2507,309"/>
<area shape="rect" href="PerCellQcMetrics_8hpp.html" title="Compute a variety of per&#45;cell quality control metrics from a count matrix." alt="" coords="2224,181,2376,221"/>
<area shape="rect" href="PerCellRnaQcMetrics_8hpp_source.html" title=" " alt="" coords="2532,269,2706,309"/>
<area shape="rect" href="average__vectors_8hpp.html" title="Average parallel elements across vectors." alt="" coords="3096,269,3228,309"/>
<area shape="rect" href="subset__vector_8hpp_source.html" title=" " alt="" coords="3083,357,3208,397"/>
<area shape="rect" href="vector__to__pointers_8hpp.html" title="Create a vector of pointers from a vector of vectors." alt="" coords="2408,86,2531,126"/>
</map>
</div>
</div>
<p><a href="macros_8hpp_source.html">Go to the source code of this file.</a></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Set common macros used through <b>libscran</b>. </p>
<p>The <code>SCRAN_CUSTOM_PARALLEL</code> macro can be set to a function that specifies a custom parallelization scheme. This function should be a template that accept three arguments:</p>
<ul>
<li><code>njobs</code>, an integer specifying the number of jobs.</li>
<li><code>fun</code>, a lambda that accepts two arguments, <code>start</code> and <code>end</code>.</li>
<li><code>nthreads</code>, an integer specifying the number of threads to use.</li>
</ul>
<p>The function should split <code>[0, njobs)</code> into any number of contiguous, non-overlapping intervals, and call <code>fun</code> on each interval, possibly in different threads. The details of the splitting and evaluation are left to the discretion of the developer defining the macro. The function should only return once all evaluations of <code>fun</code> are complete.</p>
<p>If <code>SCRAN_CUSTOM_PARALLEL</code> is set, the following macros are also set (if they are not already defined):</p>
<ul>
<li><code>TATAMI_CUSTOM_PARALLEL</code>, from the <a href="https://ltla.github.io/tatami"><b>tatami</b></a> library.</li>
<li><code>KMEANS_CUSTOM_PARALLEL</code>, from the <a href="https://ltla.github.io/CppKmeans"><b>kmeans</b></a> library.</li>
<li><code>IRLBA_CUSTOM_PARALLEL</code>, from the <a href="https://ltla.github.io/CppIrlba"><b>irlba</b></a> library.</li>
</ul>
<p>This ensures that any custom parallelization scheme is propagated to all of <b>libscran</b>'s dependencies. If these libraries are used outside of <b>libscran</b>, some care is required to ensure that the macros are consistently defined through the client library/application; otherwise, developers may observe ODR compilation errors. </p>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.6
</small></address>
</body>
</html>
